---
title: "test vctrs"
author: "Andrea"
date: "2024-03-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
example_ind_meta <- data.frame (id=c("a","b","c"),
                       population = c("pop1","pop1","pop2"))
example_genotypes <- rbind(c(1, 1, 0, 1, 1, 0), 
                        c(2, 0, 0, 0,NA, 0), 
                        c(1, 2, 0, 0, 1, 1))
example_loci <- data.frame(name=c("rs1","rs2","rs3","rs4","x1","x2"),
                   chromosome=c(1,1,1,1,2,2),
                   position=c(3,5,65,343,23,456),
                   allele_ref = c("a","t","c","g","c","t"),
                   allele_alt = c("t","c", NA,"c","g","a"))
```

We create a plink BED file:
```{r}
bed_prefix <- tempfile('test_data_')
tidypopgen::make_test_bed(example_genotypes, example_loci, example_ind_meta, bed_prefix)
bed_path <- paste0(bed_prefix,".bed")

```

```{r}
bigsnp_path <- bigsnpr::snp_readBed(bed_path, backingfile = tempfile("test_bigfile_"))

```


```{r}
 genotypes_big <- bigsnpr::snp_attach(bigsnp_path)

```


```{r eval=FALSE}
new_vctrs_test <- function(x = character(), some_attr, another_attr=NULL) {

  vctrs::new_vctr(x, my_attr = some_attr, my_attr2=another_attr, class = "vctrs_test")
}
foo <- new_vctrs_test(1:3,c("blah blah"),genotypes_big)
```



```{r}
foo <- new_vctrs_big(1:3,genotypes_big,c("blah blah"))
```



Now use it in a tibble

```{r}
ind_meta<-list(id=paste0("id_",1:3), pop=rep("pop_1",3))
bar3 <- gen_tibble(ind_meta, foo)
```

```{r}
bar3

```


```{r eval=FALSE}
bar2 <-   tibble::new_tibble(
   list(col1 = 1:3,col2=foo,col3=paste0(1:3,"_a")),
    class = "gen_tbl2"
  )
bar2
```

```{r}
attr(bar2$col2,"bigsnp")

```

Given information about the individuals, their genotypes, and the loci:
