test_genotypes <- rbind(
  c(1, 1, 0, 1, 1, 0),
  c(2, 2, 0, NA, 0, 0),
  c(2, NA, 0, 0, 1, 1),
  c(2, 0, 0, 2, 0, 0),
  c(1, 2, 0, 1, 2, 1),
  c(0, 0, 0, 0, NA, 2),
  c(0, 1, 2, 0, 1, NA)
)
test_indiv_meta <- data.frame(
  id = c("a", "b", "c", "d", "e", "f", "g"),
  population = c("pop1", "pop1", "pop2", "pop2", "pop1", "pop3", "pop3")
)
test_loci <- data.frame(
  name = paste0("rs", 1:6),
  chromosome = paste0("chr", c(1, 1, 1, 1, 2, 2)),
  position = as.integer(c(3, 5, 65, 343, 23, 456)),
  genetic_dist = as.double(rep(0, 6)),
  allele_ref = c("A", "T", "C", "G", "C", "T"),
  allele_alt = c("T", "C", NA, "C", "G", "A")
)

test_gt <- gen_tibble(
  x = test_genotypes,
  loci = test_loci,
  indiv_meta = test_indiv_meta,
  quiet = TRUE
)

test_that("gt_pseudohaploid correctly deals with ploidy", {
  # confirm that the gen_tibble is diploid for the moment
  expect_equal(show_ploidy(test_gt), 2)
  # now use gt_pseudohaploid
  test_gt_pseudo <- gt_pseudohaploid(test_gt)
  # now check that ploidy is -2
  expect_equal(show_ploidy(test_gt_pseudo), -2)
  ## check the individual ploidies
  expect_equal(
    indiv_ploidy(test_gt_pseudo),
    c(2L, 1L, 2L, 1L, 2L, 1L, 2L)
  )

  # check that we can reprocess it (use a different number of loci to get different result)
  test_gt_pseudo2 <- gt_pseudohaploid(test_gt_pseudo, test_n_loci = 4)
  expect_equal(
    indiv_ploidy(test_gt_pseudo2),
    c(2L, 1L, 1L, 1L, 2L, 1L, 2L)
  )

  ## TODO now test which functions work and which fail with pseudohaploid data (we should get the right frequencies (grouped and ungrouped, and pairwise pop fst))
  # missingness shoudl also work
  # but most other indiv stats will fail, and so will pop estimates that requires heterozygote counts
  expect_error(
    indiv_het_obs(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    indiv_inbreeding(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    loci_hwe(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    pop_fis(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    pop_fst(test_gt_pseudo),
    "this function only works on diploid data"
  )
  # now check frequencies
  pseudo_freq <- loci_alt_freq(test_gt_pseudo)
  ploidy_hap <- c(2L, 1L, 2L, 1L, 2L, 1L, 2L)
  # convert genotypes to alt counts by ploidy
  test_genotypes_hap <- sweep(test_genotypes, 1,
    3 - ploidy_hap,
    FUN = "/"
  )
  test_valid_alleles <- matrix(ploidy_hap,
    ncol = ncol(test_genotypes_hap),
    nrow = nrow(test_genotypes_hap)
  )
  test_valid_alleles[is.na(test_genotypes_hap)] <- NA
  test_valid_alleles <- colSums(test_valid_alleles, na.rm = TRUE)
  expect_identical(
    pseudo_freq,
    colSums(test_genotypes_hap, na.rm = TRUE) /
      test_valid_alleles
  )
  counts <- loci_alt_freq(test_gt_pseudo, as_counts = TRUE)
  expect_true(
    all(
      cbind(
        colSums(test_genotypes_hap, na.rm = TRUE),
        test_valid_alleles
      ) ==
        loci_alt_freq(test_gt_pseudo, as_counts = TRUE)
    )
  )

  # and missingness
  expect_true(
    all(
      loci_missingness(test_gt_pseudo) ==
        colSums(is.na(test_genotypes_hap)) / nrow(test_genotypes_hap)
    )
  )
})


## now group it
test_gt <- test_gt %>% group_by(population)

test_that("gt_pseudohaploid on grouped tibble correctly deals with ploidy", {
  # confirm that the gen_tibble is diploid for the moment
  expect_equal(show_ploidy(test_gt), 2)
  # now use gt_pseudohaploid
  test_gt_pseudo <- gt_pseudohaploid(test_gt)
  # now check that ploidy is -2
  expect_equal(show_ploidy(test_gt_pseudo), -2)
  ## check the individual ploidies
  expect_equal(
    indiv_ploidy(test_gt_pseudo),
    c(2L, 1L, 2L, 1L, 2L, 1L, 2L)
  )

  # check that we can reprocess it (use a different number of loci to get different result)
  test_gt_pseudo2 <- gt_pseudohaploid(test_gt_pseudo, test_n_loci = 4)
  expect_equal(
    indiv_ploidy(test_gt_pseudo2),
    c(2L, 1L, 1L, 1L, 2L, 1L, 2L)
  )

  ## TODO now test which functions work and which fail with pseudohaploid data (we should get the right frequencies (grouped and ungrouped, and pairwise pop fst))
  # missingness shoudl also work
  # but most other indiv stats will fail, and so will pop estimates that requires heterozygote counts
  expect_error(
    indiv_het_obs(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    indiv_inbreeding(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    loci_hwe(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    pop_fis(test_gt_pseudo),
    "this function only works on diploid data"
  )
  expect_error(
    pop_fst(test_gt_pseudo),
    "this function only works on diploid data"
  )
  # now check frequencies
  pseudo_freq <- loci_alt_freq(test_gt_pseudo, type = "matrix")
  # # convert genotypes to alt counts by ploidy
  # 3-c(2L, 1L, 2L, 1L, 2L, 1L, 2L)
  #
  #
  # expect_equal(
  #   pseudo_freq$value[1:3],
  #   c(0.4285714, 0.2857143, 0.2857143, 0.2857143, 0.4285714, 0.2857143)
  # )
})
