<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Constructor for a gen_tibble — gen_tibble • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Constructor for a gen_tibble — gen_tibble"><meta name="description" content="A gen_tibble stores genotypes for individuals in a tidy format. DESCRIBE
here the format"><meta property="og:description" content="A gen_tibble stores genotypes for individuals in a tidy format. DESCRIBE
here the format"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Projection of aDNA samples onto modern PCA</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Constructor for a <code>gen_tibble</code></h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/dev/R/gen_tibble.R" class="external-link"><code>R/gen_tibble.R</code></a></small>
      <div class="d-none name"><code>gen_tibble.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A <code>gen_tibble</code> stores genotypes for individuals in a tidy format. DESCRIBE
here the format</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gen_tibble</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span><span class="op">)</span>,</span>
<span>  missing_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"."</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'character'</span></span>
<span><span class="fu">gen_tibble</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  parser <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cpp"</span>, <span class="st">"vcfR"</span><span class="op">)</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  chunk_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span><span class="op">)</span>,</span>
<span>  missing_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"."</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'matrix'</span></span>
<span><span class="fu">gen_tibble</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">indiv_meta</span>,</span>
<span>  <span class="va">loci</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span><span class="op">)</span>,</span>
<span>  missing_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"."</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>can be:</p><ul><li><p>a string giving the path to a PLINK BED or PED file. The associated
BIM and FAM files for the BED, or MAP for PED are expected to be in the
same directory and have the same file name.</p></li>
<li><p>a string giving the path to a RDS file storing a <code>bigSNP</code> object from
the <code>bigsnpr</code> package (usually created with <code><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">bigsnpr::snp_readBed()</a></code>)</p></li>
<li><p>a string giving the path to a vcf file. Note that we currently read the
whole vcf in memory with <code>vcfR</code>, so only smallish <em>VCF</em> can be imported.
Only biallelic SNPs will be considered.</p></li>
<li><p>a string giving the path to a <em>packedancestry</em> .geno file. The associated
.ind and .snp files are expected to be in the same directory and share the
same file name prefix.</p></li>
<li><p>a genotype matrix of dosages (0, 1, 2, NA) giving the dosage of the
alternate allele.</p></li>
</ul></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>if <code>x</code> is the name of a vcf file, additional arguments passed to
<code><a href="https://rdrr.io/pkg/vcfR/man/io_vcfR.html" class="external-link">vcfR::read.vcfR()</a></code>. Otherwise, unused.</p></dd>


<dt id="arg-valid-alleles">valid_alleles<a class="anchor" aria-label="anchor" href="#arg-valid-alleles"></a></dt>
<dd><p>a vector of valid allele values; it defaults to 'A','T',
'C' and 'G'.</p></dd>


<dt id="arg-missing-alleles">missing_alleles<a class="anchor" aria-label="anchor" href="#arg-missing-alleles"></a></dt>
<dd><p>a vector of values in the BIM file/loci dataframe that
indicate a missing value for the allele value (e.g. when we have a
monomorphic locus with only one allele). It defaults to '0' and '.' (the
same as PLINK 1.9).</p></dd>


<dt id="arg-backingfile">backingfile<a class="anchor" aria-label="anchor" href="#arg-backingfile"></a></dt>
<dd><p>the path, including the file name without extension, for
backing files used to store the data (they will be given a .bk and .RDS
automatically). This is not needed if <code>x</code> is already an .RDS file. If <code>x</code>
is a .BED or a <em>VCF</em> file and <code>backingfile</code> is left NULL, the backing file
will be saved in the same directory as the bed file, using the same file
name but with a different file type (.bk rather than .bed). If <code>x</code> is a
genotype matrix and <code>backingfile</code> is NULL, then a temporary file will be
created (but note that R will delete it at the end of the session!)</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>provide information on the files used to store the data</p></dd>


<dt id="arg-parser">parser<a class="anchor" aria-label="anchor" href="#arg-parser"></a></dt>
<dd><p>the name of the parser used for <em>VCF</em>, either "cpp" to use a
fast C++ parser (the default), or "vcfR" to use the R package <code>vcfR</code>. The
latter is slower but more robust; if "cpp" gives an error, try using "vcfR"
in case your
<em>VCF</em> has an unusual structure.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>the number of cores to use for parallel processing</p></dd>


<dt id="arg-chunk-size">chunk_size<a class="anchor" aria-label="anchor" href="#arg-chunk-size"></a></dt>
<dd><p>the number of loci or individuals (depending on the format)
processed at a time (currently used if <code>x</code> is a vcf with parser "vcfR")</p></dd>


<dt id="arg-indiv-meta">indiv_meta<a class="anchor" aria-label="anchor" href="#arg-indiv-meta"></a></dt>
<dd><p>a list, data.frame or tibble with compulsory columns 'id'
and 'population', plus any additional metadata of interest. This is only
used if <code>x</code> is a genotype matrix. Otherwise this information is extracted
directly from the files.</p></dd>


<dt id="arg-loci">loci<a class="anchor" aria-label="anchor" href="#arg-loci"></a></dt>
<dd><p>a data.frame or tibble, with compulsory columns 'name',
'chromosome', and 'position','genetic_dist', 'allele_ref' and 'allele_alt'.
This is only used if <code>x</code> is a genotype matrix. Otherwise this information
is extracted directly from the files.</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>the ploidy of the samples (either a single value, or
a vector of values for mixed ploidy). Only used if creating
a gen_tibble from a matrix of data; otherwise, ploidy is determined
automatically from the data as they are read.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of the class <code>gen_tbl</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<ul><li><p><em>VCF</em> files: the fast <code>cpp</code> parser is used by default. Both <code>cpp</code> and
<code>vcfR</code> parsers attempt to establish ploidy from the first variant; if that
variant is found in a sex chromosome (or mtDNA), the parser will fail with
'Error: a genotype has more than max_ploidy alleles...'. To successful import
such a <em>VCF</em>, change the order of variants so that the first chromosome is an
autosome using a tool such as <code>vcftools</code>. Currently, only biallelic SNPs are
supported. If haploid variants (e.g. sex chromosomes) are included in the
<em>VCF</em>, they are not transformed into homozygous calls. Instead, reference
alleles will be counted as 0 and alternative alleles will be counted as 1.</p></li>
<li><p><em>packedancestry</em> files: When loading <em>packedancestry</em> files,
missing alleles will be converted from 'X' to NA</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

