<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute and store blocked f2 statistics for ADMIXTOOLS 2 — gt_extract_f2 • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute and store blocked f2 statistics for ADMIXTOOLS 2 — gt_extract_f2"><meta property="og:description" content="This function prepares data for various ADMIXTOOLS 2 functions fro the
package ADMIXTOOLS 2. It
takes a gen_tibble,
computes allele frequencies and blocked f2-statistics,
and writes the results to outdir. It is equivalent to
admixtools::extract_f2()."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9014</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/a01_overview.html">overview</a>
    </li>
    <li>
      <a href="../articles/a02_qc.html">QC</a>
    </li>
    <li>
      <a href="../articles/a03_example_clustering_and_dapc.html">example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/a99_plink_cheatsheet.html">plink cheatsheet</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute and store blocked f2 statistics for ADMIXTOOLS 2</h1>
    
    <div class="hidden name"><code>gt_extract_f2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function prepares data for various <em>ADMIXTOOLS 2</em> functions fro the
package <em>ADMIXTOOLS 2</em>. It
takes a <code><a href="gen_tibble.html">gen_tibble</a></code>,
computes allele frequencies and blocked f2-statistics,
and writes the results to <code>outdir</code>. It is equivalent to
<code><a href="https://rdrr.io/pkg/admixtools/man/extract_f2.html" class="external-link">admixtools::extract_f2()</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gt_extract_f2</span><span class="op">(</span></span>
<span>  <span class="va">.x</span>,</span>
<span>  outdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blgsize <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  maxmem <span class="op">=</span> <span class="fl">8000</span>,</span>
<span>  maxmiss <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  minmaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxmaf <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  minac2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outpop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outpop_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transitions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transversions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  adjust_pseudohaploid <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fst <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  afprod <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  poly_only <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"f2"</span><span class="op">)</span>,</span>
<span>  apply_corr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>.x</dt>
<dd><p>a <code><a href="gen_tibble.html">gen_tibble</a></code></p></dd>


<dt>outdir</dt>
<dd><p>Directory where data will be stored.</p></dd>


<dt>blgsize</dt>
<dd><p>SNP block size in Morgan. Default is 0.05 (5 cM). If <code>blgsize</code>
is 100 or greater, if will be interpreted as base pair distance rather than
centimorgan distance.</p></dd>


<dt>maxmem</dt>
<dd><p>Maximum amount of memory to be used. If the required amount
of memory exceeds <code>maxmem</code>, allele frequency data will be split into blocks,
and the computation will be performed separately on each block pair.
This doesn't put a precise cap on the amount of memory used (it used to
at some point). Set this parameter to lower values if you run out of memory
while running this function. Set it to higher values if this function is
too slow and you have lots of memory.</p></dd>


<dt>maxmiss</dt>
<dd><p>Discard SNPs which are missing in a fraction of populations
higher than <code>maxmiss</code></p></dd>


<dt>minmaf</dt>
<dd><p>Discard SNPs with minor allele frequency less than <code>minmaf</code></p></dd>


<dt>maxmaf</dt>
<dd><p>Discard SNPs with minor allele frequency greater than
than <code>maxmaf</code></p></dd>


<dt>minac2</dt>
<dd><p>Discard SNPs with allele count lower than 2 in any population
(default <code>FALSE</code>). This option should be set to <code>TRUE</code> when computing
f3-statistics where one population consists mostly of pseudohaploid samples.
Otherwise heterozygosity estimates and thus f3-estimates can be biased.
<code>minac2 == 2</code> will discard SNPs with allele count lower than 2 in any
non-singleton population (this option is experimental and is based on the
hypothesis that using SNPs with allele count lower than 2 only leads to
biases in non-singleton populations). Note that, While the <code>minac2</code> option
discards
SNPs with allele count lower than 2 in any population, the <code>qp3pop</code>
function will only discard SNPs with allele count lower than 2 in the
first (target) population (when the first argument is the prefix of
a genotype file; i.e. it is applied directly to a genotype file, not via
precomputing f2 from a <code><a href="gen_tibble.html">gen_tibble</a></code>).</p></dd>


<dt>outpop</dt>
<dd><p>Keep only SNPs which are heterozygous in this population</p></dd>


<dt>outpop_scale</dt>
<dd><p>Scale f2-statistics by the inverse <code>outpop</code>
heterozygosity (<code>1/(p*(1-p))</code>). Providing <code>outpop</code> and setting
<code>outpop_scale</code> to <code>TRUE</code> will give the same results as the original
<em>qpGraph</em> when the <code>outpop</code> parameter has been set, but it has the
disadvantage of treating one population different from the others. This
may limit the use of these f2-statistics for other models.</p></dd>


<dt>transitions</dt>
<dd><p>Set this to <code>FALSE</code> to exclude transition SNPs</p></dd>


<dt>transversions</dt>
<dd><p>Set this to <code>FALSE</code> to exclude transversion SNPs</p></dd>


<dt>overwrite</dt>
<dd><p>Overwrite existing files in <code>outdir</code></p></dd>


<dt>adjust_pseudohaploid</dt>
<dd><p>Genotypes of pseudohaploid samples are
usually coded as <code>0</code> or <code>2</code>, even though only one allele is observed.
<code>adjust_pseudohaploid</code> ensures that the observed allele count increases
only by <code>1</code> for each pseudohaploid sample. If <code>TRUE</code> (default), samples
that don't have any genotypes coded as <code>1</code> among the first 1000 SNPs are
automatically identified as pseudohaploid. This leads to slightly more
accurate estimates of f-statistics. Setting this parameter to <code>FALSE</code>
treats all samples as diploid and is equivalent to the <em>ADMIXTOOLS</em> <code> inbreed: NO</code> option. Setting <code>adjust_pseudohaploid</code> to an integer <code>n</code>
will check the first <code>n</code> SNPs instead of the first 1000 SNPs.</p></dd>


<dt>fst</dt>
<dd><p>Write files with pairwise FST for every population pair.
Setting this to FALSE can make <code>extract_f2</code> faster and will require less memory.</p></dd>


<dt>afprod</dt>
<dd><p>Write files with allele frequency products for every
population pair. Setting this to FALSE can make <code>extract_f2</code> faster and
will require less memory.</p></dd>


<dt>poly_only</dt>
<dd><p>Specify whether SNPs with identical allele frequencies in
every population should be discarded (<code>poly_only = TRUE</code>), or whether they
should be used (<code>poly_only = FALSE</code>). By default (<code>poly_only = c("f2")</code>),
these SNPs will be used to compute FST and allele frequency products, but
not to compute f2 (this is the default option in the original ADMIXTOOLS).</p></dd>


<dt>apply_corr</dt>
<dd><p>Apply small-sample-size correction when computing
f2-statistics (default <code>TRUE</code>)</p></dd>


<dt>n_cores</dt>
<dd><p>Parallelize computation across <code>n_cores</code> cores.</p></dd>


<dt>quiet</dt>
<dd><p>Suppress printing of progress updates</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>SNP metadata (invisibly)</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

