<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>PCA controlling for LD for gen_tibble objects — gt_pca_autoSVD • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PCA controlling for LD for gen_tibble objects — gt_pca_autoSVD"><meta name="description" content="This function performs Principal Component Analysis on a gen_tibble,
using a fast truncated SVD with initial pruning and then iterative removal
of long-range LD regions. This function is a wrapper for bigsnpr::snp_autoSVD()"><meta property="og:description" content="This function performs Principal Component Analysis on a gen_tibble,
using a fast truncated SVD with initial pruning and then iterative removal
of long-range LD regions. This function is a wrapper for bigsnpr::snp_autoSVD()"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9018</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">Overview</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Example workflow with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>PCA controlling for LD for <code>gen_tibble</code> objects</h1>

      <div class="d-none name"><code>gt_pca_autoSVD.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs Principal Component Analysis on a <code>gen_tibble</code>,
using a fast truncated SVD with initial pruning and then iterative removal
of long-range LD regions. This function is a wrapper for <code><a href="https://privefl.github.io/bigsnpr/reference/snp_autoSVD.html" class="external-link">bigsnpr::snp_autoSVD()</a></code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gt_pca_autoSVD</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  fun_scaling <span class="op">=</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_scaleBinom.html" class="external-link">snp_scaleBinom</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  thr_r2 <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  use_positions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">100</span><span class="op">/</span><span class="va">thr_r2</span>,</span>
<span>  roll_size <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  int_min_size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  alpha_tukey <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  min_mac <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a <code>gen_tbl</code> object</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Number of singular vectors/values to compute. Default is <code>10</code>.
<strong>This algorithm should be used to compute a few singular vectors/values.</strong></p></dd>


<dt id="arg-fun-scaling">fun_scaling<a class="anchor" aria-label="anchor" href="#arg-fun-scaling"></a></dt>
<dd><p>Usually this  can be left unset, as it defaults to
<code><a href="https://privefl.github.io/bigsnpr/reference/snp_scaleBinom.html" class="external-link">bigsnpr::snp_scaleBinom()</a></code>, which is the appropriate function for biallelic SNPs.
Alternatively it is possible to use  custom function
(see <code><a href="https://privefl.github.io/bigsnpr/reference/snp_autoSVD.html" class="external-link">bigsnpr::snp_autoSVD()</a></code> for details.</p></dd>


<dt id="arg-thr-r-">thr_r2<a class="anchor" aria-label="anchor" href="#arg-thr-r-"></a></dt>
<dd><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>. Use <code>NA</code> if you want to skip the clumping step.
size</p></dd>


<dt id="arg-use-positions">use_positions<a class="anchor" aria-label="anchor" href="#arg-use-positions"></a></dt>
<dd><p>a boolean on whether the position is used to define <code>size</code>,
or whether the size should be in number of SNPs. Default is TRUE</p></dd>


<dt id="arg-size">size<a class="anchor" aria-label="anchor" href="#arg-size"></a></dt>
<dd><p>For one SNP, window size around this SNP to compute correlations.
Default is 100 / thr_r2 for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt; 200).
If not providing infos.pos (NULL, the default), this is a window in number
of SNPs, otherwise it is a window in kb (genetic distance). I recommend
that you provide the positions if available.</p></dd>


<dt id="arg-roll-size">roll_size<a class="anchor" aria-label="anchor" href="#arg-roll-size"></a></dt>
<dd><p>Radius of rolling windows to smooth log-p-values.
Default is <code>50</code>.</p></dd>


<dt id="arg-int-min-size">int_min_size<a class="anchor" aria-label="anchor" href="#arg-int-min-size"></a></dt>
<dd><p>Minimum number of consecutive outlier SNPs
in order to be reported as long-range LD region. Default is <code>20</code>.</p></dd>


<dt id="arg-alpha-tukey">alpha_tukey<a class="anchor" aria-label="anchor" href="#arg-alpha-tukey"></a></dt>
<dd><p>Default is <code>0.1</code>. The type-I error rate in outlier
detection (that is further corrected for multiple testing).</p></dd>


<dt id="arg-min-mac">min_mac<a class="anchor" aria-label="anchor" href="#arg-min-mac"></a></dt>
<dd><p>Minimum minor allele count (MAC) for variants to be included.
Default is <code>10</code>.</p></dd>


<dt id="arg-max-iter">max_iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>Maximum number of iterations of outlier detection.
Default is <code>5</code>.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Number of cores used. Default doesn't use parallelism.
You may use <code><a href="https://privefl.github.io/bigstatsr/reference/reexports.html" class="external-link">bigstatsr::nb_cores()</a></code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Output some information on the iterations? Default is <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code>gt_pca</code> object, which is a subclass of <code>bigSVD</code>; this is
an S3 list with elements:
A named list (an S3 class "big_SVD") of</p><ul><li><p><code>d</code>, the eigenvalues (singular values, i.e. as variances),</p></li>
<li><p><code>u</code>, the scores for each sample on each component (the left singular vectors)</p></li>
<li><p><code>v</code>, the loadings (the right singular vectors)</p></li>
<li><p><code>center</code>, the centering vector,</p></li>
<li><p><code>scale</code>, the scaling vector,</p></li>
<li><p><code>method</code>, a string defining the method (in this case 'autoSVD'),</p></li>
<li><p><code>call</code>, the call that generated the object.</p></li>
</ul><p>Note: rather than accessing these elements directly, it is better to use
<code>tidy</code> and <code>augment</code>. See <code><a href="tidy_gt_pca.html">gt_pca_tidiers</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Using gt_pca_autoSVD requires a reasonably large dataset, as the function
iteratively removes regions of long range LD.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

