<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Detect runs of homozygosity using a sliding-window approach — gt_roh_window • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detect runs of homozygosity using a sliding-window approach — gt_roh_window"><meta name="description" content="This function uses a sliding-window approach to look for runs of homozygosity (or
heterozygosity) in a diploid genome. This function uses the package selectRUNS,
which implements an approach equivalent to the one in PLINK."><meta property="og:description" content="This function uses a sliding-window approach to look for runs of homozygosity (or
heterozygosity) in a diploid genome. This function uses the package selectRUNS,
which implements an approach equivalent to the one in PLINK."><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9017</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">Overview</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Example workflow with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Detect runs of homozygosity using a sliding-window approach</h1>

      <div class="d-none name"><code>gt_roh_window.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function uses a sliding-window approach to look for runs of homozygosity (or
heterozygosity) in a diploid genome. This function uses the package <code>selectRUNS</code>,
which implements an approach equivalent to the one in PLINK.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gt_roh_window</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  min_snp <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  heterozygosity <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max_opp_window <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_miss_window <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_gap <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span>,</span>
<span>  min_length_bps <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  min_density <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">1000</span>,</span>
<span>  max_opp_run <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_miss_run <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a <a href="gen_tibble.html">gen_tibble</a></p></dd>


<dt id="arg-window-size">window_size<a class="anchor" aria-label="anchor" href="#arg-window-size"></a></dt>
<dd><p>the size of sliding window (number of SNP loci) (default = 15)</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>the threshold of overlapping windows of the same state
(homozygous/heterozygous) to call a SNP in a RUN (default = 0.05)</p></dd>


<dt id="arg-min-snp">min_snp<a class="anchor" aria-label="anchor" href="#arg-min-snp"></a></dt>
<dd><p>minimum n. of SNP in a RUN (default = 3)</p></dd>


<dt id="arg-heterozygosity">heterozygosity<a class="anchor" aria-label="anchor" href="#arg-heterozygosity"></a></dt>
<dd><p>should we look for runs of heterozygosity (instead of homozygosity? (default = FALSE)</p></dd>


<dt id="arg-max-opp-window">max_opp_window<a class="anchor" aria-label="anchor" href="#arg-max-opp-window"></a></dt>
<dd><p>max n. of SNPs of the opposite type (e.g. heterozygous
snps for runs of homozygosity) in the
sliding window (default = 1)</p></dd>


<dt id="arg-max-miss-window">max_miss_window<a class="anchor" aria-label="anchor" href="#arg-max-miss-window"></a></dt>
<dd><p>max. n. of missing SNP in the sliding window (default = 1)</p></dd>


<dt id="arg-max-gap">max_gap<a class="anchor" aria-label="anchor" href="#arg-max-gap"></a></dt>
<dd><p>max distance between consecutive SNP to be still considered a
potential run (default = 10^6 bps)</p></dd>


<dt id="arg-min-length-bps">min_length_bps<a class="anchor" aria-label="anchor" href="#arg-min-length-bps"></a></dt>
<dd><p>minimum length of run in bps (defaults to 1000 bps = 1 kbps)</p></dd>


<dt id="arg-min-density">min_density<a class="anchor" aria-label="anchor" href="#arg-min-density"></a></dt>
<dd><p>minimum n. of SNP per kbps (defaults to 0.1 = 1 SNP every 10 kbps)</p></dd>


<dt id="arg-max-opp-run">max_opp_run<a class="anchor" aria-label="anchor" href="#arg-max-opp-run"></a></dt>
<dd><p>max n. of opposite genotype SNPs in the run (optional)</p></dd>


<dt id="arg-max-miss-run">max_miss_run<a class="anchor" aria-label="anchor" href="#arg-max-miss-run"></a></dt>
<dd><p>max n. of missing SNPs in the run (optional)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe with RUNs of Homozygosity or Heterozygosity in the analysed dataset.
The returned dataframe contains the following seven columns: "group", "id", "chrom",
"nSNP", "from", "to", "lengthBps" (group: population, breed, case/control etc.;
id: individual identifier; chrom: chromosome on which the run is located;
nSNP: number of SNPs in the run; from: starting position of the run, in bps;
to: end position of the run, in bps; lengthBps: size of the run)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function returns a data frame with all runs detected in the dataset.
This data frame can then be written out to a csv file.
The data frame is, in turn, the input for other functions of the detectRUNS package
that create plots and produce statistics from the results
(see plots and statistics functions in this manual,
and/or refer to the detectRUNS vignette).</p>
<p>If the <code><a href="gen_tibble.html">gen_tibble</a></code> is grouped, then the grouping variable is used to fill in the
group table. Otherwise, the group 'column' is filled with the same values as the 'id'
column</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># don't run the example</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="va">sheep_ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Kijas2016_Sheep_subset.ped"</span>,</span></span>
<span class="r-in"><span>    package<span class="op">=</span><span class="st">"detectRUNS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sheep_gt</span> <span class="op">&lt;-</span> <span class="fu">tidypopgen</span><span class="fu">::</span><span class="fu"><a href="gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">sheep_ped</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sheep_gt</span> <span class="op">&lt;-</span> <span class="va">sheep_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">group_by</span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sheep_roh</span> <span class="op">&lt;-</span> <span class="fu">gt_roh_window</span><span class="op">(</span><span class="va">sheep_gt</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">detectRUNS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/detectRUNS/man/plot_Runs.html" class="external-link">plot_Runs</a></span><span class="op">(</span>runs <span class="op">=</span> <span class="va">sheep_roh</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

