<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run SNMF from R in tidypopgen — gt_snmf • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run SNMF from R in tidypopgen — gt_snmf"><meta name="description" content="Run SNMF from R in tidypopgen"><meta property="og:description" content="Run SNMF from R in tidypopgen"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run SNMF from R in tidypopgen</h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/R/gt_snmf.R" class="external-link"><code>R/gt_snmf.R</code></a></small>
      <div class="d-none name"><code>gt_snmf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run SNMF from R in tidypopgen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gt_snmf</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">k</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"continue"</span>,</span>
<span>  n_runs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">alpha</span>,</span>
<span>  tolerance <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  percentage <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">I</span>,</span>
<span>  iterations <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  seed <span class="op">=</span> <span class="op">-</span><span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a <code>gen_tibble</code> or a character giving the path to the input geno file</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>an integer giving the number of clusters</p></dd>


<dt id="arg-project">project<a class="anchor" aria-label="anchor" href="#arg-project"></a></dt>
<dd><p>one of "continue", "new", and "force": "continue" stores files
in the current project, "new" creates a new project, and "force" stores
results in the current project even if the .geno input file has been
altered,</p></dd>


<dt id="arg-n-runs">n_runs<a class="anchor" aria-label="anchor" href="#arg-n-runs"></a></dt>
<dd><p>the number of runs for each k value (defaults to 1)</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>numeric snmf regularization parameter. See LEA::snmf for details</p></dd>


<dt id="arg-tolerance">tolerance<a class="anchor" aria-label="anchor" href="#arg-tolerance"></a></dt>
<dd><p>numeric value of tolerance (default 0.00001)</p></dd>


<dt id="arg-entropy">entropy<a class="anchor" aria-label="anchor" href="#arg-entropy"></a></dt>
<dd><p>boolean indicating whether to estimate cross-entropy</p></dd>


<dt id="arg-percentage">percentage<a class="anchor" aria-label="anchor" href="#arg-percentage"></a></dt>
<dd><p>numeric value indicating percentage of masked genotypes,
ranging between 0 and 1, to be used when entropy = TRUE</p></dd>


<dt id="arg-i">I<a class="anchor" aria-label="anchor" href="#arg-i"></a></dt>
<dd><p>number of SNPs for initialising the snmf algorithm</p></dd>


<dt id="arg-iterations">iterations<a class="anchor" aria-label="anchor" href="#arg-iterations"></a></dt>
<dd><p>numeric integer for maximum iterations (default 200)</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>the ploidy of the input data (defaults to 2)</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>the seed for the random number generator</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of class <code>gt_admix</code> consisting of a list with the following
elements:</p><ul><li><p><code>k</code> the number of clusters</p></li>
<li><p><code>Q</code> a matrix with the admixture proportions</p></li>
<li><p><code>P</code> a matrix with the allele frequencies</p></li>
<li><p><code>log</code> a log of the output generated by ADMIXTURE (usually printed
on the screen when running from the command line)</p></li>
<li><p><code>cv</code> the masked cross-entropy (if <code>entropy</code> is TRUE)</p></li>
<li><p><code>loglik</code> the log likelihood of the model</p></li>
<li><p><code>id</code> the id column of the input <code>gen_tibble</code> (if applicable)</p></li>
<li><p><code>group</code> the group column of the input <code>gen_tibble</code> (if applicable)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This is a wrapper for the function snmf from R package LEA.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="example_gt.html">example_gt</a></span><span class="op">(</span><span class="st">"gen_tbl"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To run SNMF on a gen_tibble:</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">gt_snmf</span><span class="op">(</span></span></span>
<span class="r-in"><span>  k <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, project <span class="op">=</span> <span class="st">"force"</span>, entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  percentage <span class="op">=</span> <span class="fl">0.5</span>, n_runs <span class="op">=</span> <span class="fl">1</span>, seed <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">100</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $k</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1 2 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Q</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Q[[1]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      .Q1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [7,]   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Q[[2]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              .Q1         .Q2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.000274926 0.999725000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.000100000 0.999900000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 0.000100000 0.999900000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 0.999900000 0.000100000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 0.999742000 0.000257853</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,] 0.999719000 0.000281284</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [7,] 0.000317214 0.999683000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Q[[3]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              .Q1         .Q2         .Q3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.000240022 0.999520000 0.000240022</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.000222869 0.999554000 0.000222869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 0.000222869 0.999554000 0.000222869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 0.999538000 0.000231010 0.000231010</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 0.999800000 0.000099990 0.000099990</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,] 0.000275505 0.000275505 0.999449000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [7,] 0.000275525 0.000275525 0.999449000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $P</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> list()</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $log</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1] "The project is saved into :"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2] "  "                                                                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4] "To load the project, use:"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5] " project = load.snmfProject(\"\")"                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7] "To remove the project, use:"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [8] " remove.snmfProject(\"\")"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [9] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [10] "[1] 1"                                                                                                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [11] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [12] "[1] \"*          create.dataset            *\""                                                                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [13] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [14] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [15] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [16] "        -n (number of individuals)                 7"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [17] "        -L (number of loci)                        6"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [18] "        -s (seed random init)                      1"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [19] "        -r (percentage of masked data)             0.5"                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [20] "        -x (genotype file in .geno format)         /tmp/RtmpyN1xne/file22a2345f1541.geno"                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [21] "        -o (output file in .geno format)           /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [22] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [23] " Write genotype file with masked data, /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno:\t\tOK."          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [24] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [25] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [26] "[1] \"* sNMF K = 1  repetition 1      *\""                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [27] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [28] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [29] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [30] "        -n (number of individuals)             7"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [31] "        -L (number of loci)                    6"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [32] "        -K (number of ancestral pops)          1"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [33] "        -x (input file)                        /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [34] "        -q (individual admixture file)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.Q"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [35] "        -g (ancestral frequencies file)        /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.G"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [36] "        -i (number max of iterations)          200"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [37] "        -a (regularization parameter)          100"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [38] "        -s (seed random init)                  94308891885569"                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [39] "        -e (tolerance error)                   1E-05"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [40] "        -p (number of processes)               1"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [41] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [42] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [43] "Read genotype file /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno:\t\tOK."                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [44] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [45] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [46] "Main algorithm:"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [47] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [48] "Least-square error: 13.714286"                                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [49] "Write individual ancestry coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.Q:\t\tOK."     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [50] "Write ancestral allele frequency coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.G:\tOK."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [51] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [52] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [53] "[1] \"*    cross-entropy estimation       *\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [54] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [55] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [56] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [57] "        -n (number of individuals)         7"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [58] "        -L (number of loci)                6"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [59] "        -K (number of ancestral pops)      1"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [60] "        -x (genotype file)                 /tmp/RtmpyN1xne/file22a2345f1541.geno"                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [61] "        -q (individual admixture)          /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.Q"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [62] "        -g (ancestral frequencies)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K1/run1/file22a2345f1541_r1.1.G"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [63] "        -i (with masked genotypes)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [64] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [65] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [66] "Cross-Entropy (all data):\t 0.706731"                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [67] "Cross-Entropy (masked data):\t 4.42151"                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [68] "The project is saved into :"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [69] "  "                                                                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [70] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [71] "To load the project, use:"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [72] " project = load.snmfProject(\"\")"                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [73] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [74] "To remove the project, use:"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [75] " remove.snmfProject(\"\")"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [76] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [77] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [78] "[1] \"* sNMF K = 2  repetition 1      *\""                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [79] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [80] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [81] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [82] "        -n (number of individuals)             7"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [83] "        -L (number of loci)                    6"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [84] "        -K (number of ancestral pops)          2"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [85] "        -x (input file)                        /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [86] "        -q (individual admixture file)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.Q"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [87] "        -g (ancestral frequencies file)        /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.G"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [88] "        -i (number max of iterations)          200"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [89] "        -a (regularization parameter)          100"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [90] "        -s (seed random init)                  8589934593"                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [91] "        -e (tolerance error)                   1E-05"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [92] "        -p (number of processes)               1"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [93] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [94] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [95] "Read genotype file /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno:\t\tOK."                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [96] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [97] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [98] "Main algorithm:"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [99] "\t[                                                                           ]"                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [100] "\t[======]"                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [101] "Number of iterations: 16"                                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [102] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [103] "Least-square error: 14.000528"                                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [104] "Write individual ancestry coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.Q:\t\tOK."     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [105] "Write ancestral allele frequency coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.G:\tOK."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [106] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [107] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [108] "[1] \"*    cross-entropy estimation       *\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [109] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [110] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [111] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [112] "        -n (number of individuals)         7"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [113] "        -L (number of loci)                6"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [114] "        -K (number of ancestral pops)      2"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [115] "        -x (genotype file)                 /tmp/RtmpyN1xne/file22a2345f1541.geno"                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [116] "        -q (individual admixture)          /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.Q"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [117] "        -g (ancestral frequencies)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K2/run1/file22a2345f1541_r1.2.G"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [118] "        -i (with masked genotypes)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [119] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [120] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [121] "Cross-Entropy (all data):\t 0.69943"                                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [122] "Cross-Entropy (masked data):\t 4.79389"                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [123] "The project is saved into :"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [124] "  "                                                                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [125] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [126] "To load the project, use:"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [127] " project = load.snmfProject(\"\")"                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [128] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [129] "To remove the project, use:"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [130] " remove.snmfProject(\"\")"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [131] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [132] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [133] "[1] \"* sNMF K = 3  repetition 1      *\""                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [134] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [135] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [136] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [137] "        -n (number of individuals)             7"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [138] "        -L (number of loci)                    6"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [139] "        -K (number of ancestral pops)          3"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [140] "        -x (input file)                        /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [141] "        -q (individual admixture file)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.Q"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [142] "        -g (ancestral frequencies file)        /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.G"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [143] "        -i (number max of iterations)          200"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [144] "        -a (regularization parameter)          100"                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [145] "        -s (seed random init)                  94308891885569"                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [146] "        -e (tolerance error)                   1E-05"                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [147] "        -p (number of processes)               1"                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [148] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [149] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [150] "Read genotype file /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno:\t\tOK."                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [151] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [152] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [153] "Main algorithm:"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [154] "\t[                                                                           ]"                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [155] "\t[======]"                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [156] "Number of iterations: 16"                                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [157] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [158] "Least-square error: 8.583557"                                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [159] "Write individual ancestry coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.Q:\t\tOK."     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [160] "Write ancestral allele frequency coefficient file /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.G:\tOK."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [161] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [162] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [163] "[1] \"*    cross-entropy estimation       *\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [164] "[1] \"*************************************\""                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [165] "summary of the options:"                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [166] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [167] "        -n (number of individuals)         7"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [168] "        -L (number of loci)                6"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [169] "        -K (number of ancestral pops)      3"                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [170] "        -x (genotype file)                 /tmp/RtmpyN1xne/file22a2345f1541.geno"                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [171] "        -q (individual admixture)          /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.Q"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [172] "        -g (ancestral frequencies)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/K3/run1/file22a2345f1541_r1.3.G"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [173] "        -i (with masked genotypes)         /tmp/RtmpyN1xne/file22a2345f1541.snmf/masked/file22a2345f1541_I.geno"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [174] "        - diploid"                                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [175] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [176] "Cross-Entropy (all data):\t 0.349945"                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [177] "Cross-Entropy (masked data):\t 5.78159"                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [178] "The project is saved into :"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [179] "  "                                                                                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [180] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [181] "To load the project, use:"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [182] " project = load.snmfProject(\"\")"                                                                                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [183] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [184] "To remove the project, use:"                                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [185] " remove.snmfProject(\"\")"                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [186] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [187] "snmf Project"                                                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [188] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [189] "snmfProject file:                 file22a2345f1541.snmfProject "                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [190] "project directory:                /tmp/RtmpyN1xne/ "                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [191] "snmf results directory:           file22a2345f1541.snmf/ "                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [192] "date of creation:                 1749627446 "                                                                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [193] "input file:                       file22a2345f1541.geno "                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [194] "number of individuals:            7 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [195] "number of loci:                   6 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [196] "number of ancestral populations:  1 2 3 "                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [197] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [198] "***** run *****"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [199] "snmf class"                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [200] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [201] "file directory:                   K1/run1/ "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [202] "Q output file:                    file22a2345f1541_r1.1.Q "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [203] "G output file:                    file22a2345f1541_r1.1.G "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [204] "snmfClass file:                   file22a2345f1541_r1.1.snmfClass "                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [205] "number of ancestral populations:  1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [206] "run number:                       1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [207] "regularization parameter:         100 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [208] "number of CPUs:                   1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [209] "seed:                             1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [210] "maximal number of iterations:     200 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [211] "tolerance error:                  1e-05 "                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [212] "Q input file:                      "                                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [213] "cross-Entropy:                    4.421509 "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [214] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [215] "***** run *****"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [216] "snmf class"                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [217] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [218] "file directory:                   K2/run1/ "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [219] "Q output file:                    file22a2345f1541_r1.2.Q "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [220] "G output file:                    file22a2345f1541_r1.2.G "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [221] "snmfClass file:                   file22a2345f1541_r1.2.snmfClass "                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [222] "number of ancestral populations:  2 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [223] "run number:                       1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [224] "regularization parameter:         100 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [225] "number of CPUs:                   1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [226] "seed:                             1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [227] "maximal number of iterations:     200 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [228] "tolerance error:                  1e-05 "                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [229] "Q input file:                      "                                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [230] "cross-Entropy:                    4.793895 "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [231] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [232] "***** run *****"                                                                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [233] "snmf class"                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [234] ""                                                                                                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [235] "file directory:                   K3/run1/ "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [236] "Q output file:                    file22a2345f1541_r1.3.Q "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [237] "G output file:                    file22a2345f1541_r1.3.G "                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [238] "snmfClass file:                   file22a2345f1541_r1.3.snmfClass "                                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [239] "number of ancestral populations:  3 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [240] "run number:                       1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [241] "regularization parameter:         100 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [242] "number of CPUs:                   1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [243] "seed:                             1 "                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [244] "maximal number of iterations:     200 "                                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [245] "tolerance error:                  1e-05 "                                                                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [246] "Q input file:                      "                                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [247] "cross-Entropy:                    5.781592 "                                                                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $loglik</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric(0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G[[1]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            .Q1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] 0.285714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] 0.428571</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] 0.285714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] 0.285714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] 0.714286</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] 0.857143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] 0.142857</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] 0.999900</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13,] 0.285714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14,] 0.714286</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16,] 0.857143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17,] 0.142857</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18,] 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G[[2]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            .Q1      .Q2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] 0.666578 0.749950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] 0.333422 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] 0.000100 0.250050</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] 0.333359 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] 0.666641 0.999900</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] 0.500021 0.249966</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] 0.000100 0.250034</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] 0.499979 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13,] 0.999900 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14,] 0.000100 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16,] 0.333262 0.750059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17,] 0.666738 0.249941</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18,] 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G[[3]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            .Q1      .Q2      .Q3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] 0.000100 0.000100 0.999900</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] 0.999900 0.333173 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] 0.000100 0.666827 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] 0.500116 0.000100 0.999900</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] 0.499884 0.999900 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] 0.500014 0.500029 0.249862</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] 0.000100 0.000100 0.250138</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] 0.499986 0.499971 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13,] 0.999900 0.000100 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14,] 0.000100 0.999900 0.500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16,] 0.999900 0.999900 0.499724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17,] 0.000100 0.000100 0.500276</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18,] 0.000100 0.000100 0.000100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "q_matrix" "matrix"   "array"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $cv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 4.42151 4.79389 5.78159</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "a" "b" "c" "d" "e" "f" "g"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $algorithm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "SNMF"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "gt_admix" "list"    </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

