<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9014</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/a01_overview.html">Overview</a>
    </li>
    <li>
      <a href="../articles/a02_qc.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/a03_example_clustering_and_dapc.html">Example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/a99_plink_cheatsheet.html">plink cheatsheet</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="annotate_group_info.html">annotate_group_info()</a></code> </p>
        </td>
        <td><p>Annotate an adxmiture barplot with information about the populations</p></td>
      </tr><tr><td>
          <p><code><a href="as_q_matrix.html">as_q_matrix()</a></code> </p>
        </td>
        <td><p>Convert a Q matrix into a <code>q_matrix</code> obejct</p></td>
      </tr><tr><td>
          <p><code><a href="augment.gt_dapc.html">augment(<i>&lt;gt_dapc&gt;</i>)</a></code> </p>
        </td>
        <td><p>Augment data with information from a gt_dapc object</p></td>
      </tr><tr><td>
          <p><code><a href="augment_gt_pca.html">augment(<i>&lt;gt_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Augment data with information from a gt_pca object</p></td>
      </tr><tr><td>
          <p><code><a href="augment_loci.html">augment_loci()</a></code> </p>
        </td>
        <td><p>Augment the loci table with information from a analysis object</p></td>
      </tr><tr><td>
          <p><code><a href="augment_loci_gt_pca.html">augment_loci(<i>&lt;gt_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Augment the loci table with information from a gt_pca object</p></td>
      </tr><tr><td>
          <p><code><a href="autoplot.qc_report_indiv.html">autoplot(<i>&lt;qc_report_indiv&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autoplots for <code>qc_report_indiv</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="autoplot.qc_report_loci.html">autoplot(<i>&lt;qc_report_loci&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autoplots for <code>qc_report_loci</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="autoplot_gt_cluster_pca.html">autoplot(<i>&lt;gt_cluster_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autoplots for <code>gt_cluster_pca</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="autoplot_gt_dapc.html">autoplot(<i>&lt;gt_dapc&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autoplots for <code>gt_dapc</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="autoplot_gt_pca.html">autoplot(<i>&lt;gt_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autoplots for <code>gt_pca</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="count_loci.html">count_loci()</a></code> </p>
        </td>
        <td><p>Count the number of loci in a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="distruct_colours.html">distruct_colours</a></code> </p>
        </td>
        <td><p>Distruct colours</p></td>
      </tr><tr><td>
          <p><code><a href="filter_high_relatedness.html">filter_high_relatedness()</a></code> </p>
        </td>
        <td><p>Filter individuals based on a relationship threshold</p></td>
      </tr><tr><td>
          <p><code><a href="gen_tibble.html">gen_tibble()</a></code> </p>
        </td>
        <td><p>Constructor for a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="gt_as_genind.html">gt_as_genind()</a></code> </p>
        </td>
        <td><p>Convert a <code>gen_tibble</code> to a <code>genind</code> object from <code>adegenet</code></p></td>
      </tr><tr><td>
          <p><code><a href="gt_as_genlight.html">gt_as_genlight()</a></code> </p>
        </td>
        <td><p>Convert a <code>gen_tibble</code> to a <code>genlight</code> object from <code>adegenet</code></p></td>
      </tr><tr><td>
          <p><code><a href="gt_as_geno_lea.html">gt_as_geno_lea()</a></code> </p>
        </td>
        <td><p>Convert a <code>gentibble</code> to a .geno file for sNMF from the LEA package</p></td>
      </tr><tr><td>
          <p><code><a href="gt_as_hierfstat.html">gt_as_hierfstat()</a></code> </p>
        </td>
        <td><p>Convert a <code>gen_tibble</code> to a data.frame compatible with <code>hierfstat</code></p></td>
      </tr><tr><td>
          <p><code><a href="gt_as_plink.html">gt_as_plink()</a></code> </p>
        </td>
        <td><p>Export a <code>gen_tibble</code> object to PLINK bed format</p></td>
      </tr><tr><td>
          <p><code><a href="gt_cluster_pca.html">gt_cluster_pca()</a></code> </p>
        </td>
        <td><p>Run K-clustering on principal components</p></td>
      </tr><tr><td>
          <p><code><a href="gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code> </p>
        </td>
        <td><p>Find the best number of clusters based on principal components</p></td>
      </tr><tr><td>
          <p><code><a href="gt_dapc.html">gt_dapc()</a></code> </p>
        </td>
        <td><p>Discriminant Analysis of Principal Components for gen_tibble</p></td>
      </tr><tr><td>
          <p><code><a href="gt_extract_f2.html">gt_extract_f2()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics for ADMIXTOOLS 2</p></td>
      </tr><tr><td>
          <p><code><a href="gt_get_file_names.html">gt_get_file_names()</a></code> </p>
        </td>
        <td><p>Get the names of files storing the genotypes of a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="gt_has_imputed.html">gt_has_imputed()</a></code> </p>
        </td>
        <td><p>Checks if a <code>gen_tibble</code> has been imputed</p></td>
      </tr><tr><td>
          <p><code><a href="gt_impute_simple.html">gt_impute_simple()</a></code> </p>
        </td>
        <td><p>Simple imputation based on allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="gt_king.html">gt_king()</a></code> </p>
        </td>
        <td><p>Compute the KING-robust Matrix for a a <code>gen_tibble</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="gt_load.html">gt_load()</a></code> </p>
        </td>
        <td><p>Load a gen_tibble</p></td>
      </tr><tr><td>
          <p><code><a href="gt_pca.html">gt_pca</a></code> </p>
        </td>
        <td><p>Principal Component Analysis for <code>gen_tibble</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="gt_pca_autoSVD.html">gt_pca_autoSVD()</a></code> </p>
        </td>
        <td><p>PCA controlling for LD for <code>gen_tibble</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="gt_pca_partialSVD.html">gt_pca_partialSVD()</a></code> </p>
        </td>
        <td><p>PCA for <code>gen_tibble</code> objects by partial SVD</p></td>
      </tr><tr><td>
          <p><code><a href="gt_pca_randomSVD.html">gt_pca_randomSVD()</a></code> </p>
        </td>
        <td><p>PCA for <code>gen_tibble</code> objects by randomized partial SVD</p></td>
      </tr><tr><td>
          <p><code><a href="gt_roh_window.html">gt_roh_window()</a></code> </p>
        </td>
        <td><p>Detect runs of homozygosity using a sliding-window approach</p></td>
      </tr><tr><td>
          <p><code><a href="gt_save.html">gt_save()</a></code> </p>
        </td>
        <td><p>Save a gen_tibble</p></td>
      </tr><tr><td>
          <p><code><a href="gt_set_imputed.html">gt_set_imputed()</a></code> </p>
        </td>
        <td><p>Sets a <code>gen_tibble</code> to use imputed data</p></td>
      </tr><tr><td>
          <p><code><a href="gt_uses_imputed.html">gt_uses_imputed()</a></code> </p>
        </td>
        <td><p>Checks if a <code>gen_tibble</code> uses imputed data</p></td>
      </tr><tr><td>
          <p><code><a href="indiv_het_obs.html">indiv_het_obs()</a></code> </p>
        </td>
        <td><p>Estimate individual observed heterozygosity</p></td>
      </tr><tr><td>
          <p><code><a href="indiv_missingness.html">indiv_missingness()</a></code> </p>
        </td>
        <td><p>Estimate individual missingness</p></td>
      </tr><tr><td>
          <p><code><a href="indiv_ploidy.html">indiv_ploidy()</a></code> </p>
        </td>
        <td><p>Return individual ploidy</p></td>
      </tr><tr><td>
          <p><code><a href="loci_alt_freq.html">loci_alt_freq()</a></code> <code><a href="loci_alt_freq.html">loci_maf()</a></code> </p>
        </td>
        <td><p>Estimate allele frequencies at each each locus</p></td>
      </tr><tr><td>
          <p><code><a href="loci_chromosomes.html">loci_chromosomes()</a></code> </p>
        </td>
        <td><p>Get the chromosomes of loci in a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="loci_hwe.html">loci_hwe()</a></code> </p>
        </td>
        <td><p>Test Hardy-Weinberg equilibrium at each locus</p></td>
      </tr><tr><td>
          <p><code><a href="loci_ld_clump.html">loci_ld_clump()</a></code> </p>
        </td>
        <td><p>Clump loci based on a Linkage Disequilibrium threshold</p></td>
      </tr><tr><td>
          <p><code><a href="loci_missingness.html">loci_missingness()</a></code> </p>
        </td>
        <td><p>Estimate missingness at each locus</p></td>
      </tr><tr><td>
          <p><code><a href="loci_names.html">loci_names()</a></code> </p>
        </td>
        <td><p>Get the names of loci in a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="loci_transitions.html">loci_transitions()</a></code> </p>
        </td>
        <td><p>Find transitions</p></td>
      </tr><tr><td>
          <p><code><a href="loci_transversions.html">loci_transversions()</a></code> </p>
        </td>
        <td><p>Find transversions</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_allele_sharing.html">pairwise_allele_sharing()</a></code> </p>
        </td>
        <td><p>Compute the Pairwise Allele Sharing Matrix for a <code>gen_tibble</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_ibs.html">pairwise_ibs()</a></code> </p>
        </td>
        <td><p>Compute the Identity by State Matrix for a <code>gen_tibble</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_pop_fst.html">pairwise_pop_fst()</a></code> </p>
        </td>
        <td><p>Compute pairwise population Fst</p></td>
      </tr><tr><td>
          <p><code><a href="pop_fis.html">pop_fis()</a></code> </p>
        </td>
        <td><p>Compute population specific FIS</p></td>
      </tr><tr><td>
          <p><code><a href="pop_fst.html">pop_fst()</a></code> </p>
        </td>
        <td><p>Compute population specific Fst</p></td>
      </tr><tr><td>
          <p><code><a href="predict_gt_pca.html">predict(<i>&lt;gt_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Predict scores of a PCA</p></td>
      </tr><tr><td>
          <p><code><a href="qc_report_indiv.html">qc_report_indiv()</a></code> </p>
        </td>
        <td><p>Create a Quality Control report for individuals</p></td>
      </tr><tr><td>
          <p><code><a href="qc_report_loci.html">qc_report_loci()</a></code> </p>
        </td>
        <td><p>Create a Quality Control report for loci</p></td>
      </tr><tr><td>
          <p><code><a href="rbind.gen_tbl.html">rbind(<i>&lt;gen_tbl&gt;</i>)</a></code> </p>
        </td>
        <td><p>Combine two gen_tibbles</p></td>
      </tr><tr><td>
          <p><code><a href="rbind_dry_run.html">rbind_dry_run()</a></code> </p>
        </td>
        <td><p>Generate a report of what would happen to each SNP in a merge</p></td>
      </tr><tr><td>
          <p><code><a href="scale_fill_distruct.html">scale_fill_distruct()</a></code> </p>
        </td>
        <td><p>Scale constructor using the distruct colours</p></td>
      </tr><tr><td>
          <p><code><a href="select_loci.html">select_loci()</a></code> </p>
        </td>
        <td><p>The <code>select</code> verb for <code>loci</code></p></td>
      </tr><tr><td>
          <p><code><a href="select_loci_if.html">select_loci_if()</a></code> </p>
        </td>
        <td><p>The <code>select_if</code> verb for <code>loci</code></p></td>
      </tr><tr><td>
          <p><code><a href="show_genotypes.html">show_genotypes()</a></code> </p>
        </td>
        <td><p>Show the genotypes of a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="show_loci.html">show_loci()</a></code> <code><a href="show_loci.html">`show_loci&lt;-`()</a></code> </p>
        </td>
        <td><p>Show the loci information of a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="show_ploidy.html">show_ploidy()</a></code> </p>
        </td>
        <td><p>Show the ploidy information of a <code>gen_tibble</code></p></td>
      </tr><tr><td>
          <p><code><a href="snp_allele_sharing.html">snp_allele_sharing()</a></code> </p>
        </td>
        <td><p>Compute the Pairwise Allele Sharing Matrix for a bigSNP object</p></td>
      </tr><tr><td>
          <p><code><a href="snp_ibs.html">snp_ibs()</a></code> </p>
        </td>
        <td><p>Compute the Identity by State Matrix for a bigSNP object</p></td>
      </tr><tr><td>
          <p><code><a href="snp_king.html">snp_king()</a></code> </p>
        </td>
        <td><p>Compute the KING-robust Matrix for a bigSNP object</p></td>
      </tr><tr><td>
          <p><code><a href="summary_rbind_dry_run.html">summary(<i>&lt;rbind_report&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print a summary of a merge report</p></td>
      </tr><tr><td>
          <p><code><a href="theme_distruct.html">theme_distruct()</a></code> </p>
        </td>
        <td><p>A theme to match the output of distruct</p></td>
      </tr><tr><td>
          <p><code><a href="tidy.gt_dapc.html">tidy(<i>&lt;gt_dapc&gt;</i>)</a></code> </p>
        </td>
        <td><p>Tidy a <code>gt_dapc</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="tidy.q_matrix.html">tidy(<i>&lt;q_matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Tidy a Q matrix</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_gt_pca.html">tidy(<i>&lt;gt_pca&gt;</i>)</a></code> </p>
        </td>
        <td><p>Tidy a <code>gt_pca</code> object</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

