<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute pairwise population Fst — pairwise_pop_fst • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute pairwise population Fst — pairwise_pop_fst"><meta name="description" content="This function computes pairwise Fst. The following methods are implemented:
'Hudson': Hudson's formulation, as derived in Bhatia et al (2013)
for diploids. This is the only method that can also be used with
pseudohaploid data.
'Nei87' : Fst according to Nei (1987) - includes the correction for
heterozygosity when computing Ht (it uses the same formulation as in
hierfstat::pairwise.neifst()),
'WC84' : Weir and Cockerham (1984), correcting for missing data (it uses
the same formulation as in hierfstat::pairwise.WCfst()).

"><meta property="og:description" content="This function computes pairwise Fst. The following methods are implemented:
'Hudson': Hudson's formulation, as derived in Bhatia et al (2013)
for diploids. This is the only method that can also be used with
pseudohaploid data.
'Nei87' : Fst according to Nei (1987) - includes the correction for
heterozygosity when computing Ht (it uses the same formulation as in
hierfstat::pairwise.neifst()),
'WC84' : Weir and Cockerham (1984), correcting for missing data (it uses
the same formulation as in hierfstat::pairwise.WCfst()).

"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
    <li><a class="dropdown-item" href="../articles/interpolating.html">Interpolating genetic data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute pairwise population Fst</h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/R/pairwise_pop_fst.R" class="external-link"><code>R/pairwise_pop_fst.R</code></a></small>
      <div class="d-none name"><code>pairwise_pop_fst.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes pairwise Fst. The following methods are implemented:</p><ul><li><p>'Hudson': Hudson's formulation, as derived in Bhatia et al (2013)
for diploids. This is the only method that can also be used with
pseudohaploid data.</p></li>
<li><p>'Nei87' : Fst according to Nei (1987) - includes the correction for
heterozygosity when computing Ht (it uses the same formulation as in
<code><a href="https://rdrr.io/pkg/hierfstat/man/pairwise.neifst.html" class="external-link">hierfstat::pairwise.neifst()</a></code>),</p></li>
<li><p>'WC84' : Weir and Cockerham (1984), correcting for missing data (it uses
the same formulation as in <code><a href="https://rdrr.io/pkg/hierfstat/man/pairwise.WCfst.html" class="external-link">hierfstat::pairwise.WCfst()</a></code>).</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pairwise_pop_fst</span><span class="op">(</span></span>
<span>  <span class="va">.x</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tidy"</span>, <span class="st">"pairwise"</span><span class="op">)</span>,</span>
<span>  by_locus <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  by_locus_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tidy"</span>, <span class="st">"matrix"</span>, <span class="st">"list"</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Hudson"</span>, <span class="st">"Nei87"</span>, <span class="st">"WC84"</span><span class="op">)</span>,</span>
<span>  return_num_dem <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fu">bigstatsr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg--x">.x<a class="anchor" aria-label="anchor" href="#arg--x"></a></dt>
<dd><p>a grouped <code><a href="gen_tibble.html">gen_tibble</a></code> (as obtained by using <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">dplyr::group_by()</a></code>)</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>type of object to return One of "tidy" or "pairwise" for a
pairwise matrix of populations. Default is "tidy".</p></dd>


<dt id="arg-by-locus">by_locus<a class="anchor" aria-label="anchor" href="#arg-by-locus"></a></dt>
<dd><p>boolean, determining whether Fst should be returned by
locus(TRUE), or as a single genome wide value obtained by taking the ratio
of the mean numerator and denominator (FALSE, the default).</p></dd>


<dt id="arg-by-locus-type">by_locus_type<a class="anchor" aria-label="anchor" href="#arg-by-locus-type"></a></dt>
<dd><p>type of object to return. One of "tidy", "matrix" or
"list". Default is "tidy".</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>one of 'Hudson', 'Nei87', and 'WC84'</p></dd>


<dt id="arg-return-num-dem">return_num_dem<a class="anchor" aria-label="anchor" href="#arg-return-num-dem"></a></dt>
<dd><p>returns a list of numerators and denominators for each
locus. This is useful for creating windowed estimates of Fst (as we need to
compute the mean numerator and denominator within each window). Default is
FALSE.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>number of cores to be used, it defaults to
<code><a href="https://privefl.github.io/bigstatsr/reference/reexports.html" class="external-link">bigstatsr::nb_cores()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>if <code>type=tidy</code>, a tibble of genome-wide pairwise Fst values with
each pairwise combination as a row if "by_locus=FALSE", else a list
including the tibble of genome-wide values as well as a matrix with
pairwise Fst by locus with loci as rows and and pairwise combinations as
columns. If <code>type=pairwise</code>, a matrix of genome-wide pairwise Fst values is
returned.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For all formulae, the genome wide estimate is obtained by taking the ratio of
the mean numerators and denominators over all relevant SNPs.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Bhatia G, Patterson N, Sankararaman S, Price AL. (2013)
Estimating and Interpreting FST: The Impact of Rare Variants. Genome
Research, 23(9):1514–1521.</p>
<p>Nei, M. (1987) Molecular Evolutionary Genetics. Columbia University Press</p>
<p>Weir, B. S., &amp; Cockerham, C. C. (1984). Estimating F-statistics for the
analysis of population structure. Evolution, 38(6): 1358–1370.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/hierfstat/man/pairwise.neifst.html" class="external-link">hierfstat::pairwise.neifst()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="load_example_gt.html">load_example_gt</a></span><span class="op">(</span><span class="st">"gen_tbl"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For a basic global pairwise Fst calculation:</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">pairwise_pop_fst</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"Nei87"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   population_1 population_2  value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> pop1         pop2         0.032<span style="text-decoration: underline;">0</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> pop1         pop3         0.143 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> pop2         pop3         0.050<span style="text-decoration: underline;">0</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With a pairwise matrix:</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">pairwise_pop_fst</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"Nei87"</span>, type <span class="op">=</span> <span class="st">"pairwise"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           pop1      pop2      pop3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pop1        NA 0.0320197 0.1428571</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pop2 0.0320197        NA 0.0500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pop3 0.1428571 0.0500000        NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To calculate Fst by locus:</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">pairwise_pop_fst</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"Hudson"</span>, by_locus <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Fst_by_locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 18 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    loci  stat_name       value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> rs1   fst_pop1.pop2  -<span style="color: #BB0000;">0.24</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> rs1   fst_pop1.pop3   0.6  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> rs1   fst_pop2.pop3   0.667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> rs2   fst_pop1.pop2   0.6  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> rs2   fst_pop1.pop3   0.114</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> rs2   fst_pop2.pop3   0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> rs3   fst_pop1.pop2 <span style="color: #BB0000;">NaN</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> rs3   fst_pop1.pop3   0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> rs3   fst_pop2.pop3   0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> rs4   fst_pop1.pop2  -<span style="color: #BB0000;">0.167</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> rs4   fst_pop1.pop3   0.333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> rs4   fst_pop2.pop3   0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> rs5   fst_pop1.pop2  -<span style="color: #BB0000;">0.1</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> rs5   fst_pop1.pop3  -<span style="color: #BB0000;">0.6</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> rs5   fst_pop2.pop3  -<span style="color: #BB0000;">0.5</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> rs6   fst_pop1.pop2  -<span style="color: #BB0000;">0.25</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> rs6   fst_pop1.pop3  -<span style="color: #BB0000;">0.333</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> rs6   fst_pop2.pop3  -<span style="color: #BB0000;">0.5</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Fst</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   population_1 population_2  value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> pop1         pop2         0.034<span style="text-decoration: underline;">5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> pop1         pop3         0.055<span style="text-decoration: underline;">6</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> pop2         pop3         0     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

