<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute basic population global statistics — pop_global_stats • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute basic population global statistics — pop_global_stats"><meta name="description" content="This function computes basic population global statistics, following the
notation in Nei 1987 (which in turn is based on Nei and Chesser 1983):
observed heterozygosity ( \(\hat{h}_o\), column header Ho)
expected heterozygosity, also known as gene diversity
( \(\hat{h}_s\), Hs)
total heterozygosity ( \(\hat{h}_t\), Ht)
genetic differentiation between subpopulations (\(D_{st}\), Dst)
corrected total population diversity (\(h'_t\), Htp)
corrected genetic differentiation between subpopulations
(\(D'_{st}\), Dstp)
\(\hat{F}_{ST}\) (column header, Fst)
corrected \(\hat{F'}_{ST}\) (column header Fstp)
\(\hat{F}_{IS}\) (column header, Fis)
Jost's \(\hat{D}\) (column header, Dest)

"><meta property="og:description" content="This function computes basic population global statistics, following the
notation in Nei 1987 (which in turn is based on Nei and Chesser 1983):
observed heterozygosity ( \(\hat{h}_o\), column header Ho)
expected heterozygosity, also known as gene diversity
( \(\hat{h}_s\), Hs)
total heterozygosity ( \(\hat{h}_t\), Ht)
genetic differentiation between subpopulations (\(D_{st}\), Dst)
corrected total population diversity (\(h'_t\), Htp)
corrected genetic differentiation between subpopulations
(\(D'_{st}\), Dstp)
\(\hat{F}_{ST}\) (column header, Fst)
corrected \(\hat{F'}_{ST}\) (column header Fstp)
\(\hat{F}_{IS}\) (column header, Fis)
Jost's \(\hat{D}\) (column header, Dest)

"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute basic population global statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/R/pop_global_stats.R" class="external-link"><code>R/pop_global_stats.R</code></a></small>
      <div class="d-none name"><code>pop_global_stats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes basic population global statistics, following the
notation in Nei 1987 (which in turn is based on Nei and Chesser 1983):</p><ul><li><p>observed heterozygosity ( \(\hat{h}_o\), column header <code>Ho</code>)</p></li>
<li><p>expected heterozygosity, also known as gene diversity
( \(\hat{h}_s\), <code>Hs</code>)</p></li>
<li><p>total heterozygosity ( \(\hat{h}_t\), <code>Ht</code>)</p></li>
<li><p>genetic differentiation between subpopulations (\(D_{st}\), <code>Dst</code>)</p></li>
<li><p>corrected total population diversity (\(h'_t\), <code>Htp</code>)</p></li>
<li><p>corrected genetic differentiation between subpopulations
(\(D'_{st}\), <code>Dstp</code>)</p></li>
<li><p>\(\hat{F}_{ST}\) (column header, <code>Fst</code>)</p></li>
<li><p>corrected \(\hat{F'}_{ST}\) (column header <code>Fstp</code>)</p></li>
<li><p>\(\hat{F}_{IS}\) (column header, <code>Fis</code>)</p></li>
<li><p>Jost's \(\hat{D}\) (column header, <code>Dest</code>)</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pop_global_stats</span><span class="op">(</span><span class="va">.x</span>, by_locus <span class="op">=</span> <span class="cn">FALSE</span>, n_cores <span class="op">=</span> <span class="fu">bigstatsr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg--x">.x<a class="anchor" aria-label="anchor" href="#arg--x"></a></dt>
<dd><p>a <code><a href="gen_tibble.html">gen_tibble</a></code> (usually grouped, as obtained by using
<code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">dplyr::group_by()</a></code>; use on a single population will return a number of
quantities as NA/NaN)</p></dd>


<dt id="arg-by-locus">by_locus<a class="anchor" aria-label="anchor" href="#arg-by-locus"></a></dt>
<dd><p>boolean, determining whether the statistics should be
returned by locus(TRUE), or as a single genome wide value (FALSE, the
default).</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>number of cores to be used, it defaults to
<code><a href="https://privefl.github.io/bigstatsr/reference/reexports.html" class="external-link">bigstatsr::nb_cores()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a tibble of population statistics, with populations as rows and
statistics as columns</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>We use the notation of Nei 1987. That notation was for loci with
\(m\) alleles, but in our case we only have two alleles, so <code>m=2</code>.</p><ul><li><p>Within population observed heterozygosity \(\hat{h}_o\) for a locus with
\(m\) alleles is defined as:<br>
\(\hat{h}_o= 1-\sum_{k=1}^{s} \sum_{i=1}^{m}  \hat{X}_{kii}/s\)<br>
where<br> \(\hat{X}_{kii}\) represents the proportion of homozygote
\(i\) in the sample for the \(k\)th population and<br> \(s\) the
number of populations,<br> following equation 7.38 in Nei(1987) on
pp.164.<br></p></li>
<li><p>Within population expected heterozygosity (gene diversity)
\(\hat{h}_s\) for a locus with \(m\) alleles is defined as:<br>
\(\hat{h}_s=(\tilde{n}/(\tilde{n}-1))[1-\sum_{i=1}^{m}\bar{\hat{x}_i^2}-\hat{h}_o/2\tilde{n}]\)<br> #nolint
where <br> \(\tilde{n}=s/\sum_k 1/n_k\) (i.e the harmonic mean of
\(n_k\)) and<br> \(\bar{\hat{x}_i^2}=\sum_k \hat{x}_{ki}^2/s\)<br>
following equation 7.39 in Nei(1987) on pp.164.</p></li>
<li><p>Total heterozygosity (total gene diversity)
\(\hat{h}_t\) for a locus with \(m\) alleles is defined as:<br>
\(\hat{h}_t = 1-\sum_{i=1}^{m} \bar{\hat{x}_i^2} +
  \hat{h}_s/(\tilde{n}s) - \hat{h}_o/(2\tilde{n}s)\)<br> where <br>
\(\hat{x}_i=\sum_k \hat{x}_{ki}/s\)<br> following equation 7.40 in
Nei(1987) on pp.164.<br></p></li>
<li><p>The amount of gene diversity among samples \(D_{ST}\) is defined as:<br>
\(D_{ST} = \hat{h}_t - \hat{h}_s\)<br> following the equation provided in
the text at the top of page 165 in Nei(1987).</p></li>
<li><p>The corrected amount of gene diversity among samples
\(D'_{ST}\) is defined as:<br>
\(D'_{ST} = (s/(s-1))D'_{ST}\)<br> following the equation provided in the
text at the top of page 165 in Nei(1987).</p></li>
<li><p>Total corrected heterozygosity (total gene diversity)
\(\hat{h}_t\) is defined as:<br>
\(\hat{h'}_t = \hat{h}_s + D'_{ST}\)<br> following the equation provided
in the text at the top of page 165 in Nei(1987).</p></li>
<li><p>\(\hat{F}_{IS}\) is defined as:<br>
\(\hat{F}_{IS} = 1 - \hat{h}_o/\hat{h}_s\)<br> following equation 7.41 in
Nei(1987) on pp.164.<br></p></li>
<li><p>\(\hat{F}_{ST}\) is defined as:<br>
\(\hat{F}_{ST} = 1 - \hat{h}_s/\hat{h}_t = D_{ST}/\hat{h}_t\)<br>
following equation 7.43 in Nei(1987) on pp.165.<br></p></li>
<li><p>\(\hat{F'}_{ST}\) is defined as:<br>
\(\hat{F'}_{ST} = D'_{ST}/\hat{h'}_t\)<br> following the explanation
provided in the text at the top of page 165 in Nei(1987).</p></li>
<li><p>Jost's \(\hat{D}\) is defined as:<br>
\(\hat{D} = (s/(s-1))((\hat{h'}_t-\hat{h}_s)/(1-\hat{h}_s))\)<br> as
defined by Jost(2008)</p>
<p>All these statistics are first computed by locus, and then averaged across
loci (including any monorphic locus) to obtain genome-wide values. The
function uses the same algorithm as <code><a href="https://rdrr.io/pkg/hierfstat/man/basic.stats.html" class="external-link">hierfstat::basic.stats()</a></code> but is
optimized for speed and memory usage.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Nei M, Chesser R (1983) Estimation of fixation indexes and gene
diversities. Annals of Human Genetics, 47, 253-259. Nei M. (1987) Molecular
Evolutionary Genetics. Columbia University Press, pp. 164-165. Jost L
(2008) GST and its relatives do not measure differentiation. Molecular
Ecology, 17, 4015-4026.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

