<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate a report of what would happen to each SNP in a merge — rbind_dry_run • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a report of what would happen to each SNP in a merge — rbind_dry_run"><meta name="description" content="This function provides an overview of the fate of each SNP in two
gen_tibble objects in the case of a merge. Only SNPs found in both
objects will be kept. One object is used as a reference, and SNPs in the
other dataset will be flipped and/or alleles swapped as needed. SNPs that
have different alleles in the two datasets will also be dropped."><meta property="og:description" content="This function provides an overview of the fate of each SNP in two
gen_tibble objects in the case of a merge. Only SNPs found in both
objects will be kept. One object is used as a reference, and SNPs in the
other dataset will be flipped and/or alleles swapped as needed. SNPs that
have different alleles in the two datasets will also be dropped."><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generate a report of what would happen to each SNP in a merge</h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/R/rbind_dry_run.R" class="external-link"><code>R/rbind_dry_run.R</code></a></small>
      <div class="d-none name"><code>rbind_dry_run.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function provides an overview of the fate of each SNP in two
<code><a href="gen_tibble.html">gen_tibble</a></code> objects in the case of a merge. Only SNPs found in both
objects will be kept. One object is used as a <code>reference</code>, and SNPs in the
other dataset will be flipped and/or alleles swapped as needed. SNPs that
have different alleles in the two datasets will also be dropped.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rbind_dry_run</span><span class="op">(</span></span>
<span>  <span class="va">ref</span>,</span>
<span>  <span class="va">target</span>,</span>
<span>  use_position <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  flip_strand <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ref">ref<a class="anchor" aria-label="anchor" href="#arg-ref"></a></dt>
<dd><p>either a <code><a href="gen_tibble.html">gen_tibble</a></code> object, or the path to the PLINK bim file;
the alleles in this objects will be used as template to flip the ones in
<code>target</code> and/or swap their order as necessary.</p></dd>


<dt id="arg-target">target<a class="anchor" aria-label="anchor" href="#arg-target"></a></dt>
<dd><p>either a <code><a href="gen_tibble.html">gen_tibble</a></code> object, or the path to the PLINK bim
file</p></dd>


<dt id="arg-use-position">use_position<a class="anchor" aria-label="anchor" href="#arg-use-position"></a></dt>
<dd><p>boolean of whether a combination of chromosome and
position should be used for matching SNPs. By default, <code>rbind</code> uses the
locus name, so this is set to FALSE. When using 'use_position=TRUE', make
sure chromosomes are coded in the same way in both <code>gen_tibbles</code> (a mix of
e.g. 'chr1', '1' or 'chromosome1' can be the reasons if an unexpectedly
large number variants are dropped when merging).</p></dd>


<dt id="arg-flip-strand">flip_strand<a class="anchor" aria-label="anchor" href="#arg-flip-strand"></a></dt>
<dd><p>boolean on whether strand flipping should be checked to
match the two datasets. Ambiguous SNPs (i.e. A/T and C/G) will also be
removed.  It defaults to FALSE</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>boolean whether to omit reporting to screen</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list with two <code>data.frames</code>, named <code>target</code> and <code>ref</code>. Each
data.frame has <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code> equal to the number of loci in the respective
dataset, a column <code>id</code> with the locus name, and boolean columns <code>to_keep</code>
(the valid loci that will be kept in the merge), <code>alleles_mismatched</code> (loci
found in both datasets but with mismatched alleles, leading to those loci
being dropped), <code>to_flip</code> (loci that need to be flipped to align the two
datasets, only found in <code>target</code> data.frame) and <code>to_swap</code> (loci for which
the order of alleles needs to be swapped to align the two datasets,
<code>target</code> data.frame)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="example_gt.html">example_gt</a></span><span class="op">(</span><span class="st">"gen_tbl"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create a second gen_tibble to merge</span></span></span>
<span class="r-in"><span><span class="va">test_indiv_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pop1"</span>, <span class="st">"pop1"</span>, <span class="st">"pop2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test_loci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rs"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  chromosome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">65</span>, <span class="fl">343</span>, <span class="fl">23</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  genetic_dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/double.html" class="external-link">as.double</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  allele_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"C"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  allele_alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T"</span>, <span class="st">"C"</span>, <span class="cn">NA</span>, <span class="st">"C"</span>, <span class="st">"G"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">test_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="gen_tibble.html">gen_tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  x <span class="op">=</span> <span class="va">test_genotypes</span>,</span></span>
<span class="r-in"><span>  loci <span class="op">=</span> <span class="va">test_loci</span>,</span></span>
<span class="r-in"><span>  indiv_meta <span class="op">=</span> <span class="va">test_indiv_meta</span>,</span></span>
<span class="r-in"><span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  quiet <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create an rbind report using rbind_dry_run</span></span></span>
<span class="r-in"><span><span class="fu">rbind_dry_run</span><span class="op">(</span><span class="va">example_gt</span>, <span class="va">test_gt</span>, flip_strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> harmonising loci between two datasets</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> flip_strand =  TRUE  ; remove_ambiguous =  TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -----------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dataset: reference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> number of SNPs: 6 reduced to 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ( 4 are ambiguous, of which 4  were removed)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -----------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dataset: target </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> number of SNPs: 5 reduced to 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ( 0 were flipped to match the reference set)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ( 3 are ambiguous, of which 3 were removed)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $target</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   id new_id name to_flip to_swap missing_allele ambiguous</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  1     NA  rs1   FALSE   FALSE             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  2      1  rs2   FALSE   FALSE             NA     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  3      2  rs3   FALSE   FALSE             NA     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  4     NA  rs4   FALSE   FALSE             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  5     NA  rs5   FALSE   FALSE             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ref</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   id new_id name missing_allele ambiguous</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  1     NA  rs1             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  2      1  rs2             NA     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  3      2  rs3             NA     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  4     NA  rs4             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  5     NA  rs5             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  6     NA  rs6             NA      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "rbind_report" "list"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"flip_strand")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"remove_ambiguous")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

