<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"><meta name="robots" content="noindex"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.3.2.9004</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="-gen-tibble-"><code>gen_tibble</code><a class="anchor" aria-label="anchor" href="#-gen-tibble-"></a></h2>

      <div class="section-desc"><p>Functions for creating, saving, and loading <code>gen_tibble</code> objects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gen_tibble.html">gen_tibble()</a></code>

        </dt>
        <dd>Constructor for a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="gt_save.html">gt_save()</a></code>

        </dt>
        <dd>Save a gen_tibble</dd>
      </dl><dl><dt>

          <code><a href="gt_load.html">gt_load()</a></code>

        </dt>
        <dd>Load a gen_tibble</dd>
      </dl><dl><dt>

          <code><a href="gt_get_file_names.html">gt_get_file_names()</a></code>

        </dt>
        <dd>Get the names of files storing the genotypes of a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="gt_order_loci.html">gt_order_loci()</a></code>

        </dt>
        <dd>Order the loci table of a gen_tibble</dd>
      </dl><dl><dt>

          <code><a href="gt_update_backingfile.html">gt_update_backingfile()</a></code>

        </dt>
        <dd>Update the backing matrix</dd>
      </dl><dl><dt>

          <code><a href="gt_add_sf.html">gt_add_sf()</a></code>

        </dt>
        <dd>Add an simple feature geometry to a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="gt_pseudohaploid.html">gt_pseudohaploid()</a></code>

        </dt>
        <dd>Set the ploidy of a <code>gen_tibble</code> to include pseudohaploids</dd>
      </dl></div><div class="section level2">
      <h2 id="show">Show<a class="anchor" aria-label="anchor" href="#show"></a></h2>

      <div class="section-desc"><p>Functions to view loci and genotype information stored in a <code>gen_tibble</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="show_genotypes.html">show_genotypes()</a></code>

        </dt>
        <dd>Show the genotypes of a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="show_loci.html">show_loci()</a></code> <code><a href="show_loci.html">`show_loci&lt;-`()</a></code>

        </dt>
        <dd>Show the loci information of a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="show_ploidy.html">show_ploidy()</a></code>

        </dt>
        <dd>Show the ploidy information of a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="count_loci.html">count_loci()</a></code>

        </dt>
        <dd>Count the number of loci in a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="augment_loci.html">augment_loci()</a></code>

        </dt>
        <dd>Augment the loci table with information from a analysis object</dd>
      </dl><dl><dt>

          <code><a href="is_loci_table_ordered.html">is_loci_table_ordered()</a></code>

        </dt>
        <dd>Test if the loci table is ordered</dd>
      </dl><dl><dt>

          <code><a href="find_duplicated_loci.html">find_duplicated_loci()</a></code>

        </dt>
        <dd>Find duplicates in the loci table</dd>
      </dl></div><div class="section level2">
      <h2 id="selecting-loci">Selecting loci<a class="anchor" aria-label="anchor" href="#selecting-loci"></a></h2>

      <div class="section-desc"><p>Functions to select loci from a <code>gen_tibble</code>, either based on loci information, or genotype information.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="select_loci.html">select_loci()</a></code>

        </dt>
        <dd>The <code>select</code> verb for <code>loci</code></dd>
      </dl><dl><dt>

          <code><a href="select_loci_if.html">select_loci_if()</a></code>

        </dt>
        <dd>The <code>select_if</code> verb for <code>loci</code></dd>
      </dl></div><div class="section level2">
      <h2 id="imputing">Imputing<a class="anchor" aria-label="anchor" href="#imputing"></a></h2>

      <div class="section-desc"><p>Functions for imputing missing data in a <code>gen_tibble</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_impute_simple.html">gt_impute_simple()</a></code>

        </dt>
        <dd>Simple imputation based on allele frequencies</dd>
      </dl><dl><dt>

          <code><a href="gt_impute_xgboost.html">gt_impute_xgboost()</a></code>

        </dt>
        <dd>Imputation based XGBoost</dd>
      </dl><dl><dt>

          <code><a href="gt_has_imputed.html">gt_has_imputed()</a></code>

        </dt>
        <dd>Checks if a <code>gen_tibble</code> has been imputed</dd>
      </dl><dl><dt>

          <code><a href="gt_uses_imputed.html">gt_uses_imputed()</a></code>

        </dt>
        <dd>Checks if a <code>gen_tibble</code> uses imputed data</dd>
      </dl><dl><dt>

          <code><a href="gt_set_imputed.html">gt_set_imputed()</a></code>

        </dt>
        <dd>Sets a <code>gen_tibble</code> to use imputed data</dd>
      </dl></div><div class="section level2">
      <h2 id="merging">Merging<a class="anchor" aria-label="anchor" href="#merging"></a></h2>

      <div class="section-desc"><p>Functions for merging <code>gen_tibble</code> objects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rbind_dry_run.html">rbind_dry_run()</a></code>

        </dt>
        <dd>Generate a report of what would happen to each SNP in a merge</dd>
      </dl><dl><dt>

          <code><a href="rbind.gen_tbl.html">rbind(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>Combine two gen_tibbles</dd>
      </dl><dl><dt>

          <code><a href="summary_rbind_dry_run.html">summary(<i>&lt;rbind_report&gt;</i>)</a></code>

        </dt>
        <dd>Print a summary of a merge report</dd>
      </dl><dl><dt>

          <code><a href="cbind.gen_tbl.html">cbind(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>Combine a gen_tibble to a data.frame or tibble by column</dd>
      </dl></div><div class="section level2">
      <h2 id="exporting">Exporting<a class="anchor" aria-label="anchor" href="#exporting"></a></h2>

      <div class="section-desc"><p>Functions for exporting a <code>gen_tibble</code> to another format.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_as_genind.html">gt_as_genind()</a></code>

        </dt>
        <dd>Convert a <code>gen_tibble</code> to a <code>genind</code> object from <code>adegenet</code></dd>
      </dl><dl><dt>

          <code><a href="gt_as_genlight.html">gt_as_genlight()</a></code>

        </dt>
        <dd>Convert a <code>gen_tibble</code> to a <code>genlight</code> object from <code>adegenet</code></dd>
      </dl><dl><dt>

          <code><a href="gt_as_geno_lea.html">gt_as_geno_lea()</a></code>

        </dt>
        <dd>Convert a <code>gentibble</code> to a .geno file for sNMF from the LEA package</dd>
      </dl><dl><dt>

          <code><a href="gt_as_hierfstat.html">gt_as_hierfstat()</a></code>

        </dt>
        <dd>Convert a <code>gen_tibble</code> to a data.frame compatible with <code>hierfstat</code></dd>
      </dl><dl><dt>

          <code><a href="gt_as_plink.html">gt_as_plink()</a></code>

        </dt>
        <dd>Export a <code>gen_tibble</code> object to PLINK bed format</dd>
      </dl><dl><dt>

          <code><a href="gt_as_vcf.html">gt_as_vcf()</a></code>

        </dt>
        <dd>Convert a <code>gen_tibble</code> to a VCF</dd>
      </dl></div><div class="section level2">
      <h2 id="importing">Importing<a class="anchor" aria-label="anchor" href="#importing"></a></h2>

      <div class="section-desc"><p>Functions for importing another format into a <code>gen_tibble</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_from_genlight.html">gt_from_genlight()</a></code>

        </dt>
        <dd>Convert a <code>genlight</code> object from adegenet to a <code>gen_tibble</code></dd>
      </dl></div><div class="section level2">
      <h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a></h2>

      <div class="section-desc"><p>Functions reporting summary statistics for a <code>gen_tibble</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="qc_report_loci.html">qc_report_loci()</a></code>

        </dt>
        <dd>Create a Quality Control report for loci</dd>
      </dl><dl><dt>

          <code><a href="qc_report_indiv.html">qc_report_indiv()</a></code>

        </dt>
        <dd>Create a Quality Control report for individuals</dd>
      </dl><dl><dt>

          <code><a href="autoplot.qc_report_loci.html">autoplot(<i>&lt;qc_report_loci&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>qc_report_loci</code> objects</dd>
      </dl><dl><dt>

          <code><a href="autoplot.qc_report_indiv.html">autoplot(<i>&lt;qc_report_indiv&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>qc_report_indiv</code> objects</dd>
      </dl><dl><dt>

          <code><a href="filter_high_relatedness.html">filter_high_relatedness()</a></code>

        </dt>
        <dd>Filter individuals based on a relationship threshold</dd>
      </dl></div><div class="section level2">
      <h2 id="loci">Loci<a class="anchor" aria-label="anchor" href="#loci"></a></h2>

      <div class="section-desc"><p>Functions operating across loci.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="loci_alt_freq.html">loci_alt_freq()</a></code> <code><a href="loci_alt_freq.html">loci_maf()</a></code>

        </dt>
        <dd>Estimate allele frequencies at each locus</dd>
      </dl><dl><dt>

          <code><a href="loci_chromosomes.html">loci_chromosomes()</a></code>

        </dt>
        <dd>Get the chromosomes of loci in a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="loci_hwe.html">loci_hwe()</a></code>

        </dt>
        <dd>Test Hardy-Weinberg equilibrium at each locus</dd>
      </dl><dl><dt>

          <code><a href="loci_ld_clump.html">loci_ld_clump()</a></code>

        </dt>
        <dd>Clump loci based on a Linkage Disequilibrium threshold</dd>
      </dl><dl><dt>

          <code><a href="loci_missingness.html">loci_missingness()</a></code>

        </dt>
        <dd>Estimate missingness at each locus</dd>
      </dl><dl><dt>

          <code><a href="loci_names.html">loci_names()</a></code>

        </dt>
        <dd>Get the names of loci in a <code>gen_tibble</code></dd>
      </dl><dl><dt>

          <code><a href="loci_pi.html">loci_pi()</a></code>

        </dt>
        <dd>Estimate nucleotide diversity (pi) at each locus</dd>
      </dl><dl><dt>

          <code><a href="loci_transitions.html">loci_transitions()</a></code>

        </dt>
        <dd>Find transitions</dd>
      </dl><dl><dt>

          <code><a href="loci_transversions.html">loci_transversions()</a></code>

        </dt>
        <dd>Find transversions</dd>
      </dl></div><div class="section level2">
      <h2 id="individual">Individual<a class="anchor" aria-label="anchor" href="#individual"></a></h2>

      <div class="section-desc"><p>Functions operating across individuals.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="indiv_het_obs.html">indiv_het_obs()</a></code>

        </dt>
        <dd>Estimate individual observed heterozygosity</dd>
      </dl><dl><dt>

          <code><a href="indiv_inbreeding.html">indiv_inbreeding()</a></code>

        </dt>
        <dd>Individual inbreeding coefficient</dd>
      </dl><dl><dt>

          <code><a href="indiv_missingness.html">indiv_missingness()</a></code>

        </dt>
        <dd>Estimate individual missingness</dd>
      </dl><dl><dt>

          <code><a href="indiv_ploidy.html">indiv_ploidy()</a></code>

        </dt>
        <dd>Return individual ploidy</dd>
      </dl></div><div class="section level2">
      <h2 id="population">Population<a class="anchor" aria-label="anchor" href="#population"></a></h2>

      <div class="section-desc"><p>Functions operating on a <code>gen_tibble</code> grouped by population.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="pop_fis.html">pop_fis()</a></code>

        </dt>
        <dd>Compute population specific FIS</dd>
      </dl><dl><dt>

          <code><a href="pop_fst.html">pop_fst()</a></code>

        </dt>
        <dd>Compute population specific Fst</dd>
      </dl><dl><dt>

          <code><a href="pop_global_stats.html">pop_global_stats()</a></code>

        </dt>
        <dd>Compute basic population global statistics</dd>
      </dl><dl><dt>

          <code><a href="pop_het_exp.html">pop_het_exp()</a></code> <code><a href="pop_het_exp.html">pop_gene_div()</a></code>

        </dt>
        <dd>Compute the population expected heterozygosity</dd>
      </dl><dl><dt>

          <code><a href="pop_het_obs.html">pop_het_obs()</a></code>

        </dt>
        <dd>Compute the population observed heterozygosity</dd>
      </dl><dl><dt>

          <code><a href="pop_tajimas_d.html">pop_tajimas_d()</a></code>

        </dt>
        <dd>Estimate Tajima's D for the whole genome</dd>
      </dl></div><div class="section level2">
      <h2 id="pairwise">Pairwise<a class="anchor" aria-label="anchor" href="#pairwise"></a></h2>

      <div class="section-desc"><p>Functions comparing individuals or populations.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="pairwise_allele_sharing.html">pairwise_allele_sharing()</a></code>

        </dt>
        <dd>Compute the Pairwise Allele Sharing Matrix for a <code>gen_tibble</code> object</dd>
      </dl><dl><dt>

          <code><a href="pairwise_grm.html">pairwise_grm()</a></code>

        </dt>
        <dd>Compute the Genomic Relationship Matrix for a <code>gen_tibble</code> object</dd>
      </dl><dl><dt>

          <code><a href="pairwise_ibs.html">pairwise_ibs()</a></code>

        </dt>
        <dd>Compute the Identity by State Matrix for a <code>gen_tibble</code> object</dd>
      </dl><dl><dt>

          <code><a href="pairwise_king.html">pairwise_king()</a></code>

        </dt>
        <dd>Compute the KING-robust Matrix for a <code>gen_tibble</code> object</dd>
      </dl><dl><dt>

          <code><a href="pairwise_pop_fst.html">pairwise_pop_fst()</a></code>

        </dt>
        <dd>Compute pairwise population Fst</dd>
      </dl><dl><dt>

          <code><a href="nwise_pop_pbs.html">nwise_pop_pbs()</a></code>

        </dt>
        <dd>Compute the Population Branch Statistics for each combination of populations</dd>
      </dl></div><div class="section level2">
      <h2 id="windows">Windows<a class="anchor" aria-label="anchor" href="#windows"></a></h2>

      <div class="section-desc"><p>Functions to compute statistics in sliding windows across loci.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="windows_indiv_roh.html">windows_indiv_roh()</a></code> <code><a href="windows_indiv_roh.html">gt_roh_window()</a></code>

        </dt>
        <dd>Detect runs of homozygosity using a sliding-window approach</dd>
      </dl><dl><dt>

          <code><a href="windows_nwise_pop_pbs.html">windows_nwise_pop_pbs()</a></code>

        </dt>
        <dd>Compute the Population Branch Statistics over a sliding window</dd>
      </dl><dl><dt>

          <code><a href="windows_pairwise_pop_fst.html">windows_pairwise_pop_fst()</a></code>

        </dt>
        <dd>Compute pairwise Fst for a sliding window</dd>
      </dl><dl><dt>

          <code><a href="windows_pop_tajimas_d.html">windows_pop_tajimas_d()</a></code>

        </dt>
        <dd>Compute Tajima's D for a sliding window</dd>
      </dl><dl><dt>

          <code><a href="windows_stats_generic.html">windows_stats_generic()</a></code>

        </dt>
        <dd>Estimate window statistics from per locus estimates</dd>
      </dl></div><div class="section level2">
      <h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a></h2>

      <div class="section-desc"><p>Functions for performing Principal Component Analysis.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_pca.html">gt_pca</a></code>

        </dt>
        <dd>Principal Component Analysis for <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="gt_pca_autoSVD.html">gt_pca_autoSVD()</a></code>

        </dt>
        <dd>PCA controlling for LD for <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="gt_pca_partialSVD.html">gt_pca_partialSVD()</a></code>

        </dt>
        <dd>PCA for <code>gen_tibble</code> objects by partial SVD</dd>
      </dl><dl><dt>

          <code><a href="gt_pca_randomSVD.html">gt_pca_randomSVD()</a></code>

        </dt>
        <dd>PCA for <code>gen_tibble</code> objects by randomized partial SVD</dd>
      </dl><dl><dt>

          <code><a href="predict_gt_pca.html">predict(<i>&lt;gt_pca&gt;</i>)</a></code>

        </dt>
        <dd>Predict scores of a PCA</dd>
      </dl><dl><dt>

          <code><a href="tidy_gt_pca.html">tidy(<i>&lt;gt_pca&gt;</i>)</a></code>

        </dt>
        <dd>Tidy a <code>gt_pca</code> object</dd>
      </dl><dl><dt>

          <code><a href="augment_gt_pca.html">augment(<i>&lt;gt_pca&gt;</i>)</a></code>

        </dt>
        <dd>Augment data with information from a gt_pca object</dd>
      </dl><dl><dt>

          <code><a href="augment_loci_gt_pca.html">augment_loci(<i>&lt;gt_pca&gt;</i>)</a></code>

        </dt>
        <dd>Augment the loci table with information from a gt_pca object</dd>
      </dl><dl><dt>

          <code><a href="autoplot_gt_pca.html">autoplot(<i>&lt;gt_pca&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>gt_pca</code> objects</dd>
      </dl></div><div class="section level2">
      <h2 id="dapc">DAPC<a class="anchor" aria-label="anchor" href="#dapc"></a></h2>

      <div class="section-desc"><p>Functions for performing Discriminant Analysis of Principal Components.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_dapc.html">gt_dapc()</a></code>

        </dt>
        <dd>Discriminant Analysis of Principal Components for gen_tibble</dd>
      </dl><dl><dt>

          <code><a href="gt_cluster_pca.html">gt_cluster_pca()</a></code>

        </dt>
        <dd>Run K-clustering on principal components</dd>
      </dl><dl><dt>

          <code><a href="gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>

        </dt>
        <dd>Find the best number of clusters based on principal components</dd>
      </dl><dl><dt>

          <code><a href="tidy.gt_dapc.html">tidy(<i>&lt;gt_dapc&gt;</i>)</a></code>

        </dt>
        <dd>Tidy a <code>gt_dapc</code> object</dd>
      </dl><dl><dt>

          <code><a href="augment.gt_dapc.html">augment(<i>&lt;gt_dapc&gt;</i>)</a></code>

        </dt>
        <dd>Augment data with information from a gt_dapc object</dd>
      </dl><dl><dt>

          <code><a href="autoplot_gt_dapc.html">autoplot(<i>&lt;gt_dapc&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>gt_dapc</code> objects</dd>
      </dl><dl><dt>

          <code><a href="autoplot_gt_cluster_pca.html">autoplot(<i>&lt;gt_cluster_pca&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>gt_cluster_pca</code> objects</dd>
      </dl></div><div class="section level2">
      <h2 id="k-clustering-and-admxiture">K-clustering and ADMXITURE<a class="anchor" aria-label="anchor" href="#k-clustering-and-admxiture"></a></h2>

      <div class="section-desc"><p>Functions for creating, tidying, and visualising <code>gt_admix</code> and <code>q_matrix</code> objects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_admixture.html">gt_admixture()</a></code>

        </dt>
        <dd>Run ADMIXTURE from R</dd>
      </dl><dl><dt>

          <code><a href="gt_snmf.html">gt_snmf()</a></code>

        </dt>
        <dd>Run SNMF from R in tidypopgen</dd>
      </dl><dl><dt>

          <code><a href="autoplot_gt_admix.html">autoplot(<i>&lt;gt_admix&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>gt_admix</code> objects</dd>
      </dl><dl><dt>

          <code><a href="summary.gt_admix.html">summary(<i>&lt;gt_admix&gt;</i>)</a></code>

        </dt>
        <dd>Summary method for gt_admix objects</dd>
      </dl><dl><dt>

          <code><a href="c.gt_admix.html">c(<i>&lt;gt_admix&gt;</i>)</a></code>

        </dt>
        <dd>Combine method for gt_admix objects</dd>
      </dl><dl><dt>

          <code><a href="gt_admix_reorder_q.html">gt_admix_reorder_q()</a></code>

        </dt>
        <dd>Reorder the q matrices based on the grouping variable</dd>
      </dl><dl><dt>

          <code><a href="q_matrix.html">q_matrix()</a></code>

        </dt>
        <dd>Convert a standard matrix to a <code>q_matrix</code> object</dd>
      </dl><dl><dt>

          <code><a href="read_q_files.html">read_q_files()</a></code>

        </dt>
        <dd>Read and structure .Q files or existing matrices as <code>q_matrix</code> or <code>gt_admix</code> objects.</dd>
      </dl><dl><dt>

          <code><a href="get_q_matrix.html">get_q_matrix()</a></code>

        </dt>
        <dd>Return a single Q matrix from a <code>gt_admix</code> object</dd>
      </dl><dl><dt>

          <code><a href="get_p_matrix.html">get_p_matrix()</a></code>

        </dt>
        <dd>Return a single P matrix from a <code>gt_admix</code> object</dd>
      </dl><dl><dt>

          <code><a href="tidy.q_matrix.html">tidy(<i>&lt;q_matrix&gt;</i>)</a></code>

        </dt>
        <dd>Tidy a Q matrix</dd>
      </dl><dl><dt>

          <code><a href="augment_q_matrix.html">augment(<i>&lt;q_matrix&gt;</i>)</a></code>

        </dt>
        <dd>Augment data with information from a q_matrix object</dd>
      </dl><dl><dt>

          <code><a href="autoplot_q_matrix.html">autoplot(<i>&lt;q_matrix&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>q_matrix</code> objects</dd>
      </dl><dl><dt>

          <code><a href="distruct_colours.html">distruct_colours</a></code>

        </dt>
        <dd>Distruct colours</dd>
      </dl><dl><dt>

          <code><a href="scale_fill_distruct.html">scale_fill_distruct()</a></code>

        </dt>
        <dd>Scale constructor using the distruct colours</dd>
      </dl><dl><dt>

          <code><a href="theme_distruct.html">theme_distruct()</a></code>

        </dt>
        <dd>A theme to match the output of distruct</dd>
      </dl></div><div class="section level2">
      <h2 id="other-analyses-e-g-admixtools-pcadapt-">Other analyses (e.g. ADMIXTOOLS2, pcadapt)<a class="anchor" aria-label="anchor" href="#other-analyses-e-g-admixtools-pcadapt-"></a></h2>

      <div class="section-desc"><p>Functions for other analyses within <code>R</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gt_extract_f2.html">gt_extract_f2()</a></code>

        </dt>
        <dd>Compute and store blocked f2 statistics for ADMIXTOOLS 2</dd>
      </dl><dl><dt>

          <code><a href="gt_pcadapt.html">gt_pcadapt()</a></code>

        </dt>
        <dd>pcadapt analysis on a <code>gen_tibble</code> object</dd>
      </dl><dl><dt>

          <code><a href="autoplot_gt_pcadapt.html">autoplot(<i>&lt;gt_pcadapt&gt;</i>)</a></code>

        </dt>
        <dd>Autoplots for <code>gt_pcadapt</code> objects</dd>
      </dl></div><div class="section level2">
      <h2 id="-bigsnp-"><code>bigSNP</code><a class="anchor" aria-label="anchor" href="#-bigsnp-"></a></h2>

      <div class="section-desc"><p>Mostly for developers: Functions for operating directly on <code>bigSNP</code> objects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="snp_allele_sharing.html">snp_allele_sharing()</a></code>

        </dt>
        <dd>Compute the Pairwise Allele Sharing Matrix for a bigSNP object</dd>
      </dl><dl><dt>

          <code><a href="snp_ibs.html">snp_ibs()</a></code>

        </dt>
        <dd>Compute the Identity by State Matrix for a bigSNP object</dd>
      </dl><dl><dt>

          <code><a href="snp_king.html">snp_king()</a></code>

        </dt>
        <dd>Compute the KING-robust Matrix for a bigSNP object</dd>
      </dl></div><div class="section level2">
      <h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a></h2>

      <div class="section-desc"><p>Function for loading an example <code>gen_tibble</code>.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="load_example_gt.html">load_example_gt()</a></code>

        </dt>
        <dd>Load example gen_tibble</dd>
      </dl></div><div class="section level2">
      <h2 id="tidyverse-methods">tidyverse methods<a class="anchor" aria-label="anchor" href="#tidyverse-methods"></a></h2>

      <div class="section-desc"><p>Methods for tidyverse verbs for <code>gen_tibble</code> objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="arrange.gen_tbl.html">arrange(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>An arrange method for <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="arrange.grouped_gen_tbl.html">arrange(<i>&lt;grouped_gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>An arrange method for grouped <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="filter.gen_tbl.html">filter(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>Tidyverse methods for gt objects</dd>
      </dl><dl><dt>

          <code><a href="filter.grouped_gen_tbl.html">filter(<i>&lt;grouped_gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>A filter method for grouped <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="mutate.gen_tbl.html">mutate(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>A mutate method for <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="mutate.grouped_gen_tbl.html">mutate(<i>&lt;grouped_gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>A mutate method for grouped <code>gen_tibble</code> objects</dd>
      </dl><dl><dt>

          <code><a href="cash-set-.gen_tbl.html">`$&lt;-`(<i>&lt;gen_tbl&gt;</i>)</a></code>

        </dt>
        <dd>A $ method for <code>gen_tibble</code> objects</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

