<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Clump loci based on a Linkage Disequilibrium threshold — loci_ld_clump • tidypopgen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Clump loci based on a Linkage Disequilibrium threshold — loci_ld_clump"><meta name="description" content='This function uses clumping to remove SNPs at high LD. When used with its
default options, clumping based on MAF is similar to standard pruning (as
done by PLINK with "–indep-pairwise (size+1) 1 thr.r2", but it results in a
better spread of SNPs over the chromosome.'><meta property="og:description" content='This function uses clumping to remove SNPs at high LD. When used with its
default options, clumping based on MAF is similar to standard pruning (as
done by PLINK with "–indep-pairwise (size+1) 1 thr.r2", but it results in a
better spread of SNPs over the chromosome.'><meta property="og:image" content="https://evolecolgroup.github.io/tidypopgen/logo.png"><meta name="robots" content="noindex"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.3.0.9001</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Clump loci based on a Linkage Disequilibrium threshold</h1>
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/dev/R/loci_ld_clump.R" class="external-link"><code>R/loci_ld_clump.R</code></a></small>
      <div class="d-none name"><code>loci_ld_clump.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function uses clumping to remove SNPs at high LD. When used with its
default options, clumping based on MAF is similar to standard pruning (as
done by PLINK with "–indep-pairwise (size+1) 1 thr.r2", but it results in a
better spread of SNPs over the chromosome.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">loci_ld_clump</span><span class="op">(</span><span class="va">.x</span>, .col <span class="op">=</span> <span class="st">"genotypes"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'tbl_df'</span></span>
<span><span class="fu">loci_ld_clump</span><span class="op">(</span><span class="va">.x</span>, .col <span class="op">=</span> <span class="st">"genotypes"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'vctrs_bigSNP'</span></span>
<span><span class="fu">loci_ld_clump</span><span class="op">(</span></span>
<span>  <span class="va">.x</span>,</span>
<span>  .col <span class="op">=</span> <span class="st">"genotypes"</span>,</span>
<span>  S <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thr_r2 <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">100</span><span class="op">/</span><span class="va">thr_r2</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_positions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  return_id <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg--x">.x<a class="anchor" aria-label="anchor" href="#arg--x"></a></dt>
<dd><p>a <code><a href="gen_tibble.html">gen_tibble</a></code> object</p></dd>


<dt id="arg--col">.col<a class="anchor" aria-label="anchor" href="#arg--col"></a></dt>
<dd><p>the column to be used when a tibble (or grouped tibble is passed
directly to the function). This defaults to "genotypes" and can only take
that value. There is no need for the user to set it, but it is included to
resolve certain tidyselect operations.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>currently not used.</p></dd>


<dt id="arg-s">S<a class="anchor" aria-label="anchor" href="#arg-s"></a></dt>
<dd><p>A vector of loci statistics which express the importance of each SNP
(the more important is the SNP, the greater should be the corresponding
statistic).<br> For example, if <code>S</code> follows the standard normal
distribution, and "important" means significantly different from 0, you
must use <code>abs(S)</code> instead.<br><strong>If not specified, MAFs are computed and used.</strong></p></dd>


<dt id="arg-thr-r-">thr_r2<a class="anchor" aria-label="anchor" href="#arg-thr-r-"></a></dt>
<dd><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>.</p></dd>


<dt id="arg-size">size<a class="anchor" aria-label="anchor" href="#arg-size"></a></dt>
<dd><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>100 / thr.r2</code> for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt;
200). If <code>use_positions = FALSE</code>, this is a window in number of SNPs,
otherwise it is a window in kb (genetic distance). Ideally, use positions,
as they provide a more sensible approach.</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>Vector of SNP indices to exclude anyway. For example, can be
used to exclude long-range LD regions (see Price2008). Another use can be
for thresholding with respect to p-values associated with <code>S</code>.</p></dd>


<dt id="arg-use-positions">use_positions<a class="anchor" aria-label="anchor" href="#arg-use-positions"></a></dt>
<dd><p>boolean, if TRUE (the default), <code>size</code> is in kb, if
FALSE size is the number of SNPs.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>number of cores to be used</p></dd>


<dt id="arg-return-id">return_id<a class="anchor" aria-label="anchor" href="#arg-return-id"></a></dt>
<dd><p>boolean on whether the id of SNPs to keep should be
returned. It defaults to FALSE, which returns a vector of booleans (TRUE or
FALSE)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a boolean vector indicating whether the SNP should be kept (if
'return_id = FALSE', the default), else a vector of SNP indices to be kept
(if 'return_id = TRUE')</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Any missing values in the genotypes of a <code>gen_tibble</code> passed to
<code>loci_ld_clump</code> will cause an error. To deal with missingness, see
<code><a href="gt_impute_simple.html">gt_impute_simple()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="load_example_gt.html">load_example_gt</a></span><span class="op">(</span><span class="st">"gen_tbl"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To return a boolean vector indicating whether the SNP should be kept</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">loci_ld_clump</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]  TRUE  TRUE FALSE  TRUE  TRUE FALSE</span>
<span class="r-in"><span><span class="co"># To return a vector of SNP indices to be kept</span></span></span>
<span class="r-in"><span><span class="va">example_gt</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">loci_ld_clump</span><span class="op">(</span>return_id <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1 2 4 5</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

