<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>example workflow with tidypopgen • tidypopgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="example workflow with tidypopgen">
<meta property="og:description" content="tidypopgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9013</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_overview.html">overview</a>
    </li>
    <li>
      <a href="../articles/02_qc.html">QC</a>
    </li>
    <li>
      <a href="../articles/03_example_clustering_and_dapc.html">example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/99_plink_cheatsheet.html">plink cheatsheet</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>example workflow with tidypopgen</h1>
            
      
      
      <div class="hidden name"><code>03_example_clustering_and_dapc.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="an-example-workflow-with-real-data">An example workflow with real data<a class="anchor" aria-label="anchor" href="#an-example-workflow-with-real-data"></a>
</h3>
<p>We will explore the genetic structure of <em>Anolis punctatus</em> in
South America, using data from Prates et al 2018. We downloaded the vcf
file of the genotypes from “<a href="https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true" class="external-link uri">https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true</a>”
and compressed it to a vcf.gz file.</p>
<p>We read in the data from the compressed vcf with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidypopgen</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: tibble</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/punctatus_t70_s10_n46_filtered.recode.vcf.gz"</span>,</span>
<span>                          package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">vcf_path</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"anolis_"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s inspect our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 3</span></span></span>
<span><span class="co">#&gt;    id                population  genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> punc_BM288                 0          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> punc_GN71                  0          2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> punc_H1907                 0          3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> punc_H1911                 0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> punc_H2546                 0          5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361          0          6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> punc_ICST764               0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> punc_JFT459                0          8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> punc_JFT773                0          9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> punc_LG1299                0         10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>We can see that we have 46 individuals, from 3249 loci. Note that we
don’t have any information on population from the vcf. That information
can be found from another file on the github repository (“<a href="https://github.com/ivanprates/2018_Anolis_EcolEvol/raw/master/data/plot_order_punctatus_n46.csv" class="external-link uri">https://github.com/ivanprates/2018_Anolis_EcolEvol/raw/master/data/plot_order_punctatus_n46.csv</a>).
We will have add the population information manually. Let’s start by
reading the file:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/plot_order_punctatus_n46.csv"</span>,</span>
<span>                        package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">pops_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">46</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (2): ID, pop</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (1): plot_order</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 46 × 3</span></span></span>
<span><span class="co">#&gt;    ID           plot_order pop  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288                18 Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71                 21 Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907                 6 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911                 4 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546                 7 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361         34 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764              31 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459               37 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773               38 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299               44 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>The ids from the VCF are in a different format than the ones we just
got from the pop csv. We need a bit of string wrangling, but it looks
easy, we just need to remove “punc_”:</p>
<p>Let us simplify the ids, which will have a “punc_” prefix</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'punc_'</span>,<span class="st">""</span>,<span class="va">.data</span><span class="op">$</span><span class="va">id</span>,<span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 3</span></span></span>
<span><span class="co">#&gt;    id           population  genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288                 0          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71                  0          2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907                 0          3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911                 0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546                 0          5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361          0          6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764               0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459                0          8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773                0          9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299                0         10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>Now we can bring in the pop information:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="va">pops</span><span class="op">$</span><span class="va">pop</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">pops</span><span class="op">$</span><span class="va">ID</span>,<span class="va">.data</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 3</span></span></span>
<span><span class="co">#&gt;    id           population  genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288        Eam                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71         Eam                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907        Wam                 3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911        Wam                 4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546        Wam                 5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361 AF                  6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764      AF                  7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459       AF                  8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773       AF                  9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299       AF                 10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>That was easy. The loci had already been filtered and cleaned, so we
don’t need to do any QC. Let us jump straight into analysis and run a
PCA:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: You can't have missing values in 'X'.</span></span></code></pre></div>
<p>OK, we jumped too quickly. There are missing data, and we need first
to impute them:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">anole_gt</span>, method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></code></pre></div>
<p>And now:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Let us look at the object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span></span>
<span><span class="co">#&gt;  === PCA of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Method: [1] "partialSVD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call ($call):gt_pca_partialSVD(x = ., k = 30)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($d):</span></span>
<span><span class="co">#&gt;  351.891 192.527 113.562 104.427 87.615 83.476 ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Principal component scores ($u):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 30 columns (axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings (Principal axes) ($v):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (SNPs) and 30 columns (axes)</span></span></code></pre></div>
<p>The <code>print</code> function (implicitly called when we type the
name of the object) gives us information about the most important
elements in the object (and the names of the elements in which they are
stored).</p>
<p>We can extract those elements with the <code>tidy</code> function,
which returns a tibble that can be easily used for further analysis,
e.g.:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 3</span></span></span>
<span><span class="co">#&gt;       PC std.dev cumulative</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1   51.9        51.9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2   28.4        80.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3   16.7        97.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4   15.4       112. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5   12.9       125. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6   12.3       138. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7   10.2       148. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8    9.92      158. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9    9.13      167. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10    8.80      176. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more rows</span></span></span></code></pre></div>
<p>We can return information on the <em>eigenvalues</em>,
<em>scores</em> and <em>loadings</em> of the pca. There is also an
<code>autoplot</code> method that allows to visualise those elements
(type <em>screeplot</em> for <em>eigenvalues</em>, type <em>scores</em>
for <em>scores</em>, and <em>loadings</em> for <em>loadings</em>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type<span class="op">=</span><span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>To plot the sample in principal coordinates space, we can simply
use:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span><span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p><code>autoplots</code> are deliberately kept simple: they are just a
way to quickly inspect the results. They generate <code>ggplot2</code>
objects, and so they can be further embellished with the usual
<code>ggplot2</code> grammar:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-14-1.png" width="700">
For more complex/publication ready plots, we will want to add the PC
scores to the tibble, so that we can create a custom plot with
<code>ggplot2</code>. We can easily add the data with the
<code>augment</code> method:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span><span class="va">anole_pca</span> , data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
<p>And now we can use <code>ggplot2</code> directly to generate our
plot:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.fittedPC1</span>, <span class="va">.fittedPC2</span>, color <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-16-1.png" width="700">
We can see that the three population do separate nicely on the PCA, with
just one individual from Wam sitting in-between the other Wam
individuals and those from Eam.</p>
<p>It is also possible to inspect which loci contribute the most to a
given component:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>By using information from the loci table, we could easily embellish
the plot, for example colouring by chromosome or maf.</p>
<p>For more complex plots, we can augment the loci table with the
loadings using <code><a href="../reference/augment_loci.html">augment_loci()</a></code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/augment_loci.html">augment_loci</a></span><span class="op">(</span><span class="va">anole_pca</span>, data<span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="explore-population-structure-with-dapc">Explore population structure with DAPC<a class="anchor" aria-label="anchor" href="#explore-population-structure-with-dapc"></a>
</h2>
<p>DAPC is a powerful tool to investigate population structure. It has
the advantage of scaling well to very large datasets. It does not have
the assumptions of STRUCTURE or ADMIXTURE (which also limits its
power).</p>
<p>The first step is to determine the number of genetic clusters in the
dataset. DAPC can be either used to test a a-priori hypothesis, or we
can use the data to suggest the number of clusters. In this case, we did
not have any strong expectations of structure in our study system, so we
will let the data inform the number of possible genetic clusters. We
will use a k-clustering algorithm applied to the principal components
(allowing us to reduce the dimensions from the thousands of loci to just
a few tens of components). We need to decide how many components to use;
this decision is often made based on a plot of the cumulative explained
variance of the components. Using <code>tidy</code> on the
<code>gt_pca</code> object allows us easily obtain those quantities, and
it is then trivial to plot them:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>,matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC</span>, y<span class="op">=</span><span class="va">cumulative</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Note that, as we were working with a truncated SVD algorithm for our
PCA, we can not easily phrase the eigenvalues in terms of proportion of
total variance, so the cumulative y axis simply shows the cumulative sum
of the eigen values. Ideally, we are looking for the point where the
curve starts flattening. In this case, we can not see a very clear
flattening, but by PC 10 the increase in explained variance has markedly
decelerated. We can now find clusters based on those 10 PCs:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca.html">gt_cluster_pca</a></span><span class="op">(</span><span class="va">anole_pca</span>, n_pca <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>As we did not define the <em>k</em> values to explore, the default 1
to 5 was used (we can change that by setting the <code>k</code>
parameter to change the range). To choose an appropriate <em>k</em>, we
plot the number of clusters against a measure of fit. BIC has been shown
to be a very good metric under many scenarios:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>We are looking for the minimum value of BIC. There is no clear elbow
(a minimum after which BIC increases with increasing k). However, we
notice that there is a quick levelling off in the decrease in BIC at 3
clusters. Arguably, these are sufficient to capture the main structure
(and that makes sense given what we saw in the PCA). We can also use a
number of algorithmic approaches (based on the original
<code><a href="https://rdrr.io/pkg/adegenet/man/find.clusters.html" class="external-link">find.clusters()</a></code> function in <code>adegenet</code>) to
choose the best <em>k</em> value from this plot through
<code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>. We will use the defaults (BIC with
“diffNgroup”, see the help page for <code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>
for a description of the various options):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using BIC with criterion diffNgroup: 3 clusters</span></span></code></pre></div>
<p>The algorithm confirms our choice. Note that this function simply
adds an element <code>$best_k</code> to the <code>gt_cluster_pca</code>
object:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span><span class="op">$</span><span class="va">best_k</span></span>
<span><span class="co">#&gt; [1] 3</span></span></code></pre></div>
<p>If we decided that we wanted to explore a different value, we could
simply overwrite that number with
<code>anole_clusters$best_k&lt;-5</code></p>
<p>In this case, we are happy with the option of 3 clusters, and we can
run a DAPC:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_dapc.html">gt_dapc</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code><a href="../reference/gt_dapc.html">gt_dapc()</a></code> takes automatically the number of
clusters form the <code>anole_clusters</code> object, but can change
that behaviour by setting some of its parameters (see the help page for
<code><a href="../reference/gt_dapc.html">gt_dapc()</a></code>). When we print the object, we are given
information about the most important elements of the object and where to
find them (as we saw for <code>gt_pca</code>):</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span></span>
<span><span class="co">#&gt;  === DAPC of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Call ($call):gt_dapc(x = anole_clusters)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($eig):</span></span>
<span><span class="co">#&gt;  727.414 218.045 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; LD scores ($ind.coord):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by PC ($loadings):</span></span>
<span><span class="co">#&gt;  matrix with 2 rows (PC axes) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by locus($var.load):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (loci) and 2 columns (LD axes)</span></span></code></pre></div>
<p>Again, these elements can be obtained with tidiers (with
<code>matrix</code> equal to <code>eigenvalues</code>,
<code>scores</code>,<code>ld_loadings</code> and
<code>loci_loadings</code>):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;      LD eigenvalue cumulative</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1       727.       727.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2       218.       945.</span></span></code></pre></div>
<p>And they can be visualised with <code>autoplot</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>As for pca, there is a <code>tidy</code> method that can be used to
extract information from <code>gt_dapc</code> objects. For example, if
we want to create a bar plot of the eigenvalues (since we only have
two), we could simply use:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">LD</span>,y<span class="op">=</span><span class="va">eigenvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>We can plot the scores with:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/plot_scatter_dapc-1.png" width="700"></p>
<p>We can inspect the assignment by DAPC with <code>autoplot</code>
using the type <code>components</code>, ordering the samples by the
original population labels:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"components"</span>, group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>Because of the very clear separation we observed when plotting the LD
scores, no individual is modelled as a mixture: all assignments are with
100% probability to a single cluster.</p>
<p>Finally, we can explore which loci have the biggest impact on
separating the clusters (either because of drift or selection):</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<p>There is no strong outlier, suggesting drift across many loci has
created the signal picked up by DAPC.</p>
<p>Note that <code>anole_dapc</code> is of class <code>gt_dapc</code>,
which is a subclass of <code>dapc</code> from <code>adegenet</code>.
This means that functions written to work on <code>dapc</code> objects
should work out of the box (the only exception is
<code><a href="https://rdrr.io/pkg/adegenet/man/dapc.html" class="external-link">adegenet::predict.dapc</a></code>, which does not work because the
underlying pca object is different). For example, we can obtain the
standard dapc plot with:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ade4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    /// adegenet 2.1.10 is loaded ////////////</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    &gt; overview: '?adegenet'</span></span>
<span><span class="co">#&gt;    &gt; tutorials/doc/questions: 'adegenetWeb()' </span></span>
<span><span class="co">#&gt;    &gt; bug reports/feature requests: adegenetIssues()</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ade4/man/scatter.html" class="external-link">scatter</a></span><span class="op">(</span><span class="va">anole_dapc</span>, posi.da<span class="op">=</span><span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clustering-with-snmf">Clustering with sNMF<a class="anchor" aria-label="anchor" href="#clustering-with-snmf"></a>
</h2>
<p>sNMF is a fast clustering algorithm which provides results similar to
STRUCTURE and ADMIXTURE. We can run it directly form R. We have first to
generate a file with our genotype data:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geno_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_as_geno_lea.html">gt_as_geno_lea</a></span><span class="op">(</span><span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="va">geno_file</span></span>
<span><span class="co">#&gt; [1] "/tmp/Rtmp3aPpdt/anolis_1ced5b914ab2.geno"</span></span></code></pre></div>
<p>Note that the .geno file is placed by default in the same directory
and using the same name as the backing file of the
<code>gen_tibble</code></p>
<p>Now can run K clusters from k =1 to k=10. We will use just one
repeat, but ideally we should run multiple repetition for each K:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://membres-timc.imag.fr/Olivier.Francois/lea.html" class="external-link">LEA</a></span><span class="op">)</span></span>
<span><span class="va">anole_snmf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/LEA/man/main_sNMF.html" class="external-link">snmf</a></span><span class="op">(</span>input.file <span class="op">=</span> <span class="va">geno_file</span>,</span>
<span>                   K <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                   entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   repetitions <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   alpha <span class="op">=</span> <span class="fl">100</span></span>
<span>                  <span class="op">)</span></span></code></pre></div>
<p>We can examine the suitability of our K values by plotting with:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anole_snmf</span>, cex <span class="op">=</span> <span class="fl">1.2</span>, pch <span class="op">=</span> <span class="fl">19</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>From which we can see that K = 3 is a sensible choice, as 3
represents the ‘elbow’ in the plot. We can select the fastest run
using:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ce</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/LEA/man/crossEntropy.html" class="external-link">cross.entropy</a></span><span class="op">(</span><span class="va">anole_snmf</span>, K <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">quick_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">ce</span><span class="op">)</span></span></code></pre></div>
<p>And then to plot our clusters, we can begin by extracting the Q
matrix from our <code>snmf</code> object, and covert it to a
<code>q_matrix</code> object as used by <code>tidypopgen</code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_mat</span> <span class="op">&lt;-</span> <span class="fu">LEA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/LEA/man/Q.html" class="external-link">Q</a></span><span class="op">(</span><span class="va">anole_snmf</span>, K <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="va">quick_run</span><span class="op">)</span> </span>
<span><span class="va">q_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_q_matrix.html">as_q_matrix</a></span><span class="op">(</span><span class="va">q_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">q_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt;            .Q1       .Q2       .Q3</span></span>
<span><span class="co">#&gt; [1,] 9.999e-05 9.999e-05 9.998e-01</span></span>
<span><span class="co">#&gt; [2,] 9.999e-05 9.999e-05 9.998e-01</span></span>
<span><span class="co">#&gt; [3,] 9.998e-01 9.999e-05 9.999e-05</span></span>
<span><span class="co">#&gt; [4,] 9.998e-01 9.999e-05 9.999e-05</span></span>
<span><span class="co">#&gt; [5,] 9.998e-01 9.999e-05 9.999e-05</span></span>
<span><span class="co">#&gt; [6,] 9.999e-05 9.998e-01 9.999e-05</span></span></code></pre></div>
<p>Then we can tidy our q matrix, and attach the population and id data
from our original gen_tibble:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">q_mat</span>, <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tidy_q</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   id    group q     percentage</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;q_matrix&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> BM288 Eam   .Q1   9.999e-05 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> BM288 Eam   .Q2   9.999e-05 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> BM288 Eam   .Q3   9.998e-01 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GN71  Eam   .Q1   9.999e-05 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GN71  Eam   .Q2   9.999e-05 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> GN71  Eam   .Q3   9.998e-01</span></span></code></pre></div>
<p>And we can plot the data using ggplot</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set up the ggplot object</span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_q</span>,</span>
<span>                       <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">id</span>,</span>
<span>                                    y <span class="op">=</span> <span class="va">percentage</span>,</span>
<span>                                    fill <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># add the columns based on percentage membership to each cluster</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>reverse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="co"># set the y label</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"K = 4"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="co"># use a theme to match the distruct look, removing most decorations</span></span>
<span>  <span class="fu"><a href="../reference/theme_distruct.html">theme_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># set the colour scale to be the same as in distruct and clumppling</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span></span>
<span><span class="co">#&gt; Don't know how to automatically pick scale for object of type</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">&lt;q_matrix/matrix/array&gt;</span>. Defaulting to continuous.</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p>To clean up the plot, we can order our individuals based on their
population of origin and proportion derived from the q matrix:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># @TODO we should do this as part of tidying</span></span>
<span><span class="va">dominant_q</span> <span class="op">&lt;-</span> <span class="va">tidy_q</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>dominant_pop <span class="op">=</span> <span class="va">group</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">percentage</span><span class="op">)</span><span class="op">]</span>, dominant_q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">percentage</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">'drop'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="va">tidy_q</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">dominant_q</span>, by <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="va">tidy_q</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">group</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">dominant_q</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>plot_order <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span>,  <span class="co"># Create plot_order column</span></span>
<span>         id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">id</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">group</span>, <span class="op">-</span><span class="va">dominant_q</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>And replot it</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set up the ggplot object</span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_q</span>,</span>
<span>                       <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">id</span>,</span>
<span>                                    y <span class="op">=</span> <span class="va">percentage</span>,</span>
<span>                                    fill <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># add the columns based on percentage membership to each cluster</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>reverse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="co"># set the y label</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"K = 4"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="co"># use a theme to match the distruct look, removing most decorations</span></span>
<span>  <span class="fu"><a href="../reference/theme_distruct.html">theme_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># set the colour scale to be the same as in distruct and clumppling</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span></span>
<span><span class="co">#&gt; Don't know how to automatically pick scale for object of type</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">&lt;q_matrix/matrix/array&gt;</span>. Defaulting to continuous.</span></span></code></pre></div>
<p><img src="03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-40-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
