<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>overview • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="overview">
<meta property="og:description" content="tidypopgen">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9014</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_overview.html">overview</a>
    </li>
    <li>
      <a href="../articles/a02_qc.html">QC</a>
    </li>
    <li>
      <a href="../articles/a03_example_clustering_and_dapc.html">example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/a99_plink_cheatsheet.html">plink cheatsheet</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>overview</h1>
            
      
      
      <div class="hidden name"><code>a01_overview.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="tidy-data-in-population-genetics">tidy data in population genetics<a class="anchor" aria-label="anchor" href="#tidy-data-in-population-genetics"></a>
</h3>
<p>The fundamental tenet of tidy data that is that each observation
should have its own row and each variable its own column, such that each
value has its own cell. Applying this logic to population genetic data
means that each individual should have its own row, with individual
metadata (such as its population, sex, phenotype, etc) as the variables.
Genotypes for each locus can also be thought of as variables, however,
due to the large number of loci and the restricted values that each
genotype can take, it would be very inefficient to store them as
individual standard columns.</p>
<p>In <code>tidypopgen</code>, we represent data as a
<code>gen_tbl</code>, a subclass of <code>tibble</code> which has three
compulsory columns: <code>id</code> of the individual (as a
<code>character</code>, which must be unique for each individual),
<code>population</code> giving the population the individuals belong to
(a <code>character</code>), and <code>genotypes</code> (stored in a
compressed format as a File-Backed Matrix, with the vector in the tibble
providing the row indices of the matrix for each individual). The real
data reside on disk, and an attribute <code>bigsnp</code> of the
<code>genotype</code> column contains all the information to access it.
There is also an additional attribute , <code>loci</code> which provides
all the information about the loci, including the column indices that
represent each locus in the FBM. The vector of row indices and the table
of loci can be subsetted and reordered without changing the data on
disk; thus, any operation on the <code>gen_tibble</code> is fast as it
shapes the indices of the genotype matrix rather than the matrix
itself.<br>
The <code>loci</code> tibble includes columns <code>big_index</code> for
the index in the FBM, <code>name</code> for the locus name (a
<code>character</code>, which must be unique), <code>chromosome</code>
for the chromosome (an <code>integer</code>, if known, otherwise set to
<code>NA</code>), <code>position</code> for the position on the
chromosome (an <code>integer</code>, if known, otherwise set to
<code>NA</code>), <code>genetic_dist</code> for the genetic distance on
the chromosome (<code>numeric</code>, if known, else set to 0)
<code>allele_ref</code> for the the reference allele (a
<code>character</code>), and <code>allele_alt</code> for the alternate
allele (a <code>character</code>, which can be <code>0</code> for
monomorphic loci, following the same convention as plink). Additional
individual metadata can be stored as columns in the main
<code>gen_tbl</code>, whilst additional loci information (such as the
position in centimorgans) can be added as columns in the
<code>loci</code> attribute table.</p>
<p>In principle, it is possible to use use multiple ways to compress the
genotypes. <code>tidypopgen</code> currently uses a <code>bigSNP</code>
object from the package <code>bigsnpr</code>. It is very fast and well
documented, but it is mostly geared towards diploid data.
<code>tidypopgen</code> expands that object to deal with different
levels of ploidy, including multiple ploidy within a single dataset;
however, most functions are currently incompatible with ploidy levels
other than 2 (but they will return a clear error message and avoid
computing anything incorrectly).</p>
</div>
<div class="section level3">
<h3 id="the-grammar-of-population-genetics">The grammar of population genetics<a class="anchor" aria-label="anchor" href="#the-grammar-of-population-genetics"></a>
</h3>
<div class="section level4">
<h4 id="the-gen_tibble">The <code>gen_tibble</code><a class="anchor" aria-label="anchor" href="#the-gen_tibble"></a>
</h4>
<p>Given information about the individuals, their genotypes, and the
loci:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidypopgen</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: tibble</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_indiv_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span> <span class="op">(</span>id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>,<span class="st">"b"</span>,<span class="st">"c"</span>,<span class="st">"d"</span>,<span class="st">"e"</span><span class="op">)</span>,</span>
<span>                       population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pop1"</span>,<span class="st">"pop1"</span>,<span class="st">"pop2"</span>,<span class="st">"pop2"</span>,<span class="st">"pop2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">example_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,<span class="cn">NA</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>,<span class="cn">NA</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">example_loci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs1"</span>,<span class="st">"rs2"</span>,<span class="st">"rs3"</span>,<span class="st">"rs4"</span>,<span class="st">"x1"</span>,<span class="st">"x2"</span><span class="op">)</span>,</span>
<span>                           chromosome<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                           position<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">65</span>,<span class="fl">343</span>,<span class="fl">23</span>,<span class="fl">456</span><span class="op">)</span>,</span>
<span>                           genetic_dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span>,</span>
<span>                           allele_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"T"</span>,<span class="st">"C"</span>,<span class="st">"G"</span>,<span class="st">"C"</span>,<span class="st">"T"</span><span class="op">)</span>,</span>
<span>                           allele_alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T"</span>,<span class="st">"C"</span>, <span class="cn">NA</span>,<span class="st">"C"</span>,<span class="st">"G"</span>,<span class="st">"A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can create a simple <code>gen_tibble</code> object (of class
<code>gen_tbl</code>) with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">example_genotypes</span>, </span>
<span>                         indiv_meta <span class="op">=</span> <span class="va">example_indiv_meta</span>,</span>
<span>                         loci<span class="op">=</span><span class="va">example_loci</span>,</span>
<span>                         backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/file1ca4799d91a6.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/file1ca4799d91a6.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/file1ca4799d91a6.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/file1ca4799d91a6.gt')</span></span></code></pre></div>
<p>We are provided information on where the three files underlying the
genotype information are stored. As we don’t want to keep the files, we
used the tmp directory; normally you will want to use your working
directory so that the files will not be cleared by R at the end of the
session. It is important that these files are not deleted or moved, as
<code>gen_tibble</code> stores their paths for future use.</p>
<p>Now let’s have a look at our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 6 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 3</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> a     pop1                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> b     pop1                2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> c     pop2                3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d     pop2                4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> e     pop2                5</span></span></code></pre></div>
<p>As discussed above, in this tibble, <code>genotypes</code> contains
the indices of the individuals in the FBM as values, and the FBM as an
attribute.</p>
<p>To retrieve the genotypes (which are compressed in the FBM), we
use:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_genotypes.html">show_genotypes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6]</span></span>
<span><span class="co">#&gt; [1,]    1    1    0    1    1    0</span></span>
<span><span class="co">#&gt; [2,]    2    0    0    0   NA    0</span></span>
<span><span class="co">#&gt; [3,]    1    2    0    0    1    1</span></span>
<span><span class="co">#&gt; [4,]    0    2    0    1    2    1</span></span>
<span><span class="co">#&gt; [5,]    1    1   NA    2    1    0</span></span></code></pre></div>
<p>If we want to extract the information about the loci for which we
have genotypes (which are stored as an attribute of that column), we
say:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name  chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         1 rs1            1        3            0 A          T         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>         2 rs2            1        5            0 T          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>         3 rs3            1       65            0 C          <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>         4 rs4            1      343            0 G          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>         5 x1             2       23            0 C          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>         6 x2             2      456            0 T          A</span></span></code></pre></div>
<p>Note that, if we are passing a <code>gen_tibble</code> to a function
that works on genotypes, it is generally not necessary to pass the
column <code>genotypes</code> in the call:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6666667 0.0000000 0.5000000 0.3333333 0.6000000</span></span></code></pre></div>
<p>However, if such a function is used within a <code>dplyr</code> verb
such as <code>mutate</code>, we need to pass the <code>genotype</code>
column to the function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span> <span class="op">(</span>het_obs <span class="op">=</span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 6 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 4</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes het_obs</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> a     pop1                1   0.667</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> b     pop1                2   0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> c     pop2                3   0.5  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d     pop2                4   0.333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> e     pop2                5   0.6</span></span></code></pre></div>
<p>Or, more simply:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span> <span class="op">(</span>het_obs <span class="op">=</span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 6 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 4</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes het_obs</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> a     pop1                1   0.667</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> b     pop1                2   0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> c     pop2                3   0.5  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d     pop2                4   0.333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> e     pop2                5   0.6</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="standard-dplyr-verbs-to-manipulate-the-tibble">Standard <code>dplyr</code> verbs to manipulate the tibble<a class="anchor" aria-label="anchor" href="#standard-dplyr-verbs-to-manipulate-the-tibble"></a>
</h3>
<p>The individual metadata can then be processed with the usual
<code>tidyverse</code> grammar. So, we can filter individuals by
population with</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_pop1</span> <span class="op">&lt;-</span> <span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span> <span class="op">(</span><span class="va">population</span><span class="op">==</span><span class="st">"pop2"</span><span class="op">)</span></span>
<span><span class="va">example_pop1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 6 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     3 × 3</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> c     pop2                3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> d     pop2                4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> e     pop2                5</span></span></code></pre></div>
<p>There are a number of functions that compute population genetics
quantities for each individual, such as individual observed
heterozygosity. We can compute them simply with:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6666667 0.0000000 0.5000000 0.3333333 0.6000000</span></span></code></pre></div>
<p>Or after filtering:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span> <span class="op">(</span><span class="va">population</span><span class="op">==</span><span class="st">"pop2"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5000000 0.3333333 0.6000000</span></span></code></pre></div>
<p>We can use <code>mutate</code> to add observed heterozygosity as a
column to our <code>gen_tibble</code> (again, note that functions that
work on genotypes don’t need to be passed any arguments if the tibble is
passed directly to them, but the column <code>genotypes</code> has to be
provided when they are used within <code>dplyr</code> verbs such as
<code>mutate</code>):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>het_obs <span class="op">=</span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 6 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 4</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes het_obs</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> a     pop1                1   0.667</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> b     pop1                2   0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> c     pop2                3   0.5  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d     pop2                4   0.333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> e     pop2                5   0.6</span></span></code></pre></div>
<p>There are a number of functions that estimate quantities at the
<em>individual</em> level, and they are prefixed by
<code>indiv_</code>.</p>
</div>
<div class="section level2">
<h2 id="using-verbs-on-loci">Using verbs on loci<a class="anchor" aria-label="anchor" href="#using-verbs-on-loci"></a>
</h2>
<p>Since the genotypes of the loci are stored as a compressed list in
one column, it is not possible to use standard <code>dplyr</code> verbs
on them. However, <code>tidypopgen</code> provides a number of
specialised verbs, postfixed by <code>_loci</code>, to manipulate
loci.</p>
<p>A key operation on loci is their selection (and removal). The
compressed nature of genotypes imposes some constraints on the possible
grammar. For selection, there are two verbs: <code>select_loci</code>
and <code>select_loci_if</code>. <code>select_loci</code> understands
the concise-minilanguage spoken by standard <code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">dplyr::select</a></code>
that allows to easily refer to variables by their names. However,
<code>select_loci</code> criteria can not be based on the actual
genotypes (e.g. on heterozygosity or missingness). For that, we have to
use <code>select_loci_if</code>, which can operate on the genotypes but
is blind to the names of loci.</p>
<p>Let us start by looking at the loci names in our simple dataset:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/loci_names.html">loci_names</a></span><span class="op">(</span><span class="va">example_gt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "rs1" "rs2" "rs3" "rs4" "x1"  "x2"</span></span></code></pre></div>
<p>We can see that there are two categories of loci, one starting with
“rs” and the other with “x”. If we wanted to select only loci that have
an “rs” code, we would use:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sub</span> <span class="op">&lt;-</span> <span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci.html">select_loci</a></span> <span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"rs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">example_sub</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 4 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 3</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> a     pop1                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> b     pop1                2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> c     pop2                3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d     pop2                4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> e     pop2                5</span></span></code></pre></div>
<p>This gives us a <code>gen_tibble</code> with only 4 loci, as
expected. We can confirm that we have the correct loci with:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/loci_names.html">loci_names</a></span><span class="op">(</span><span class="va">example_sub</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "rs1" "rs2" "rs3" "rs4"</span></span></code></pre></div>
<p>Let us check that this has indeed impacted the individual
heterozygosity</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sub</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5000000 0.0000000 0.1666667 0.1666667 0.4000000</span></span></code></pre></div>
<p>We can also subset and reorder by passing indices:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci.html">select_loci</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">6</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name  chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         2 rs2            1        5            0 T          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>         6 x2             2      456            0 T          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>         1 rs1            1        3            0 A          T</span></span></code></pre></div>
<p>This operation could be helpful when merging datasets that do not
fully overlap on their loci (more on that later).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci.html">select_loci</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">6</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_genotypes.html">show_genotypes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3]</span></span>
<span><span class="co">#&gt; [1,]    1    0    1</span></span>
<span><span class="co">#&gt; [2,]    0    0    2</span></span>
<span><span class="co">#&gt; [3,]    2    1    1</span></span>
<span><span class="co">#&gt; [4,]    2    1    0</span></span>
<span><span class="co">#&gt; [5,]    1    0    1</span></span></code></pre></div>
<p>The limit of <code>select_loci</code> is that it can not directly
summarise the genotypes. We can do that separately and then feed the
result as a set of indices. For example, we might want to impose a
minimum minor allele frequency. <code><a href="../reference/loci_alt_freq.html">loci_maf()</a></code> allows us to
inspect the minimum allele frequencies in a <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.500 0.400 0.000 0.400 0.375 0.200</span></span></code></pre></div>
<p>We can now create a vector of indices of loci with a minimum allele
frequency (MAF) larger than 0.2, and use it to select:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci.html">select_loci</a></span> <span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">sel_indices</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name  chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         1 rs1            1        3            0 A          T         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>         2 rs2            1        5            0 T          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>         4 rs4            1      343            0 G          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>         5 x1             2       23            0 C          G</span></span></code></pre></div>
<p>Note that passing a variable directly to <code>select</code> is
deprecated, and so we have to use <code>all_of</code> to wrap it.</p>
<p><code>select_loci_if</code> allows us to avoid creating a temporary
variable to store indices:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt_sub</span> <span class="op">&lt;-</span> <span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">example_gt_sub</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_genotypes.html">show_genotypes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">#&gt; [1,]    1    1    1    1</span></span>
<span><span class="co">#&gt; [2,]    2    0    0   NA</span></span>
<span><span class="co">#&gt; [3,]    1    2    0    1</span></span>
<span><span class="co">#&gt; [4,]    0    2    1    2</span></span>
<span><span class="co">#&gt; [5,]    1    1    2    1</span></span></code></pre></div>
<p>Note that, as we need to tidy evaluate <code>loci_maf</code> within
the <code>select_loci_if</code> verb, we need to provide it with the
column that we want to use (even though it has to be
<code>genotypes</code>). Also note that, with
<code>select_loci_if</code>, we can not reorder the loci.</p>
<p><code>select_loci_if</code> is very flexible; for example, we could
filter loci with a MAF greater than 0.2 that are also on chromosome
2.</p>
<p>We can use a similar approach to select only alleles on a given
chromosome:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_chromosomes.html">loci_chromosomes</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">==</span><span class="fl">2</span> <span class="op">&amp;</span></span>
<span>                             <span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name  chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         5 x1             2       23            0 C          G</span></span></code></pre></div>
<p>Incidentally, <code><a href="../reference/loci_alt_freq.html">loci_maf()</a></code> is one of several functions
that compute quantities by locus; they can be identified as they start
with <code>loci_</code>.</p>
</div>
<div class="section level2">
<h2 id="grouping-individuals-in-populations">Grouping individuals in populations<a class="anchor" aria-label="anchor" href="#grouping-individuals-in-populations"></a>
</h2>
<p>In population genetics, we are generally interested in computing
quantities that describe groups of individuals (i.e. populations).
Grouping can be used in a number of ways.</p>
<p>As a starting point, we can group by population and get pop
sizes:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">tally</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   population     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop1           2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> pop2           3</span></span></code></pre></div>
<p>For functions that return one result per individual (such as
<code>indiv_het_obs</code> that we used before), we can use
<code>summarise</code>, which returns a new <code>tibble</code> with one
line per population. For example, we can count the number of individuals
per population, as well as their mean heterozygosity with :</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>n<span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, mean_het <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   population     n mean_het</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop1           2    0.333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> pop2           3    0.478</span></span></code></pre></div>
<p>However, note that this is somewhat inefficient, as computing the pop
averages requires multiple access to the data. A more efficient approach
is to:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>het_obs <span class="op">=</span> <span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>n<span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, mean_het <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">het_obs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   population     n mean_het</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop1           2    0.333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> pop2           3    0.478</span></span></code></pre></div>
<p>In this way, we compute all individual heterozygosities in one go
(optimising our file access time), and then generate the population
summaries.</p>
<p>For functions that return a quantity per locus
(e.g. <code>loci_maf</code>), we can use <code>group_by</code> together
with <code>group_map</code> to apply such functions over
populations:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_map.html" class="external-link">group_map</a></span><span class="op">(</span>.f <span class="op">=</span> <span class="op">~</span><span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0.25 0.25 0.00 0.25 0.50 0.00</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 0.3333333 0.1666667 0.0000000 0.5000000 0.3333333 0.3333333</span></span></code></pre></div>
<p>For more details on the syntax of <code>group_map</code>, see its
help page. Some functions, such as <code><a href="../reference/loci_alt_freq.html">loci_maf()</a></code>, also have a
method for grouped tibbles that allows an even easier syntax:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0.25 0.25 0.00 0.25 0.50 0.00</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 0.3333333 0.1666667 0.0000000 0.5000000 0.3333333 0.3333333</span></span></code></pre></div>
<p>In reality, such grouped methods are nothing more than a wrapper for
the the appropriate <code>group_map</code> syntax, but they do make life
a little bit simpler.</p>
<p>Certain metrics and analyses are naturally defined by a grouped
tibble as they refer to populations, such as distance metrics among
populations. For example:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># not reimplemented yet!</span></span>
<span><span class="va">example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/pairwise_pop_fst.html">pairwise_pop_fst</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Whilst this function should work, it has not been extensively tested. Check your results to ensure they make sense</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   population_1 population_2  value</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop1         pop2         0.025<span style="text-decoration: underline;">8</span></span></span></code></pre></div>
<p>These type of functions are prefixed with <code>pop</code>.</p>
</div>
<div class="section level2">
<h2 id="saving-and-reading-data">Saving and reading data<a class="anchor" aria-label="anchor" href="#saving-and-reading-data"></a>
</h2>
<p>We can save a <code>gen_tibble</code> with <code><a href="../reference/gt_save.html">gt_save()</a></code>.
This command will save a file with extension <code>.gt</code>. Together
with the <code>.rds</code> and <code>.bk</code> files, the
<code>.gt</code> file include all the information stored in the
<code>gen_tibble</code>. Note that, whilst the <code>.rds</code> and
<code>.bk</code> file have to share name, the <code>.gt</code> file can
be named differently (but, by default, if no specific name is given,
<code>gt_save</code> will use the same pattern as for the
<code>.rds</code> and <code>.bk</code> file).</p>
<p>So, let us save our file:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gt_file_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_save.html">gt_save</a></span><span class="op">(</span><span class="va">example_gt</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/file1ca4799d91a6.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/file1ca4799d91a6.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/file1ca4799d91a6.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/file1ca4799d91a6.gt')</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gt_file_name</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpUyZqki/file1ca4799d91a6.gt" </span></span>
<span><span class="co">#&gt; [2] "/tmp/RtmpUyZqki/file1ca4799d91a6.rds"</span></span>
<span><span class="co">#&gt; [3] "/tmp/RtmpUyZqki/file1ca4799d91a6.bk"</span></span></code></pre></div>
<p>In a later session, we could reload the data with:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_example_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_load.html">gt_load</a></span><span class="op">(</span><span class="va">gt_file_name</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">new_example_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_genotypes.html">show_genotypes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6]</span></span>
<span><span class="co">#&gt; [1,]    1    1    0    1    1    0</span></span>
<span><span class="co">#&gt; [2,]    2    0    0    0   NA    0</span></span>
<span><span class="co">#&gt; [3,]    1    2    0    0    1    1</span></span>
<span><span class="co">#&gt; [4,]    0    2    0    1    2    1</span></span>
<span><span class="co">#&gt; [5,]    1    1   NA    2    1    0</span></span></code></pre></div>
<p>We can see that our genotypes were recovered correctly.</p>
<p>As we saw at the beginning of this vignette, it is possible to create
a <code>gen_tibble</code> with data in data.frames and tibbles. We can
use that function to wrangle small datasets in custom formats, but more
commonly SNP data are stored as PLINK bed files or VCF files.
<code>gen_tibble</code> can directly read both types of files (including
gzipped vcf files), we just need to provide the path to file as the
first argument of <code>gen_tibble</code>; for example, if we want to
read a PLINK bed file, we can simply use:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_path_pop_a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/pop_a.bed"</span>, package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">pop_a_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">bed_path_pop_a</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"pop_a_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/pop_a_1ca46bdb3bc.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/pop_a_1ca46bdb3bc.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/pop_a_1ca46bdb3bc.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/pop_a_1ca46bdb3bc.gt')</span></span></code></pre></div>
<p>For this vignette, we don’t want to keep files, so we are using again
a temporary path for the backing files, but in normal instances, we can
simply omit the <code>backingfile</code> parameter, and the
<code>.rds</code> and <code>.bk</code> file will be saved with the same
name and path as the original <code>.bed</code> file.</p>
<p>We can also export data into various formats with the family of
functions <code>gt_as_*()</code>. Some functions, such as
<code><a href="../reference/gt_as_hierfstat.html">gt_as_hierfstat()</a></code>, <code><a href="../reference/gt_as_genind.html">gt_as_genind()</a></code> or
<code><a href="../reference/gt_as_genlight.html">gt_as_genlight()</a></code> return an object of the appropriate type;
other functions, such as <code><a href="../reference/gt_as_plink.html">gt_as_plink()</a></code> or
<code><a href="../reference/gt_as_geno_lea.html">gt_as_geno_lea()</a></code> write a file in the appropriate format,
and return the name of that file on completion. For example, to export
to a PLINK .bed file, we simply use:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_as_plink.html">gt_as_plink</a></span><span class="op">(</span><span class="va">example_gt</span>, file <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"new_bed_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpUyZqki/new_bed_1ca45aeb0bed.bed"</span></span></code></pre></div>
<p>This will also write a .bim and .fam file and save them together with
the .bed file. Note that, from the main tibble, only <code>id</code>,
<code>population</code> and <code>sex</code> will be preserved in the
.fam file. It is also possible to write .ped and .raw files by
specifying <code>type="ped"</code> or <code>type="raw"</code> in
<code><a href="../reference/gt_as_plink.html">gt_as_plink()</a></code> (see the help page for
<code><a href="../reference/gt_as_plink.html">gt_as_plink()</a></code> for details).</p>
</div>
<div class="section level2">
<h2 id="merging-data">Merging data<a class="anchor" aria-label="anchor" href="#merging-data"></a>
</h2>
<p>Merging data from different sources is a common problem, especially
in human population genetics where there is a wealth of SNP chips
available. In <code>tidypopgen</code>, merging is enacted with an
<code>rbind</code> operation between <code>gen_tibbles</code>. If the
datasets have the same loci, then the merge is trivial. If not, then it
is necessary to subset to the same loci, and ensure that the data are
coded with the same reference and alternate alleles (or swap if needed).
Additionally, if data come from SNP chips, there is the added
complication that the strand is not always consistent, so it might also
be necessary to flip strand (in that case, ambiguous SNPs have to be
filtered). The <code>rbind</code> method for <code>gen_tibbles</code>
has a number of parameters that allow us to control the behaviour of the
merge.</p>
<p>Let us start by bringing in two sample datasets (note that we use
tempfiles to store the data; in real applications, we will usually avoid
defining a backingfile and let the function create backing files where
the bed file is stored):</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_path_pop_a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/pop_a.bed"</span>, package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">bigsnp_path_a</span> <span class="op">&lt;-</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bed_path_pop_a</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"pop_a_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pop_a_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">bigsnp_path_a</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/pop_a_1ca46dce358.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/pop_a_1ca46dce358.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/pop_a_1ca46dce358.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/pop_a_1ca46dce358.gt')</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_path_pop_b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/pop_b.bed"</span>, package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">bigsnp_path_b</span> <span class="op">&lt;-</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bed_path_pop_b</span>, backingfile <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"pop_b_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pop_b_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">bigsnp_path_b</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/pop_b_1ca4374291ee.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/pop_b_1ca4374291ee.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/pop_b_1ca4374291ee.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/pop_b_1ca4374291ee.gt')</span></span></code></pre></div>
<p>And inspect them:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop_a_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 16 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 3</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> GRC24 pop_a               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GRC25 pop_a               2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> GRC26 pop_a               3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GRC27 pop_a               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GRC28 pop_a               5</span></span></code></pre></div>
<p>And the other one:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop_b_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 17 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     3 × 3</span></span></span>
<span><span class="co">#&gt;   id     population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SL088  pop_b               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> SL1329 pop_b               2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> SL1108 pop_b               3</span></span></code></pre></div>
<p>Here we are using very small datasets, but in real life,
<code>rbind</code> operations are very demanding. Before performing such
an operation, we can run <code>rbind_dry_run</code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbind_dry_run.html">rbind_dry_run</a></span><span class="op">(</span><span class="va">pop_a_gt</span>, <span class="va">pop_b_gt</span>, flip_strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonising loci between two datasets</span></span>
<span><span class="co">#&gt; flip_strand =  TRUE  ; remove_ambiguous =  TRUE </span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: reference </span></span>
<span><span class="co">#&gt; number of SNPs: 16 reduced to 12 </span></span>
<span><span class="co">#&gt; ( 2 are ambiguous, of which 2  were removed)</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: target </span></span>
<span><span class="co">#&gt; number of SNPs: 17 reduced to 12 </span></span>
<span><span class="co">#&gt; ( 5 were flipped to match the reference set)</span></span>
<span><span class="co">#&gt; ( 2 are ambiguous, of which 2 were removed)</span></span></code></pre></div>
<p>Note that, by default, <code>rbind</code> will NOT flip strand or
remove ambiguous SNPs (as they are only relevant when merging different
SNP chips), you need to set <code>flip_strand</code> to TRUE.</p>
<p>The report object contains details about why each locus was either
kept or removed, but usually the report is sufficient to make decisions
on whether we want to go ahead. If we are happy with the likely outcome,
we can proceed with the <code>rbind</code>. Note that the data will be
saved to disk. We can either provide a path and prefix, to which ‘.RDS’
and ‘.bk’ will be appended for the <code>bigSNP</code> file and its
backing file; or let the function save the files in the same path as the
original backing file of the first object).</p>
<p>NOTE: In this vignette, we save to the temporary directory, but in
real life you want to save in a directory where you will be able to
retrieve the file at a later date!!!</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># #create merge</span></span>
<span><span class="va">merged_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">pop_a_gt</span>, <span class="va">pop_b_gt</span>, flip_strand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"gt_merged"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonising loci between two datasets</span></span>
<span><span class="co">#&gt; flip_strand =  TRUE  ; remove_ambiguous =  TRUE </span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: reference </span></span>
<span><span class="co">#&gt; number of SNPs: 16 reduced to 12 </span></span>
<span><span class="co">#&gt; ( 2 are ambiguous, of which 2  were removed)</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: target </span></span>
<span><span class="co">#&gt; number of SNPs: 17 reduced to 12 </span></span>
<span><span class="co">#&gt; ( 5 were flipped to match the reference set)</span></span>
<span><span class="co">#&gt; ( 2 are ambiguous, of which 2 were removed)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/gt_merged.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/gt_merged.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/gt_merged.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/gt_merged.gt')</span></span></code></pre></div>
<p>Let’s check the resulting <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 12 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     8 × 3</span></span></span>
<span><span class="co">#&gt;   id     population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> GRC24  pop_a               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GRC25  pop_a               2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> GRC26  pop_a               3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GRC27  pop_a               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GRC28  pop_a               5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> SL088  pop_b               6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> SL1329 pop_b               7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> SL1108 pop_b               8</span></span></code></pre></div>
<p>Note that the values in the genotype column (which corresponds to the
id in the FBM file) have changed to reflect that we have a new, larger
FBM with both datasets.</p>
<p>We can look at the subsetted loci (note that we used the first
population as reference to determine the strand and order of
alleles):</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 7</span></span></span>
<span><span class="co">#&gt;    big_index name       chromosome  position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>         1 rs3094315           1    <span style="text-decoration: underline;">752</span>566            0 A          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>         2 rs3131972           1    <span style="text-decoration: underline;">752</span>721            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>         3 rs12124819          1    <span style="text-decoration: underline;">776</span>546            0 A          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>         4 rs11240777          1    <span style="text-decoration: underline;">798</span>959            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>         5 rs1110052           1    <span style="text-decoration: underline;">873</span>558            0 G          T         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>         6 rs6657048           1    <span style="text-decoration: underline;">957</span>640            0 C          T         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>         7 rs2488991           1    <span style="text-decoration: underline;">994</span>391            0 T          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>         8 rs2862633           2  61<span style="text-decoration: underline;">974</span>443            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>         9 rs28569024          2 139<span style="text-decoration: underline;">008</span>811            0 T          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>        10 rs10106770          2 235<span style="text-decoration: underline;">832</span>763            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span>        11 rs11942835          3 155<span style="text-decoration: underline;">913</span>651            0 T          C         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span>        12 rs5945676          23  51<span style="text-decoration: underline;">433</span>071            0 T          G</span></span></code></pre></div>
<p>Again, note that the <code>big_index</code> values have changed
compared to the original files, as we generated a new FBM with the
merged data.</p>
</div>
<div class="section level2">
<h2 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h2>
<p>Many genetic analysis (e.g. PCA) do not allow missing data. In many
software implementations, missing gentoypes are imputed on the fly,
often in a simplistic manner. In <code>tidypopgen</code>, we encourage
taking this step explicitly, before running any analysis.</p>
<p>Let us start with a dataset that has some missing genotypes:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example-missing.bed"</span>, package <span class="op">=</span> <span class="st">"bigsnpr"</span><span class="op">)</span></span>
<span><span class="va">missing_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">bed_file</span>,  backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"missing_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpUyZqki/missing_1ca43194da0f.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpUyZqki/missing_1ca43194da0f.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpUyZqki/missing_1ca43194da0f.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpUyZqki/missing_1ca43194da0f.gt')</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 500 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     200 × 3</span></span></span>
<span><span class="co">#&gt;    id     population  genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ind_1  fam_1               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ind_2  fam_2               2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ind_3  fam_3               3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ind_4  fam_4               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ind_5  fam_5               5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ind_6  fam_6               6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ind_7  fam_7               7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ind_8  fam_8               8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ind_9  fam_9               9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ind_10 fam_10             10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 190 more rows</span></span></span></code></pre></div>
<p>We can visualise the amount of missingness with:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a01_overview_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<p>If we attempt to run a PCA on this dataset, we get:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_pca</span> <span class="op">&lt;-</span> <span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_autoSVD.html">gt_pca_autoSVD</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: You can't have missing values in 'G'.</span></span></code></pre></div>
<p>It is possible to obviate to this problem by filtering loci with
missing data, but that might lose a lot of loci. The alternative is to
impute the missing the data. <code>tidypopgen</code> provides wrappers
for two fast imputation approaches available in <code>bigsnpr</code>, a
simple imputation (<code>gt_impute_simple</code>) based on the frequency
of the alleles at each locus (by random sampling, or using the mean or
mode), and more sophisticated approach (<code>gt_impute_xgboost</code>)
that uses boosted trees to try and predict the most likely genotype.
These methods are fine to impute a few missing genotypes, but they
should not be used for any sophisticated imputation (e.g. of low
coverage genomes).</p>
<p>We use the simple approach to fix our dataset:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">missing_gt</span>, method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></code></pre></div>
<p>We can now check that our dataset has indeed been imputed:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_has_imputed.html">gt_has_imputed</a></span><span class="op">(</span><span class="va">missing_gt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>However, note that a <code>gen_tibble</code> stores both the raw data
and the imputed data. Even after imputation, the imputed data are not
used by default:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_uses_imputed.html">gt_uses_imputed</a></span><span class="op">(</span><span class="va">missing_gt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>And indeed, if we summarise missingness, we still get:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a01_overview_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>We can manually force a <code>gen_tibble</code> to use the imputed
data:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_set_imputed.html">gt_set_imputed</a></span><span class="op">(</span><span class="va">missing_gt</span>, set <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a01_overview_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<p>However, this is generally not needed, we can keep our
<code>gen_tibble</code> set to use the raw data:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_set_imputed.html">gt_set_imputed</a></span><span class="op">(</span><span class="va">missing_gt</span>, set <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a01_overview_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>And let functions that need imputation use it automatically:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_pca</span> <span class="op">&lt;-</span> <span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">missing_pca</span></span>
<span><span class="co">#&gt;  === PCA of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Method: [1] "partialSVD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call ($call):gt_pca_partialSVD(x = .)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($d):</span></span>
<span><span class="co">#&gt;  146.859 106.219 90.352 80.983 69.332 68.427 ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Principal component scores ($u):</span></span>
<span><span class="co">#&gt;  matrix with 200 rows (individuals) and 10 columns (axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings (Principal axes) ($v):</span></span>
<span><span class="co">#&gt;  matrix with 500 rows (SNPs) and 10 columns (axes)</span></span></code></pre></div>
<p>Note that, when the function is finished, the <code>gen_tibble</code>
is back to using the raw genotypes:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_uses_imputed.html">gt_uses_imputed</a></span><span class="op">(</span><span class="va">missing_gt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missing_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a01_overview_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>More details about PCA and other analysis is found in the vignette on
population genetic analysis.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
