<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quality Control • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quality Control">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
    <li><a class="dropdown-item" href="../articles/interpolating.html">Interpolating genetic data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Quality Control</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/vignettes/a02_qc.Rmd" class="external-link"><code>vignettes/a02_qc.Rmd</code></a></small>
      <div class="d-none name"><code>a02_qc.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quality-control-for-snp-datasets">Quality control for SNP datasets<a class="anchor" aria-label="anchor" href="#quality-control-for-snp-datasets"></a>
</h2>
<p>tidypopgen has two key functions to examine the quality of data
across loci and across individuals: <code>qc_report_loci</code> and
<code>qc_report_indiv</code>. This vignette uses a simulated data set to
illustrate these methods of data cleaning.</p>
</div>
<div class="section level2">
<h2 id="read-data-into-gen_tibble-format">Read data into gen_tibble format<a class="anchor" aria-label="anchor" href="#read-data-into-gen_tibble-format"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EvolEcolGroup/tidypopgen" class="external-link">tidypopgen</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: tibble</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/related/families.bed"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"tidypopgen"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>#Quality control for individuals</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">individual_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     het_obs        missingness     </span></span>
<span><span class="co">##  Min.   :0.3467   Min.   :0.03018  </span></span>
<span><span class="co">##  1st Qu.:0.3736   1st Qu.:0.03642  </span></span>
<span><span class="co">##  Median :0.3790   Median :0.03798  </span></span>
<span><span class="co">##  Mean   :0.3799   Mean   :0.03867  </span></span>
<span><span class="co">##  3rd Qu.:0.3901   3rd Qu.:0.04110  </span></span>
<span><span class="co">##  Max.   :0.4015   Max.   :0.04579</span></span></code></pre>
<p>The output of <code>qc_report_indiv</code> supplies observed
heterozygosity per individual, and rate of missingness per individual as
standard.</p>
<p>These data can also be visualised using autoplot:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-3-1.png" alt="Scatter plot of missingness proportion and observed heterozygosity for each individual" width="700"></p>
<p>Here, the red line indicates a threshold for proportion of missing
loci, which is set as 5% by default, and can be altered using the
<code>miss_threshold</code> argument. Similarly, the blue lines are
drawn at 2 (inner) and 3 (outer) standard deviations from the mean
observed heterozygosity. These thresholds can be used to visualise
outliers and consider how to filter datasets.</p>
<p>We can see above that most individuals have low missingness, none are
above the default 5% threshold. However, if we wanted to filter
individuals to remove those with more than 4.5% of their genotypes
missing, we can use <code>filter</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_missingness.html">indiv_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.045</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10</span></span></code></pre>
<p>And if we wanted to remove outliers with particularly high or low
heterozygosity, we can again do so by using <code>filter</code>. As an
example, here we remove observations that lie more than 2 standard
deviations from the mean.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_val</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">$</span><span class="va">het_obs</span><span class="op">)</span></span>
<span><span class="va">sd_val</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">$</span><span class="va">het_obs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lower</span> <span class="op">&lt;-</span> <span class="va">mean_val</span> <span class="op">-</span> <span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="va">sd_val</span><span class="op">)</span></span>
<span><span class="va">upper</span> <span class="op">&lt;-</span> <span class="va">mean_val</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="va">sd_val</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">lower</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">upper</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9</span></span></code></pre>
<p>Next, we can look at relatedness within our sample. If the parameter
<code>kings_threshold</code> is provided to
<code><a href="../reference/qc_report_indiv.html">qc_report_indiv()</a></code>, then the report also calculates a KING
coefficient of relatedness matrix using the sample. The
<code>kings_threshold</code> is used to provide an output of the largest
possible group with no related individuals in the third column
<code>to_keep</code>. This boolean column recommends which individuals
to remove (FALSE) and to keep (TRUE) to achieve an unrelated sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">individual_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="va">data</span>, kings_threshold <span class="op">=</span> <span class="fl">0.177</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     het_obs        missingness       to_keep             id           </span></span>
<span><span class="co">##  Min.   :0.3688   Min.   :0.03018   Mode :logical   Length:9          </span></span>
<span><span class="co">##  1st Qu.:0.3774   1st Qu.:0.03642   FALSE:1         Class :character  </span></span>
<span><span class="co">##  Median :0.3795   Median :0.03746   TRUE :8         Mode  :character  </span></span>
<span><span class="co">##  Mean   :0.3851   Mean   :0.03735                                     </span></span>
<span><span class="co">##  3rd Qu.:0.3985   3rd Qu.:0.04058                                     </span></span>
<span><span class="co">##  Max.   :0.4015   Max.   :0.04266</span></span></code></pre>
<p>We can remove the recommended individuals by using:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">individual_report</span><span class="op">$</span><span class="va">id</span> <span class="op">&amp;</span> <span class="va">individual_report</span><span class="op">$</span><span class="va">to_keep</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We can now view a summary of our cleaned data set again, showing that
our data has reduced from 12 to 8 individuals.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       id             genotypes        </span></span>
<span><span class="co">##  Length:8           Length:8          </span></span>
<span><span class="co">##  Class :character   Class :character  </span></span>
<span><span class="co">##  Mode  :character   Mode  :character</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="quality-control-for-loci">Quality control for loci<a class="anchor" aria-label="anchor" href="#quality-control-for-loci"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loci_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## This gen_tibble is not grouped. For Hardy-Weinberg equilibrium, `qc_report_loci()` will assume individuals are part of the same population and HWE test p-values will be calculated across all individuals. If you wish to calculate HWE p-values within populations or groups, please use`group_by()` before calling `qc_report_loci()`.</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">loci_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     snp_id               maf          missingness          hwe_p        </span></span>
<span><span class="co">##  Length:961         Min.   :0.0000   Min.   :0.00000   Min.   :0.00272  </span></span>
<span><span class="co">##  Class :character   1st Qu.:0.1667   1st Qu.:0.00000   1st Qu.:0.32867  </span></span>
<span><span class="co">##  Mode  :character   Median :0.2500   Median :0.00000   Median :0.53333  </span></span>
<span><span class="co">##                     Mean   :0.2661   Mean   :0.03733   Mean   :0.50321  </span></span>
<span><span class="co">##                     3rd Qu.:0.3750   3rd Qu.:0.12500   3rd Qu.:0.69231  </span></span>
<span><span class="co">##                     Max.   :0.5000   Max.   :0.37500   Max.   :0.76503</span></span></code></pre>
<p>The output of <code>qc_report_loci</code> supplies minor allele
frequency, rate of missingness, and a Hardy-Weinberg exact p-value for
each SNP. These data can be visualised in autoplot :</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"overview"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: `aes_string()` was deprecated in ggplot2 3.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use tidy evaluation idioms with `aes()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> See also `vignette("ggplot2-in-packages")` for more information.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">UpSetR</span> package.</span></span>
<span><span class="co">##   Please report the issue to the authors.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">UpSetR</span> package.</span></span>
<span><span class="co">##   Please report the issue to the authors.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Warning: The `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `linewidth` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">UpSetR</span> package.</span></span>
<span><span class="co">##   Please report the issue to the authors.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-10-1.png" alt="UpSet plot giving counts of snps over the threshold for: missingness, minor allele frequency, and Hardy-Weinberg equilibrium P-value" width="700"></p>
<p>Using ‘overview’ provides an Upset plot, which is designed to show
the intersection of different sets in the same way as a Venn diagram.
SNPs can be divided into ‘sets’ that each pass predefined quality
control threshold; a set of SNPs with missingness under a given
threshold, a set of SNPs with MAF above a given threshold, and a set of
SNPs with a Hardy-Weinberg exact p-value that falls above a given
significance level.</p>
<p>The thresholds for each parameter, (percentage of missingness that is
accepted, minor allele frequency cutoff, and Hardy-Weinberg equilibrium
p-value) can be adjusted using the parameters provided in autoplot. For
example:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"overview"</span>,</span>
<span>  miss_threshold <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  maf_threshold <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  hwe_p <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-11-1.png" alt="Upset plot as above, with adjusted thresholds" width="700"></p>
<p>The upset plot then visualises our 961 SNPs within their respective
sets. The number above the second bar indicates that 262 SNPs occur in
all 3 sets, meaning 262 SNPs pass all of our QC thresholds. The combined
total of the first and second bars represents the number of SNPs that
pass our MAF and HWE thresholds, here 939 SNPs.</p>
<p>To examine each QC measure in further detail, we can plot a different
summary panel.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  miss_threshold <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  maf_threshold <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  hwe_p <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-12-1.png" alt="Four panel plot, containing: a histogram of the proportion of missing data for snps with minor allele frequency above the threshold, a histogram of the proportion of missing data for snps with minor allele freqency below the threshold, a histogram of HWE exact test p-values, and a histogram of significant HWE exact test p-values" width="700"></p>
<p>We can then begin to consider how to quality control this raw data
set. Let’s start by filtering SNPs according to their minor allele
frequency. We can visualise the MAF distribution using:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"maf"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-13-1.png" alt="Histogram of minor allele frequency" width="700"></p>
<p>Here we can see there are some monomorphic SNPs in the data set.
Let’s filter out loci with a minor allele frequency lower than 2%, by
using <code>select_loci_if</code>. Here, we select all SNPs with a MAF
greater than 2%. This operation is equivalent to plink –maf 0.02.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.02</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 931</span></span></code></pre>
<p>Following this, we can remove SNPs with a high rate of missingness.
Lets say we want to remove SNPs that are missing in more than 5% of
individuals, equivalent to using plink –geno 0.05</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"missing"</span>, miss_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-15-1.png" alt="Histogram of the proportion of missing data" width="700"></p>
<p>We can see here that most SNPs have low missingness, under our 5%
threshold, some do, however, have missingness over our threshold. To
remove these SNPs, we can again use <code>select_loci_if</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 698</span></span></code></pre>
<p>Finally, we may want to remove SNPs that show significant deviation
from Hardy-Weinberg equilibrium, if our study design requires. To
visualise SNPs with significant p-values in the Hardy-Weinberg exact
test, we can again call autoplot:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"significant hwe"</span>, hwe_p <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-17-1.png" alt="Histogram of significant HWE exact test p-values" width="700"></p>
<p>None of the SNPs in our data are significant, however there may be
circumstances where we would want to cut out the most extreme cases, if
these data were real, these cases could indicate genotyping errors.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_hwe.html">loci_hwe</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 697</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="linkage-disequilibrium">Linkage Disequilibrium<a class="anchor" aria-label="anchor" href="#linkage-disequilibrium"></a>
</h2>
<p>For further analyses, it may be necessary to control for linkage in
the data set. tidypopgen provides LD clumping. This option is similar to
the –indep-pairwise flag in plink, but results in a more even
distribution of loci when compared to LD pruning.</p>
<p>To explore why clumping is preferable to pruning, see <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="external-link uri">https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</a></p>
<p>LD clumping requires a data set with no missingness. This means we
need to create an imputed data set before LD pruning, which we can do
quickly with <code>gt_impute_simple</code>.</p>
<p>Because we have removed individuals through our filtering, we first
need to update the backingfiles with:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_update_backingfile.html">gt_update_backingfile</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Genetic distances are not sorted, setting them to zero</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## gen_backing files updated, now</span></span></code></pre>
<pre><code><span><span class="co">## using FBM RDS: /tmp/Rtmpg5cWNx/file2c30219d84e7_v2.rds</span></span></code></pre>
<pre><code><span><span class="co">## with FBM backing file: /tmp/Rtmpg5cWNx/file2c30219d84e7_v2.bk</span></span></code></pre>
<pre><code><span><span class="co">## make sure that you do NOT delete those files!</span></span></code></pre>
<p>And then we can impute using:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imputed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">data</span>, method <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span></code></pre></div>
<p>In this example, if we want to remove SNPs with a correlation greater
than 0.2 in windows of 10 SNPs at a time, we can set these parameters
with <code>thr_r2</code> and <code>size</code> respectively.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep_ld</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loci_ld_clump.html">loci_ld_clump</a></span><span class="op">(</span><span class="va">imputed_data</span>, thr_r2 <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">to_keep_ld</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE FALSE FALSE FALSE FALSE FALSE</span></span></code></pre>
<p><code>loci_ld_clump</code> provides a boolean vector the same length
as our list of SNPs, telling us which to keep in the data set. We can
then use this list to create a pruned version of our data:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_data</span> <span class="op">&lt;-</span> <span class="va">imputed_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_ld_clump.html">loci_ld_clump</a></span><span class="op">(</span><span class="va">genotypes</span>, thr_r2 <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="save">Save<a class="anchor" aria-label="anchor" href="#save"></a>
</h2>
<p>The benefit of operating on a <code>gen_tibble</code> is that each
quality control step can be observed visually, and easily reversed if
necessary.</p>
<p>When we are happy with the quality of our data, we can create and
save a final quality controlled version of our <code>gen_tibble</code>
using <code>gt_save</code>.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_save.html">gt_save</a></span><span class="op">(</span><span class="va">ld_data</span>, file_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## gen_tibble saved to /tmp/Rtmpg5cWNx/file2c30303e9b.gt</span></span></code></pre>
<pre><code><span><span class="co">## using FBM RDS: /tmp/Rtmpg5cWNx/file2c30219d84e7_v2.rds</span></span></code></pre>
<pre><code><span><span class="co">## with FBM backing file: /tmp/Rtmpg5cWNx/file2c30219d84e7_v2.bk</span></span></code></pre>
<pre><code><span><span class="co">## make sure that you do NOT delete those files!</span></span></code></pre>
<pre><code><span><span class="co">## to reload the gen_tibble in another session, use:</span></span></code></pre>
<pre><code><span><span class="co">## gt_load('/tmp/Rtmpg5cWNx/file2c30303e9b.gt')</span></span></code></pre>
<pre><code><span><span class="co">## [1] "/tmp/Rtmpg5cWNx/file2c30303e9b.gt"      </span></span>
<span><span class="co">## [2] "/tmp/Rtmpg5cWNx/file2c30219d84e7_v2.rds"</span></span>
<span><span class="co">## [3] "/tmp/Rtmpg5cWNx/file2c30219d84e7_v2.bk"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="grouping-data">Grouping data<a class="anchor" aria-label="anchor" href="#grouping-data"></a>
</h2>
<p>For some quality control measures, if you have a gen_tibble that
includes multiple datasets you may want to group by population before
running the quality control. This can be done using
<code>group_by</code>.</p>
<p>First, lets add some imaginary population data to our gen_tibble:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can then group by population and run quality control on each
group:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grouped_loci_report</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grouped_loci_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   snp_id   maf missingness hwe_p</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 2      0.438           0 0.429</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 3      0.188           0 1    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 5      0.25            0 1    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> 7      0.312           0 0.429</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> 8      0.312           0 1.14 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> 10     0.438           0 0.429</span></span></code></pre>
<p>The loci report that we receive here will calculate Hardy-Weinberg
equilibrium for each SNP within each population separately, providing a
Bonferroni corrected p-value for each SNP.</p>
<p>Similarly, we can run a quality control report for individuals within
each population:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grouped_individual_report</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span>kings_threshold <span class="op">=</span> <span class="fl">0.177</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grouped_individual_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">##   het_obs missingness id    group to_keep</span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>   0.382           0 1     A     TRUE   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>   0.377           0 4     A     TRUE   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>   0.393           0 5     A     TRUE   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>   0.382           0 6     A     TRUE   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>   0.403           0 7     B     TRUE   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>   0.382           0 8     B     TRUE</span></span></code></pre>
<p>This is important when we have related individuals, as background
population structure can affect the filtering of relatives.</p>
</div>
<div class="section level2">
<h2 id="grouped-functions">Grouped functions<a class="anchor" aria-label="anchor" href="#grouped-functions"></a>
</h2>
<p>It is also possible to run <code>loci</code> and <code>indiv</code>
functions on grouped data. This is useful when you want to run the same
quality control on each group of data, but don’t want to split the data
into separate gen_tibbles.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loci_maf_grouped</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Grouped functions are built for efficiency and surpass the use of
applying a function with <code>group_map</code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
