<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quality Control">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9016</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_overview.html">Overview</a>
    </li>
    <li>
      <a href="../articles/a02_qc.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/a03_example_clustering_and_dapc.html">Example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control</h1>
            
      

      <div class="hidden name"><code>a02_qc.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="quality-control-for-snp-datasets">Quality control for SNP datasets<a class="anchor" aria-label="anchor" href="#quality-control-for-snp-datasets"></a>
</h2>
<p>tidypopgen has two key functions to examine the quality of data
across loci and across individuals: <code>qc_report_loci</code> and
<code>qc_report_indiv</code>. This vignette uses a simulated data set to
illustrate these methods of data cleaning.</p>
</div>
<div class="section level2">
<h2 id="read-data-into-gen_tibble-format">Read data into gen_tibble format<a class="anchor" aria-label="anchor" href="#read-data-into-gen_tibble-format"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidypopgen</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: tibble</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/related/families.bed"</span>,</span>
<span>  package<span class="op">=</span><span class="st">"tidypopgen"</span><span class="op">)</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  valid_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>,<span class="st">"2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quality-control-for-loci">Quality control for loci<a class="anchor" aria-label="anchor" href="#quality-control-for-loci"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loci_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">loci_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      snp_id         maf          missingness          hwe_p         </span></span>
<span><span class="co">##  Min.   :  1   Min.   :0.0000   Min.   :0.00000   Min.   :0.004644  </span></span>
<span><span class="co">##  1st Qu.:241   1st Qu.:0.1667   1st Qu.:0.00000   1st Qu.:0.340961  </span></span>
<span><span class="co">##  Median :481   Median :0.2727   Median :0.00000   Median :0.523810  </span></span>
<span><span class="co">##  Mean   :481   Mean   :0.2696   Mean   :0.03867   Mean   :0.507695  </span></span>
<span><span class="co">##  3rd Qu.:721   3rd Qu.:0.3750   3rd Qu.:0.08333   3rd Qu.:0.739938  </span></span>
<span><span class="co">##  Max.   :961   Max.   :0.5000   Max.   :0.25000   Max.   :0.796935</span></span></code></pre>
<p>The output of <code>qc_report_loci</code> supplies minor allele
frequency, rate of missingness, and a Hardy-Weinberg exact p-value for
each SNP. These data can be visualised in autoplot :</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"overview"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Using ‘overview’ provides an Upset plot, which is designed to show
the intersection of different sets in the same way as a Venn diagram.
SNPs can be divided into ‘sets’ that each pass predefined quality
control threshold; a set of SNPs with missingness under a given
threshold, a set of SNPs with MAF above a given threshold, and a set of
SNPs with a Hardy-Weinberg exact p-value that falls above a given
significance level.</p>
<p>The upset plot then visualises our 961 SNPs within their respective
sets. The number above the first bar indicates that, of our total SNPs,
609 occur in all 3 sets, meaning 609 SNPs pass all of our automatic QC
measures. The combined total of the first and second bars represents the
number of SNPs that pass our MAF and HWE thresholds, here, 956.</p>
<p>The thresholds for each parameter, (level of missingness that is
accepted etc) can be adjusted using the parameters provided in autoplot.
For example:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"overview"</span>, miss_threshold <span class="op">=</span> <span class="fl">0.03</span>, maf_threshold <span class="op">=</span> <span class="fl">0.02</span>, hwe_p <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>To examine each QC measure in further detail, we can plot a different
summary panel.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"all"</span>, miss_threshold <span class="op">=</span> <span class="fl">0.03</span>, maf_threshold <span class="op">=</span> <span class="fl">0.02</span>, hwe_p <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We can then begin to consider how to quality control this raw data
set. Let’s start by filtering SNPs according to their minor allele
frequency. We can visualise the MAF distribution using:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"maf"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Here we can see there are some monomorphic SNPs in the data set.
Let’s filter out loci with a minor allele frequency lower than 2%, by
using <code>select_loci_if</code>. Here, we select all SNPs with a MAF
greater than 2%. This operation is equivalent to plink –maf 0.02.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.02</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 956</span></span></code></pre>
<p>Following this, we can remove SNPs with a high rate of missingness.
Lets say we want to remove SNPs that are missing in more than 5% of
individuals, equivalent to using plink –geno 0.05</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"missing"</span>, miss_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We can see here that most SNPs have low missingness, under our 5%
threshold, some do, however, have missingness over our threshold. To
remove these SNPs, we can again use <code>select_loci_if</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 609</span></span></code></pre>
<p>Finally, we may want to remove SNPs that show significant deviation
from Hardy-Weinberg equilibrium, if our study design requires. To
visualise SNPs with significant p-values in the Hardy-Weinberg exact
test, we can again call autoplot:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>,type <span class="op">=</span> <span class="st">"significant hwe"</span>, hwe_p <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Few SNPs in our data are significant, however there may be
circumstances where we would want to cut out the most extreme cases, if
these data were real, these cases could indicate genotyping errors.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_hwe.html">loci_hwe</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/count_loci.html">count_loci</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 608</span></span></code></pre>
<p>Once we have quality controlled the SNPs in our data, we can turn to
individual samples.</p>
<p>#Quality control for individuals</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">individual_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     het_obs        missingness</span></span>
<span><span class="co">##  Min.   :0.2206   Min.   :0   </span></span>
<span><span class="co">##  1st Qu.:0.2227   1st Qu.:0   </span></span>
<span><span class="co">##  Median :0.2347   Median :0   </span></span>
<span><span class="co">##  Mean   :0.2351   Mean   :0   </span></span>
<span><span class="co">##  3rd Qu.:0.2409   3rd Qu.:0   </span></span>
<span><span class="co">##  Max.   :0.2591   Max.   :0</span></span></code></pre>
<p>The output of <code>qc_report_indiv</code> supplies observed
heterozygosity per individual, and rate of missingness per individual as
standard.</p>
<p>These data can also be visualised using autoplot:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<p><img src="a02_qc_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Here we can see that most individuals have low missingness. If we
wanted to filter individuals to remove those with more than 3% of their
genotypes missing, we can use <code>filter</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_missingness.html">indiv_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">&lt;</span><span class="fl">0.03</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12</span></span></code></pre>
<p>And if we wanted to remove outliers with particularly high or low
heterozygosity, we can again do so by using <code>filter</code>. As an
example, here we remove observations that lie more than 3 standard
deviations from the mean.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_val</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">$</span><span class="va">het_obs</span><span class="op">)</span></span>
<span><span class="va">sd_val</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">$</span><span class="va">het_obs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lower</span> <span class="op">&lt;-</span> <span class="va">mean_val</span> <span class="op">-</span> <span class="fl">3</span><span class="op">*</span><span class="op">(</span><span class="va">sd_val</span><span class="op">)</span></span>
<span><span class="va">upper</span> <span class="op">&lt;-</span> <span class="va">mean_val</span> <span class="op">+</span> <span class="fl">3</span><span class="op">*</span><span class="op">(</span><span class="va">sd_val</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">lower</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_het_obs.html">indiv_het_obs</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">upper</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12</span></span></code></pre>
<p>Next, we can look at relatedness within our sample. If the parameter
<code>kings_threshold</code> is provided to
<code><a href="../reference/qc_report_indiv.html">qc_report_indiv()</a></code>, then the report also calculates a KING
coefficient of relatedness matrix using the sample. The
<code>kings_threshold</code> is used to provide an output of the largest
possible group with no related individuals in the third column
<code>to_keep</code>. This boolean column recommends which individuals
to remove (FALSE) and to keep (TRUE) to achieve an unrelated sample.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">individual_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="va">data</span>, kings_threshold <span class="op">=</span> <span class="fl">0.177</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">individual_report</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     het_obs        missingness  to_keep              id       </span></span>
<span><span class="co">##  Min.   :0.2206   Min.   :0    Mode :logical   Min.   : 1.00  </span></span>
<span><span class="co">##  1st Qu.:0.2227   1st Qu.:0    FALSE:2         1st Qu.: 3.75  </span></span>
<span><span class="co">##  Median :0.2347   Median :0    TRUE :10        Median : 6.50  </span></span>
<span><span class="co">##  Mean   :0.2351   Mean   :0                    Mean   : 6.50  </span></span>
<span><span class="co">##  3rd Qu.:0.2409   3rd Qu.:0                    3rd Qu.: 9.25  </span></span>
<span><span class="co">##  Max.   :0.2591   Max.   :0                    Max.   :12.00</span></span></code></pre>
<p>We can remove the recommended individuals by using:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">individual_report</span><span class="op">$</span><span class="va">id</span> <span class="op">&amp;</span> <span class="va">individual_report</span><span class="op">$</span><span class="va">to_keep</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We can now view a summary of our cleaned data set again, showing that
our data has reduced from 12 to 10 individuals.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        id          population     genotypes        </span></span>
<span><span class="co">##  Min.   : 1.00   Min.   : 1.00   Length:10         </span></span>
<span><span class="co">##  1st Qu.: 3.25   1st Qu.: 3.25   Class :character  </span></span>
<span><span class="co">##  Median : 5.50   Median : 5.50   Mode  :character  </span></span>
<span><span class="co">##  Mean   : 5.80   Mean   : 5.80                     </span></span>
<span><span class="co">##  3rd Qu.: 7.75   3rd Qu.: 7.75                     </span></span>
<span><span class="co">##  Max.   :12.00   Max.   :12.00</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="linkage-disequilibrium">Linkage Disequilibrium<a class="anchor" aria-label="anchor" href="#linkage-disequilibrium"></a>
</h2>
<p>For further analyses, it may be necessary to control for linkage in
the data set. tidypopgen provides LD clumping. This option is similar to
the –indep-pairwise flag in plink, but results in a more even
distribution of loci when compared to LD pruning.</p>
<p>To explore why clumping is preferable to pruning, see <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="external-link uri">https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</a></p>
<p>LD clumping requires a data set with no missingness. This means we
need to create an imputed data set before LD pruning, which we can do
quickly with <code>gt_impute_simple</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imputed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">data</span>, method <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span></code></pre></div>
<p>In this example, if we want to remove SNPs with a correlation greater
than 0.2 in windows of 10 SNPs at a time, we can set these parameters
with <code>thr_r2</code> and <code>size</code> respectively.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep_LD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loci_ld_clump.html">loci_ld_clump</a></span><span class="op">(</span><span class="va">imputed_data</span>, thr_r2 <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">to_keep_LD</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE FALSE FALSE FALSE  TRUE FALSE</span></span></code></pre>
<p><code>loci_ld_clump</code> provides a boolean vector the same length
as our list of SNPs, telling us which to keep in the data set. We can
then use this list to create a pruned version of our data:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_data</span> <span class="op">&lt;-</span> <span class="va">imputed_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_ld_clump.html">loci_ld_clump</a></span><span class="op">(</span><span class="va">genotypes</span>, thr_r2 <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="save">Save<a class="anchor" aria-label="anchor" href="#save"></a>
</h2>
<p>The benefit of operating on a <code>gen_tibble</code> is that each
quality control step can be observed visually, and easily reversed if
necessary.</p>
<p>When we are happy with the quality of our data, we can create and
save a final quality controlled version of our <code>gen_tibble</code>
using <code>gt_save</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_save.html">gt_save</a></span><span class="op">(</span><span class="va">ld_data</span>, file_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## gen_tibble saved to /tmp/RtmpSWl8gy/file20fa3e8758e9.gt</span></span></code></pre>
<pre><code><span><span class="co">## using bigSNP file: /tmp/RtmpSWl8gy/file20fa36194e6.rds</span></span></code></pre>
<pre><code><span><span class="co">## with backing file: /tmp/RtmpSWl8gy/file20fa36194e6.bk</span></span></code></pre>
<pre><code><span><span class="co">## make sure that you do NOT delete those files!</span></span></code></pre>
<pre><code><span><span class="co">## to reload the gen_tibble in another session, use:</span></span></code></pre>
<pre><code><span><span class="co">## gt_load('/tmp/RtmpSWl8gy/file20fa3e8758e9.gt')</span></span></code></pre>
<pre><code><span><span class="co">## [1] "/tmp/RtmpSWl8gy/file20fa3e8758e9.gt" "/tmp/RtmpSWl8gy/file20fa36194e6.rds"</span></span>
<span><span class="co">## [3] "/tmp/RtmpSWl8gy/file20fa36194e6.bk"</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
