<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Population genetic analysis with tidypopgen • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Population genetic analysis with tidypopgen">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Population genetic analysis with tidypopgen</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/vignettes/a03_example_clustering_and_dapc.Rmd" class="external-link"><code>vignettes/a03_example_clustering_and_dapc.Rmd</code></a></small>
      <div class="d-none name"><code>a03_example_clustering_and_dapc.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="an-example-workflow-with-real-data">An example workflow with real data<a class="anchor" aria-label="anchor" href="#an-example-workflow-with-real-data"></a>
</h3>
<p>We will explore the genetic structure of <em>Anolis punctatus</em> in
South America, using data from Prates et al 2018. We downloaded the vcf
file of the genotypes from “<a href="https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true" class="external-link uri">https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true</a>”
and compressed it to a vcf.gz file.</p>
<p>We read in the data from the compressed vcf with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EvolEcolGroup/tidypopgen" class="external-link">tidypopgen</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: tibble</span></span>
<span><span class="va">vcf_path</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/punctatus_t70_s10_n46_filtered.recode.vcf.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"tidypopgen"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">anole_gt</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">vcf_path</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"anolis_"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s inspect our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 2</span></span></span>
<span><span class="co">#&gt;    id                 genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> punc_BM288                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> punc_GN71                  2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> punc_H1907                 3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> punc_H1911                 4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> punc_H2546                 5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361          6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> punc_ICST764               7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> punc_JFT459                8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> punc_JFT773                9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> punc_LG1299               10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>We can see that we have 46 individuals, from 3249 loci. Note that we
don’t have any information on population from the vcf. That information
can be found from another csv file. We will have add the population
information manually. Let’s start by reading the file:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/punctatus_n46_meta.csv"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"tidypopgen"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">pops_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">46</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">5</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (3): id, population, pop</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (2): longitude, latitude</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="va">pops</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 46 × 5</span></span></span>
<span><span class="co">#&gt;    id                population       longitude latitude pop  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> punc_BM288        Amazonian_Forest     -<span style="color: #BB0000;">51.8</span>    -<span style="color: #BB0000;">3.32</span> Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> punc_GN71         Amazonian_Forest     -<span style="color: #BB0000;">54.6</span>    -<span style="color: #BB0000;">9.73</span> Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> punc_H1907        Amazonian_Forest     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.45</span> Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> punc_H1911        Amazonian_Forest     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.44</span> Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> punc_H2546        Amazonian_Forest     -<span style="color: #BB0000;">65.4</span>    -<span style="color: #BB0000;">9.60</span> Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361 Atlantic_Forest      -<span style="color: #BB0000;">46.0</span>   -<span style="color: #BB0000;">23.8</span>  AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> punc_ICST764      Atlantic_Forest      -<span style="color: #BB0000;">36.3</span>    -<span style="color: #BB0000;">9.81</span> AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> punc_JFT459       Atlantic_Forest      -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.3</span>  AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> punc_JFT773       Atlantic_Forest      -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.0</span>  AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> punc_LG1299       Atlantic_Forest      -<span style="color: #BB0000;">39.1</span>   -<span style="color: #BB0000;">15.3</span>  AF   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>We can now attempt to join the tables. We recommend using a
<code>left_join</code> to do so, rather than <code>cbind</code> or
<code>bind_cols</code>, as the latter functions assume that the two
tables are in the same order. In this case, we do not want to bring over
the wrong data due to mismatched ordering.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">pops</span>, by <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span></span></code></pre></div>
<p>Let us check that we have been successful:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 46</span></span>
<span><span class="co">#&gt; Columns: 6</span></span>
<span><span class="co">#&gt; A tibble: 46 × 6</span></span>
<span><span class="co">#&gt; $ id         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "punc_BM288"<span style="color: #949494;">, </span>"punc_GN71"<span style="color: #949494;">, </span>"punc_H1907"<span style="color: #949494;">, </span>"punc_H1911"<span style="color: #949494;">, </span>"pun…</span></span>
<span><span class="co">#&gt; $ genotypes  <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span> 1<span style="color: #949494;">, </span>2<span style="color: #949494;">, </span>3<span style="color: #949494;">, </span>4<span style="color: #949494;">, </span>5<span style="color: #949494;">, </span>6<span style="color: #949494;">, </span>7<span style="color: #949494;">, </span>8<span style="color: #949494;">, </span>9<span style="color: #949494;">, </span>10<span style="color: #949494;">, </span>11<span style="color: #949494;">, </span>12<span style="color: #949494;">, </span>13<span style="color: #949494;">, </span>14<span style="color: #949494;">, </span>15<span style="color: #949494;">, </span>16<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ population <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Amazonian_Forest"<span style="color: #949494;">, </span>"Amazonian_Forest"<span style="color: #949494;">, </span>"Amazonian_Forest"<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ longitude  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -51.8448<span style="color: #949494;">, </span>-54.6064<span style="color: #949494;">, </span>-64.8247<span style="color: #949494;">, </span>-64.8203<span style="color: #949494;">, </span>-65.3576<span style="color: #949494;">, </span>-46.0247<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ latitude   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -3.3228<span style="color: #949494;">, </span>-9.7307<span style="color: #949494;">, </span>-9.4459<span style="color: #949494;">, </span>-9.4358<span style="color: #949494;">, </span>-9.5979<span style="color: #949494;">, </span>-23.7564<span style="color: #949494;">, </span>-9.8…</span></span>
<span><span class="co">#&gt; $ pop        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Eam"<span style="color: #949494;">, </span>"Eam"<span style="color: #949494;">, </span>"Wam"<span style="color: #949494;">, </span>"Wam"<span style="color: #949494;">, </span>"Wam"<span style="color: #949494;">, </span>"AF"<span style="color: #949494;">, </span>"AF"<span style="color: #949494;">, </span>"AF"<span style="color: #949494;">, </span>"AF"<span style="color: #949494;">, </span>…</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="map">Map<a class="anchor" aria-label="anchor" href="#map"></a>
</h3>
<p>Lets begin by visualising our samples geographically. We have
latitudes and longitudes in our tibble; we can transform them into an
<code>sf</code> geometry with the function <code><a href="../reference/gt_add_sf.html">gt_add_sf()</a></code>.
Once we have done that, our <code>gen_tibble</code> will act as an
<code>sf</code> object, which can be plotted with
<code>ggplot2</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_add_sf.html">gt_add_sf</a></span><span class="op">(</span><span class="va">anole_gt</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"longitude"</span>, <span class="st">"latitude"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; Simple feature collection with 46 features and 6 fields</span></span>
<span><span class="co">#&gt; Geometry type: POINT</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: -75.8069 ymin: -23.7564 xmax: -35.7099 ymax: 4.4621</span></span>
<span><span class="co">#&gt; Geodetic CRS:  WGS 84</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 7</span></span></span>
<span><span class="co">#&gt;    id    genotypes population longitude latitude pop              geometry</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;vctr_SN&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;POINT [°]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> punc…         1 Amazonian…     -<span style="color: #BB0000;">51.8</span>    -<span style="color: #BB0000;">3.32</span> Eam    (-51.8448 -3.3228)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> punc…         2 Amazonian…     -<span style="color: #BB0000;">54.6</span>    -<span style="color: #BB0000;">9.73</span> Eam    (-54.6064 -9.7307)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> punc…         3 Amazonian…     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.45</span> Wam    (-64.8247 -9.4459)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> punc…         4 Amazonian…     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.44</span> Wam    (-64.8203 -9.4358)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> punc…         5 Amazonian…     -<span style="color: #BB0000;">65.4</span>    -<span style="color: #BB0000;">9.60</span> Wam    (-65.3576 -9.5979)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> punc…         6 Atlantic_…     -<span style="color: #BB0000;">46.0</span>   -<span style="color: #BB0000;">23.8</span>  AF    (-46.0247 -23.7564)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> punc…         7 Atlantic_…     -<span style="color: #BB0000;">36.3</span>    -<span style="color: #BB0000;">9.81</span> AF     (-36.2838 -9.8092)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> punc…         8 Atlantic_…     -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.3</span>  AF    (-40.5219 -20.2811)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> punc…         9 Atlantic_…     -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.0</span>  AF        (-40.52 -19.96)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> punc…        10 Atlantic_…     -<span style="color: #BB0000;">39.1</span>   -<span style="color: #BB0000;">15.3</span>  AF    (-39.0694 -15.2696)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>To visualise our samples, we can create a map of South America using
the <code>rnaturalearth</code> package. We will use the
<code><a href="https://docs.ropensci.org/rnaturalearth/reference/ne_countries.html" class="external-link">ne_countries()</a></code> function to get the countries in South
America. We will then plot the map and add our samples using the
<code><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf()</a></code> function from <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/rnaturalearth/" class="external-link">rnaturalearth</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rnaturalearth/reference/ne_countries.html" class="external-link">ne_countries</a></span><span class="op">(</span></span>
<span>  continent <span class="op">=</span> <span class="st">"South America"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"map_units"</span>, scale <span class="op">=</span> <span class="st">"medium"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">geometry</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span></span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">85</span>, <span class="op">-</span><span class="fl">30</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">30</span>, <span class="fl">15</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>From our map we can see that we have samples from the coastal
Atlantic Forest and the Amazonian Forest.</p>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>That was easy. The loci had already been filtered and cleaned, so we
don’t need to do any QC. Let us jump straight into analysis and run a
PCA:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: You can't have missing values in 'X'.</span></span></code></pre></div>
<p>OK, we jumped too quickly. There are missing data, and we need first
to impute them:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">anole_gt</span>, method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></code></pre></div>
<p>And now:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Let us look at the object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span></span>
<span><span class="co">#&gt;  === PCA of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Method: [1] "partialSVD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call ($call):gt_pca_partialSVD(x = ., k = 30)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($d):</span></span>
<span><span class="co">#&gt;  351.891 192.527 113.562 104.427 87.615 83.476 ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Principal component scores ($u):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 30 columns (axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings (Principal axes) ($v):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (SNPs) and 30 columns (axes)</span></span></code></pre></div>
<p>The <code>print</code> function (implicitly called when we type the
name of the object) gives us information about the most important
elements in the object (and the names of the elements in which they are
stored).</p>
<p>We can extract those elements with the <code>tidy</code> function,
which returns a tibble that can be easily used for further analysis,
e.g.:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>, matrix <span class="op">=</span> <span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 4</span></span></span>
<span><span class="co">#&gt;       PC std.dev percent cumulative</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1   52.5    45.9        45.9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2   28.7    13.7        59.7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3   16.9     4.78       64.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4   15.6     4.05       68.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5   13.1     2.85       71.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6   12.4     2.58       73.9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7   10.3     1.79       75.7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8   10.0     1.68       77.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9    9.23    1.42       78.8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10    8.90    1.32       80.2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more rows</span></span></span></code></pre></div>
<p>We can return information on the <em>eigenvalues</em>,
<em>scores</em> and <em>loadings</em> of the pca. There is also an
<code>autoplot</code> method that allows to visualise those elements
(type <em>screeplot</em> for <em>eigenvalues</em>, type <em>scores</em>
for <em>scores</em>, and <em>loadings</em> for <em>loadings</em>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-14-1.png" alt="Scree Plot of eigenvalues for each Principal Component" width="700"></p>
<p>To plot the sample in principal coordinates space, we can simply
use:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-15-1.png" alt="Score plot of individuals across the first and second Principal Components" width="700"></p>
<p><code>autoplots</code> are deliberately kept simple: they are just a
way to quickly inspect the results. They generate <code>ggplot2</code>
objects, and so they can be further embellished with the usual
<code>ggplot2</code> grammar:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-16-1.png" alt="Score plot of individuals across the first and second Principal Components, with individual samples coloured by population" width="700"></p>
<p>For more complex/publication ready plots, we will want to add the PC
scores to the tibble, so that we can create a custom plot with
<code>ggplot2</code>. We can easily add the data with the
<code>augment</code> method:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span><span class="va">anole_pca</span>, data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
<p>And now we can use <code>ggplot2</code> directly to generate our
plot:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.fittedPC1</span>, <span class="va">.fittedPC2</span>, color <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"PC1"</span>, y <span class="op">=</span> <span class="st">"PC2"</span>, color <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-18-1.png" alt="Another score plot of individuals across the first and second Principal Components, with individual samples coloured by population, using ggplot2" width="700"></p>
<p>We can see that the two population separate clearly in the PCA space,
with individuals from the Atlantic Forest clustering closely together
while Amazonian Forest individuals are spread across the first two
principal components.</p>
<p>It is also possible to inspect which loci contribute the most to a
given component:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-19-1.png" alt="Plot of the loadings of Principal Component 1 for each loci" width="700"></p>
<p>By using information from the loci table, we could easily embellish
the plot, for example colouring by chromosome or maf.</p>
<p>For more complex plots, we can augment the loci table with the
loadings using <code><a href="../reference/augment_loci.html">augment_loci()</a></code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/augment_loci.html">augment_loci</a></span><span class="op">(</span><span class="va">anole_pca</span>, data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="explore-population-structure-with-dapc">Explore population structure with DAPC<a class="anchor" aria-label="anchor" href="#explore-population-structure-with-dapc"></a>
</h2>
<p>DAPC is a powerful tool to investigate population structure. It has
the advantage of scaling well to very large datasets. It does not have
the assumptions of STRUCTURE or ADMIXTURE (which also limits its
power).</p>
<p>The first step is to determine the number of genetic clusters in the
dataset. DAPC can be either used to test a a-priori hypothesis, or we
can use the data to suggest the number of clusters. In this case, we did
not have any strong expectations of structure in our study system, so we
will let the data inform the number of possible genetic clusters. We
will use a k-clustering algorithm applied to the principal components
(allowing us to reduce the dimensions from the thousands of loci to just
a few tens of components). We need to decide how many components to use;
this decision is often made based on a plot of the cumulative explained
variance of the components. Using <code>tidy</code> on the
<code>gt_pca</code> object allows us easily obtain those quantities, and
it is then trivial to plot them:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>, matrix <span class="op">=</span> <span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC</span>, y <span class="op">=</span> <span class="va">cumulative</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-21-1.png" alt="Plot of the cumulative loadings for each Principal Component" width="700"></p>
<p>Note that, as we were working with a truncated SVD algorithm for our
PCA, we can not easily phrase the eigenvalues in terms of proportion of
total variance, so the cumulative y axis simply shows the cumulative sum
of the eigenvalues. Ideally, we are looking for the point where the
curve starts flattening. In this case, we can not see a very clear
flattening, but by PC 10 the increase in explained variance has markedly
decelerated. We can now find clusters based on those 10 PCs:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca.html">gt_cluster_pca</a></span><span class="op">(</span><span class="va">anole_pca</span>, n_pca <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>As we did not define the <em>k</em> values to explore, the default 1
to 5 was used (we can change that by setting the <code>k</code>
parameter to change the range). To choose an appropriate <em>k</em>, we
plot the number of clusters against a measure of fit. BIC has been shown
to be a very good metric under many scenarios:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-23-1.png" alt="Plot of BIC (Bayesian Information Criteria) for each value of k" width="700"></p>
<p>We are looking for the minimum value of BIC. There is no clear elbow
(a minimum after which BIC increases with increasing k). However, we
notice that there is a quick levelling off in the decrease in BIC at 3
clusters. Arguably, these are sufficient to capture the main structure.
We can also use a number of algorithmic approaches (based on the
original <code><a href="https://rdrr.io/pkg/adegenet/man/find.clusters.html" class="external-link">find.clusters()</a></code> function in <code>adegenet</code>)
to choose the best <em>k</em> value from this plot through
<code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>. We will use the defaults (BIC with
“diffNgroup”, see the help page for <code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>
for a description of the various options):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using BIC with criterion diffNgroup: 3 clusters</span></span></code></pre></div>
<p>The algorithm confirms our choice. Note that this function simply
adds an element <code>$best_k</code> to the <code>gt_cluster_pca</code>
object:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span><span class="op">$</span><span class="va">best_k</span></span>
<span><span class="co">#&gt; [1] 3</span></span></code></pre></div>
<p>If we decided that we wanted to explore a different value, we could
simply overwrite that number with
<code>anole_clusters$best_k&lt;-5</code></p>
<p>In this case, we are happy with the option of 3 clusters, and we can
run a DAPC:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_dapc.html">gt_dapc</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code><a href="../reference/gt_dapc.html">gt_dapc()</a></code> takes automatically the number of
clusters from the <code>anole_clusters</code> object, but can change
that behaviour by setting some of its parameters (see the help page for
<code><a href="../reference/gt_dapc.html">gt_dapc()</a></code>). When we print the object, we are given
information about the most important elements of the object and where to
find them (as we saw for <code>gt_pca</code>):</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span></span>
<span><span class="co">#&gt;  === DAPC of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Call ($call):gt_dapc(x = anole_clusters)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($eig):</span></span>
<span><span class="co">#&gt;  727.414 218.045 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; LD scores ($ind.coord):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by PC ($loadings):</span></span>
<span><span class="co">#&gt;  matrix with 2 rows (PC axes) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by locus($var.load):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (loci) and 2 columns (LD axes)</span></span></code></pre></div>
<p>Again, these elements can be obtained with tidiers (with
<code>matrix</code> equal to <code>eigenvalues</code>,
<code>scores</code>,<code>ld_loadings</code> and
<code>loci_loadings</code>):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix <span class="op">=</span> <span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;      LD eigenvalue cumulative</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1       727.       727.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2       218.       945.</span></span></code></pre></div>
<p>And they can be visualised with <code>autoplot</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type <span class="op">=</span> <span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-29-1.png" alt="Scree plot of the eigenvalues on the two discriminant axes (defined by `ld`)" width="700"></p>
<p>As for pca, there is a <code>tidy</code> method that can be used to
extract information from <code>gt_dapc</code> objects. For example, if
we want to create a bar plot of the eigenvalues (since we only have
two), we could simply use:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix <span class="op">=</span> <span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">LD</span>, y <span class="op">=</span> <span class="va">eigenvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-30-1.png" alt="Bar plot of eigenvalues against the two discriminant axes" width="700"></p>
<p>We can plot the scores with:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-31-1.png" alt="Scatterplot of the scores of each individual on the two discriminant axes (defined by `ld`)" width="700"></p>
<p>The DAPC plot shows the separation of individuals into 3
clusters.</p>
<p>We can inspect this assignment by DAPC with <code>autoplot</code>
using the type <code>components</code>, ordering the samples by their
original population labels:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type <span class="op">=</span> <span class="st">"components"</span>, group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-32-1.png" alt="Bar plot showing the probability of assignment to each cluster" width="700"></p>
<p>Here we can see that DAPC splits the Amazonian Forest individuals
into two clusters. Because of the very clear separation we observed when
plotting the LD scores, no individual is modelled as a mixture: all
assignments are with 100% probability to a single cluster.</p>
<p>Finally, we can explore which loci have the biggest impact on
separating the clusters (either because of drift or selection):</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-33-1.png" alt="Plot of the loadings of Principal Component 1 for each loci" width="700"></p>
<p>There is no strong outlier, suggesting drift across many loci has
created the signal picked up by DAPC.</p>
<p>Note that <code>anole_dapc</code> is of class <code>gt_dapc</code>,
which is a subclass of <code>dapc</code> from <code>adegenet</code>.
This means that functions written to work on <code>dapc</code> objects
should work out of the box (the only exception is
<code><a href="https://rdrr.io/pkg/adegenet/man/dapc.html" class="external-link">adegenet::predict.dapc</a></code>, which does not work because the
underlying pca object is different). For example, we can obtain the
standard dapc plot with:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ade4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    /// adegenet 2.1.11 is loaded ////////////</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    &gt; overview: '?adegenet'</span></span>
<span><span class="co">#&gt;    &gt; tutorials/doc/questions: 'adegenetWeb()' </span></span>
<span><span class="co">#&gt;    &gt; bug reports/feature requests: adegenetIssues()</span></span>
<span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/scatter.html" class="external-link">scatter</a></span><span class="op">(</span><span class="va">anole_dapc</span>, posi.da <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-34-1.png" alt="Scatterplot of individuals with ellipses surrounding each cluster of individuals identified by the DAPC analysis. Eigenvalues are displayed in an inset in the bottom right corner of the plot." width="700"></p>
<p>We can also plot these results onto the map we created earlier.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dapc <span class="op">=</span> <span class="va">anole_dapc</span><span class="op">$</span><span class="va">grp</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">geometry</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">dapc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span></span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">85</span>, <span class="op">-</span><span class="fl">30</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">30</span>, <span class="fl">15</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"DAPC cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>Atlantic forest lizards form distinct geographic cluster, separate
from the Amazonian lizards.</p>
</div>
<div class="section level2">
<h2 id="clustering-with-admixture">Clustering with ADMIXTURE<a class="anchor" aria-label="anchor" href="#clustering-with-admixture"></a>
</h2>
<p>ADMIXTURE is a fast clustering algorithm which provides results
similar to STRUCTURE. We can run it directly from tidypopgen using the
<code>gt_admixture</code> function.</p>
<p>The output of <code>gt_admixture</code> is a <code>gt_admix</code>
object. <code>gt_admix</code> objects are designed to make visualising
data from different clustering algorithms swift and easy.</p>
<p>Lets begin by running <code>gt_admixture</code>. We can run K
clusters from k=2 to k=6. We will use just one repeat, but ideally we
should run multiple repetitions for each K:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anole_adm_original</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_admixture.html">gt_admixture</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">anole_gt</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>,</span>
<span>  n_runs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  crossval <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that within this vignette we are not running the
<code>gt_admixture</code> function, as it requires the ADMIXTURE
software to be installed on your machine and available in your PATH.
Instead, we load the results from a previous run of ADMIXTURE. If you
would prefer a native clustering algorithm, we recommend using the
<code>gt_snmf</code> function, which is a wrapper for the
<code>snmf</code> function from the <code>LEA</code> package. sNMF is a
fast clustering algorithm which provides results similar to STRUCTURE
and ADMIXTURE, and is available directly using the
<code>tidypopgen</code> package. We can examine the suitability of our K
values by plotting the cross-entropy values for each K value, which is
contained in the <code>cv</code> element of the <code>gt_admix</code>
object:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_adm</span>, type <span class="op">=</span> <span class="st">"cv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-38-1.png" alt="Plot of cross-entropy for each value of K, showing K = 3 has the lowest cross-entropy" width="700"></p>
<p>Here we can see that K = 3 is a sensible choice, as K = 3 represents
the ‘elbow’ in the plot, where cross-entropy is at it’s lowest.</p>
<p>We can quickly plot this data with <code>autoplot</code> by using the
type <code>barplot</code> and selecting our chosen value of
<code>k</code>:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_adm</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-39-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>This plot is fine, but it is messy, and not a very helpful
visualisation if we want to understand how our populations are
structured.</p>
<p>To re-order this plot, grouping individuals by population, we can use
the <code>annotate_group</code> and <code>arrange_by_group</code>
arguments in autoplot:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_adm</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">anole_gt</span>,</span>
<span>  annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-40-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>This plot is much better. Now the plot is labelled by population we
can see that the individuals from each population are grouped together,
with a few individuals showing mixed ancestry between populations.</p>
<p>However, for a more visually appealing plot, we can re-order the
individuals within each population as follows:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_adm</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-41-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources, with individuals arranged according to their dominant ancestry propotion" width="700"></p>
<p>Now individuals are neatly ordered within each population.</p>
<p>For more complex/publication ready plots, we can extract the specific
K value and run that we are interested in plotting, and add this .Q
matrix to our gen_tibble object, so that we can create a custom plot
with <code>ggplot2</code>.</p>
<p>First we extract the q_matrix of interest using the
<code>get_q_matrix</code> function:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">anole_adm</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Then we can easily add the data with the <code>augment</code>
method:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_adm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span><span class="va">q_mat</span>, data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anole_gt_adm</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 6 features and 42 fields</span></span>
<span><span class="co">#&gt; Geometry type: POINT</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: -65.3576 ymin: -23.7564 xmax: -46.0247 ymax: -3.3228</span></span>
<span><span class="co">#&gt; Geodetic CRS:  WGS 84</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 43</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   population [2]</span></span></span>
<span><span class="co">#&gt;   id     genotypes population longitude latitude pop              geometry</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;vctr_SN&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;POINT [°]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> punc_…         1 Amazonian…     -<span style="color: #BB0000;">51.8</span>    -<span style="color: #BB0000;">3.32</span> Eam    (-51.8448 -3.3228)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> punc_…         2 Amazonian…     -<span style="color: #BB0000;">54.6</span>    -<span style="color: #BB0000;">9.73</span> Eam    (-54.6064 -9.7307)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> punc_…         3 Amazonian…     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.45</span> Wam    (-64.8247 -9.4459)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> punc_…         4 Amazonian…     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.44</span> Wam    (-64.8203 -9.4358)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> punc_…         5 Amazonian…     -<span style="color: #BB0000;">65.4</span>    -<span style="color: #BB0000;">9.60</span> Wam    (-65.3576 -9.5979)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> punc_…         6 Atlantic_…     -<span style="color: #BB0000;">46.0</span>   -<span style="color: #BB0000;">23.8</span>  AF    (-46.0247 -23.7564)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more variables: .rownames &lt;chr&gt;, .fittedPC1 &lt;dbl&gt;, .fittedPC2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC3 &lt;dbl&gt;, .fittedPC4 &lt;dbl&gt;, .fittedPC5 &lt;dbl&gt;, .fittedPC6 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC7 &lt;dbl&gt;, .fittedPC8 &lt;dbl&gt;, .fittedPC9 &lt;dbl&gt;, .fittedPC10 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC11 &lt;dbl&gt;, .fittedPC12 &lt;dbl&gt;, .fittedPC13 &lt;dbl&gt;, .fittedPC14 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC15 &lt;dbl&gt;, .fittedPC16 &lt;dbl&gt;, .fittedPC17 &lt;dbl&gt;, .fittedPC18 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC19 &lt;dbl&gt;, .fittedPC20 &lt;dbl&gt;, .fittedPC21 &lt;dbl&gt;, .fittedPC22 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC23 &lt;dbl&gt;, .fittedPC24 &lt;dbl&gt;, .fittedPC25 &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>Alternatively, we can convert our q matrix into <code>tidy</code>
format, which is more suitable for plotting:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">q_mat</span>, <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tidy_q</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   id         group            q     percentage</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> punc_BM288 Amazonian_Forest .Q1     0.000<span style="text-decoration: underline;">019</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> punc_BM288 Amazonian_Forest .Q2     0.000<span style="text-decoration: underline;">01</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> punc_BM288 Amazonian_Forest .Q3     1.00    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> punc_GN71  Amazonian_Forest .Q1     0.000<span style="text-decoration: underline;">01</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> punc_GN71  Amazonian_Forest .Q2     0.000<span style="text-decoration: underline;">01</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> punc_GN71  Amazonian_Forest .Q3     1.00</span></span></code></pre></div>
<p>And now we can use <code>ggplot2</code> directly to generate our
custom plot:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="va">tidy_q</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dominant_q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">percentage</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">group</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">dominant_q</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    plot_order <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">id</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_q</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">id</span>, y <span class="op">=</span> <span class="va">percentage</span>, fill <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span></span>
<span>    width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>reverse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    y <span class="op">=</span> <span class="st">"Population Structure for K = 3"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"ADMIXTURE algorithm on A. punctatus"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_distruct.html">theme_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-45-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>In the Prates et al 2018 data, anolis lizards were assigned to three
regions, described as Af, Eam, and Wam, acronyms which correspond to the
Atlantic Forest, Eastern Amazonia, and Western Amazonia of Brazil. Our
metadata also contains this assignment, in the column
<code>pop</code>.</p>
<p>Lets say we wanted to change the grouping variable of our plot to
match these regions. We can use the <code>gt_admix_reorder_q</code>
function to reorder the Q matrix by a different grouping variable:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_adm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_admix_reorder_q.html">gt_admix_reorder_q</a></span><span class="op">(</span><span class="va">anole_adm</span>, group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span></span></code></pre></div>
<p>And replot the data:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_adm</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-47-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources, labelled Atlantic forest or Amazonian forest" width="700"></p>
<p>And here we can see the clear distinction between the three regions,
with a few individuals that have admixed ancestry.</p>
</div>
<div class="section level2">
<h2 id="handling-q-matrices">Handling Q matrices<a class="anchor" aria-label="anchor" href="#handling-q-matrices"></a>
</h2>
<p>For clustering algorithms that operate outside the R environment, the
function <code><a href="../reference/q_matrix.html">q_matrix()</a></code> can take a path to a directory
containing results from multiple runs of a clustering algorithm for
multiple values of K, read the .Q files, and summarise them in a single
<code>q_matrix_list</code> object.</p>
<p>In the analysis above, through <code>gt_admxiture()</code>, we ran 1
repetition for K values 2 to 6. Let us now collate the results from a
standard run of ADMIXTURE for which we stored the files in a
directory:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adm_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"anolis_adm"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">adm_dir</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "K2run1.Q" "K3run1.Q" "K4run1.Q" "K5run1.Q" "K6run1.Q"</span></span></code></pre></div>
<p>And read them back into our R environment using
<code>q_matrix_list</code>:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_q_files.html">read_q_files</a></span><span class="op">(</span><span class="va">adm_dir</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">q_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; Admixture results for multiple runs:           </span></span>
<span><span class="co">#&gt; k 2 3 4 5 6</span></span>
<span><span class="co">#&gt; n 1 1 1 1 1</span></span>
<span><span class="co">#&gt; with slots:</span></span>
<span><span class="co">#&gt; $Q for Q matrices</span></span></code></pre></div>
<p><code>q_matrix_list</code> has read and summarised the .Q files from
our analysis, and we can access a single matrix from the list by
selecting the run number and K value of interest using
<code>get_q_matrix</code> as before.</p>
<p>For example, if we would like to view the second run of K = 3:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">q_list</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           .Q1     .Q2      .Q3</span></span>
<span><span class="co">#&gt; [1,] 0.000019 0.00001 0.999971</span></span>
<span><span class="co">#&gt; [2,] 0.000010 0.00001 0.999980</span></span>
<span><span class="co">#&gt; [3,] 0.000010 0.99998 0.000010</span></span>
<span><span class="co">#&gt; [4,] 0.000010 0.99998 0.000010</span></span>
<span><span class="co">#&gt; [5,] 0.000010 0.99998 0.000010</span></span>
<span><span class="co">#&gt; [6,] 0.999980 0.00001 0.000010</span></span></code></pre></div>
<p>And, again, we can then autoplot any matrix by selecting from the
<code>q_matrix_list</code> object:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">q_list</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  data <span class="op">=</span> <span class="va">anole_gt</span>,</span>
<span>  annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-52-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>In this way, <code>tidypopgen</code> integrates with external
clustering software seamlessly for quick, easy plotting.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
