<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Example workflow with tidypopgen • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Example workflow with tidypopgen">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9022</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">Overview</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Example workflow with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Example workflow with tidypopgen</h1>
            
      

      <div class="d-none name"><code>a03_example_clustering_and_dapc.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="an-example-workflow-with-real-data">An example workflow with real data<a class="anchor" aria-label="anchor" href="#an-example-workflow-with-real-data"></a>
</h3>
<p>We will explore the genetic structure of <em>Anolis punctatus</em> in
South America, using data from Prates et al 2018. We downloaded the vcf
file of the genotypes from “<a href="https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true" class="external-link uri">https://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true</a>”
and compressed it to a vcf.gz file.</p>
<p>We read in the data from the compressed vcf with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://evolecolgroup.github.io/tidypopgen">tidypopgen</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: tibble</span></span>
<span><span class="va">vcf_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/punctatus_t70_s10_n46_filtered.recode.vcf.gz"</span>,</span>
<span>                          package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">vcf_path</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"anolis_"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s inspect our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 2</span></span></span>
<span><span class="co">#&gt;    id                 genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> punc_BM288                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> punc_GN71                  2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> punc_H1907                 3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> punc_H1911                 4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> punc_H2546                 5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361          6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> punc_ICST764               7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> punc_JFT459                8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> punc_JFT773                9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> punc_LG1299               10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>We can see that we have 46 individuals, from 3249 loci. Note that we
don’t have any information on population from the vcf. That information
can be found from another file on the github repository (“<a href="https://github.com/ivanprates/2018_Anolis_EcolEvol/raw/master/data/plot_order_punctatus_n46.csv" class="external-link uri">https://github.com/ivanprates/2018_Anolis_EcolEvol/raw/master/data/plot_order_punctatus_n46.csv</a>).
We will have add the population information manually. Let’s start by
reading the file:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/plot_order_punctatus_n46.csv"</span>,</span>
<span>                        package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span></span>
<span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">pops_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">46</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (2): ID, pop</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (1): plot_order</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="va">pops</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 46 × 3</span></span></span>
<span><span class="co">#&gt;    ID           plot_order pop  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288                18 Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71                 21 Eam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907                 6 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911                 4 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546                 7 Wam  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361         34 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764              31 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459               37 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773               38 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299               44 AF   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>The ids from the VCF are in a different format than the ones we just
got from the pop csv. We need a bit of string wrangling, but it looks
easy, we just need to remove “punc_”:</p>
<p>Let us simplify the ids, which will have a “punc_” prefix</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'punc_'</span>,<span class="st">""</span>,<span class="va">.data</span><span class="op">$</span><span class="va">id</span>,<span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 2</span></span></span>
<span><span class="co">#&gt;    id            genotypes</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71                  2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907                 3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911                 4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546                 5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361          6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764               7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459                8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773                9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299               10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
<p>Now we can bring in the pop information:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="va">pops</span><span class="op">$</span><span class="va">pop</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">pops</span><span class="op">$</span><span class="va">ID</span>,<span class="va">.data</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">anole_gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     46 × 3</span></span></span>
<span><span class="co">#&gt;    id            genotypes population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM288                 1 Eam       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GN71                  2 Eam       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> H1907                 3 Wam       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1911                 4 Wam       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2546                 5 Wam       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IBSPCRIB0361          6 AF        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ICST764               7 AF        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> JFT459                8 AF        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> JFT773                9 AF        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LG1299               10 AF        </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>That was easy. The loci had already been filtered and cleaned, so we
don’t need to do any QC. Let us jump straight into analysis and run a
PCA:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: You can't have missing values in 'X'.</span></span></code></pre></div>
<p>OK, we jumped too quickly. There are missing data, and we need first
to impute them:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">anole_gt</span>, method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></code></pre></div>
<p>And now:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span>k<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Let us look at the object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_pca</span></span>
<span><span class="co">#&gt;  === PCA of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Method: [1] "partialSVD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call ($call):gt_pca_partialSVD(x = ., k = 30)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($d):</span></span>
<span><span class="co">#&gt;  351.891 192.527 113.562 104.427 87.615 83.476 ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Principal component scores ($u):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 30 columns (axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings (Principal axes) ($v):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (SNPs) and 30 columns (axes)</span></span></code></pre></div>
<p>The <code>print</code> function (implicitly called when we type the
name of the object) gives us information about the most important
elements in the object (and the names of the elements in which they are
stored).</p>
<p>We can extract those elements with the <code>tidy</code> function,
which returns a tibble that can be easily used for further analysis,
e.g.:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 3</span></span></span>
<span><span class="co">#&gt;       PC std.dev cumulative</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1   51.9        51.9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2   28.4        80.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3   16.7        97.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4   15.4       112. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5   12.9       125. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6   12.3       138. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7   10.2       148. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8    9.92      158. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9    9.13      167. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10    8.80      176. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more rows</span></span></span></code></pre></div>
<p>We can return information on the <em>eigenvalues</em>,
<em>scores</em> and <em>loadings</em> of the pca. There is also an
<code>autoplot</code> method that allows to visualise those elements
(type <em>screeplot</em> for <em>eigenvalues</em>, type <em>scores</em>
for <em>scores</em>, and <em>loadings</em> for <em>loadings</em>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type<span class="op">=</span><span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-12-1.png" alt="Scree Plot of eigenvalues for each Principal Component" width="700"></p>
<p>To plot the sample in principal coordinates space, we can simply
use:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span><span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-13-1.png" alt="Score plot of individuals across the first and second Principal Components" width="700"></p>
<p><code>autoplots</code> are deliberately kept simple: they are just a
way to quickly inspect the results. They generate <code>ggplot2</code>
objects, and so they can be further embellished with the usual
<code>ggplot2</code> grammar:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-14-1.png" alt="Score plot of individuals across the first and second Principal Components, with individual samples coloured by population" width="700"></p>
<p>For more complex/publication ready plots, we will want to add the PC
scores to the tibble, so that we can create a custom plot with
<code>ggplot2</code>. We can easily add the data with the
<code>augment</code> method:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span><span class="va">anole_pca</span> , data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
<p>And now we can use <code>ggplot2</code> directly to generate our
plot:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.fittedPC1</span>, <span class="va">.fittedPC2</span>, color <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-16-1.png" alt="Another score plot of individuals across the first and second Principal Components, with individual samples coloured by population, using ggplot2" width="700"></p>
<p>We can see that the three population do separate nicely on the PCA,
with just one individual from Wam sitting in-between the other Wam
individuals and those from Eam.</p>
<p>It is also possible to inspect which loci contribute the most to a
given component:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_pca</span>, type <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-17-1.png" alt="Plot of the loadings of Principal Component 1 for each loci" width="700"></p>
<p>By using information from the loci table, we could easily embellish
the plot, for example colouring by chromosome or maf.</p>
<p>For more complex plots, we can augment the loci table with the
loadings using <code><a href="../reference/augment_loci.html">augment_loci()</a></code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/augment_loci.html">augment_loci</a></span><span class="op">(</span><span class="va">anole_pca</span>, data<span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="explore-population-structure-with-dapc">Explore population structure with DAPC<a class="anchor" aria-label="anchor" href="#explore-population-structure-with-dapc"></a>
</h2>
<p>DAPC is a powerful tool to investigate population structure. It has
the advantage of scaling well to very large datasets. It does not have
the assumptions of STRUCTURE or ADMIXTURE (which also limits its
power).</p>
<p>The first step is to determine the number of genetic clusters in the
dataset. DAPC can be either used to test a a-priori hypothesis, or we
can use the data to suggest the number of clusters. In this case, we did
not have any strong expectations of structure in our study system, so we
will let the data inform the number of possible genetic clusters. We
will use a k-clustering algorithm applied to the principal components
(allowing us to reduce the dimensions from the thousands of loci to just
a few tens of components). We need to decide how many components to use;
this decision is often made based on a plot of the cumulative explained
variance of the components. Using <code>tidy</code> on the
<code>gt_pca</code> object allows us easily obtain those quantities, and
it is then trivial to plot them:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_pca</span>,matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC</span>, y<span class="op">=</span><span class="va">cumulative</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-19-1.png" alt="Plot of the cumulative loadings for each Principal Component" width="700"></p>
<p>Note that, as we were working with a truncated SVD algorithm for our
PCA, we can not easily phrase the eigenvalues in terms of proportion of
total variance, so the cumulative y axis simply shows the cumulative sum
of the eigenvalues. Ideally, we are looking for the point where the
curve starts flattening. In this case, we can not see a very clear
flattening, but by PC 10 the increase in explained variance has markedly
decelerated. We can now find clusters based on those 10 PCs:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca.html">gt_cluster_pca</a></span><span class="op">(</span><span class="va">anole_pca</span>, n_pca <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>As we did not define the <em>k</em> values to explore, the default 1
to 5 was used (we can change that by setting the <code>k</code>
parameter to change the range). To choose an appropriate <em>k</em>, we
plot the number of clusters against a measure of fit. BIC has been shown
to be a very good metric under many scenarios:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-21-1.png" alt="Plot of BIC (Bayesian Information Criteria) for each value of k" width="700"></p>
<p>We are looking for the minimum value of BIC. There is no clear elbow
(a minimum after which BIC increases with increasing k). However, we
notice that there is a quick levelling off in the decrease in BIC at 3
clusters. Arguably, these are sufficient to capture the main structure
(and that makes sense given what we saw in the PCA). We can also use a
number of algorithmic approaches (based on the original
<code><a href="https://rdrr.io/pkg/adegenet/man/find.clusters.html" class="external-link">find.clusters()</a></code> function in <code>adegenet</code>) to
choose the best <em>k</em> value from this plot through
<code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>. We will use the defaults (BIC with
“diffNgroup”, see the help page for <code><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k()</a></code>
for a description of the various options):</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_cluster_pca_best_k.html">gt_cluster_pca_best_k</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using BIC with criterion diffNgroup: 3 clusters</span></span></code></pre></div>
<p>The algorithm confirms our choice. Note that this function simply
adds an element <code>$best_k</code> to the <code>gt_cluster_pca</code>
object:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clusters</span><span class="op">$</span><span class="va">best_k</span></span>
<span><span class="co">#&gt; [1] 3</span></span></code></pre></div>
<p>If we decided that we wanted to explore a different value, we could
simply overwrite that number with
<code>anole_clusters$best_k&lt;-5</code></p>
<p>In this case, we are happy with the option of 3 clusters, and we can
run a DAPC:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_dapc.html">gt_dapc</a></span><span class="op">(</span><span class="va">anole_clusters</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code><a href="../reference/gt_dapc.html">gt_dapc()</a></code> takes automatically the number of
clusters from the <code>anole_clusters</code> object, but can change
that behaviour by setting some of its parameters (see the help page for
<code><a href="../reference/gt_dapc.html">gt_dapc()</a></code>). When we print the object, we are given
information about the most important elements of the object and where to
find them (as we saw for <code>gt_pca</code>):</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_dapc</span></span>
<span><span class="co">#&gt;  === DAPC of gen_tibble object ===</span></span>
<span><span class="co">#&gt; Call ($call):gt_dapc(x = anole_clusters)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues ($eig):</span></span>
<span><span class="co">#&gt;  727.414 218.045 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; LD scores ($ind.coord):</span></span>
<span><span class="co">#&gt;  matrix with 46 rows (individuals) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by PC ($loadings):</span></span>
<span><span class="co">#&gt;  matrix with 2 rows (PC axes) and 2 columns (LD axes) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loadings by locus($var.load):</span></span>
<span><span class="co">#&gt;  matrix with 3249 rows (loci) and 2 columns (LD axes)</span></span></code></pre></div>
<p>Again, these elements can be obtained with tidiers (with
<code>matrix</code> equal to <code>eigenvalues</code>,
<code>scores</code>,<code>ld_loadings</code> and
<code>loci_loadings</code>):</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;      LD eigenvalue cumulative</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1       727.       727.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2       218.       945.</span></span></code></pre></div>
<p>And they can be visualised with <code>autoplot</code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"screeplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-27-1.png" alt="Scree plot of the eigenvalues on the two discriminant axes (defined by `ld`)" width="700"></p>
<p>As for pca, there is a <code>tidy</code> method that can be used to
extract information from <code>gt_dapc</code> objects. For example, if
we want to create a bar plot of the eigenvalues (since we only have
two), we could simply use:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">anole_dapc</span>, matrix<span class="op">=</span><span class="st">"eigenvalues"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">LD</span>,y<span class="op">=</span><span class="va">eigenvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-28-1.png" alt="Bar plot of eigenvalues against the two discriminant axes" width="700"></p>
<p>We can plot the scores with:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-29-1.png" alt="Scatterplot of the scores of each individual on the two discriminant axes (defined by `ld`)" width="700"></p>
<p>We can inspect the assignment by DAPC with <code>autoplot</code>
using the type <code>components</code>, ordering the samples by the
original population labels:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, type<span class="op">=</span><span class="st">"components"</span>, group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-30-1.png" alt="Bar plot showing the probability of assignment to each cluster" width="700"></p>
<p>Because of the very clear separation we observed when plotting the LD
scores, no individual is modelled as a mixture: all assignments are with
100% probability to a single cluster.</p>
<p>Finally, we can explore which loci have the biggest impact on
separating the clusters (either because of drift or selection):</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_dapc</span>, <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-31-1.png" alt="Plot of the loadings of Principal Component 1 for each loci" width="700"></p>
<p>There is no strong outlier, suggesting drift across many loci has
created the signal picked up by DAPC.</p>
<p>Note that <code>anole_dapc</code> is of class <code>gt_dapc</code>,
which is a subclass of <code>dapc</code> from <code>adegenet</code>.
This means that functions written to work on <code>dapc</code> objects
should work out of the box (the only exception is
<code><a href="https://rdrr.io/pkg/adegenet/man/dapc.html" class="external-link">adegenet::predict.dapc</a></code>, which does not work because the
underlying pca object is different). For example, we can obtain the
standard dapc plot with:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ade4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    /// adegenet 2.1.11 is loaded ////////////</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    &gt; overview: '?adegenet'</span></span>
<span><span class="co">#&gt;    &gt; tutorials/doc/questions: 'adegenetWeb()' </span></span>
<span><span class="co">#&gt;    &gt; bug reports/feature requests: adegenetIssues()</span></span>
<span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/scatter.html" class="external-link">scatter</a></span><span class="op">(</span><span class="va">anole_dapc</span>, posi.da<span class="op">=</span><span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-32-1.png" alt="Scatterplot of individuals with ellipses surrounding each cluster of individuals identified by the DAPC analysis. Eigenvalues are displayed in an inset in the bottom right corner of the plot." width="700"></p>
</div>
<div class="section level2">
<h2 id="clustering-with-snmf">Clustering with sNMF<a class="anchor" aria-label="anchor" href="#clustering-with-snmf"></a>
</h2>
<p>sNMF is a fast clustering algorithm which provides results similar to
STRUCTURE and ADMIXTURE. We can run it directly from tidypopgen using
the <code>gt_snmf</code> function.</p>
<p>The function <code>gt_snmf</code> creates a .geno file to be used by
LEA. This is placed by default in the same directory as the backing file
of the <code>gen_tibble</code>, and uses the same name.</p>
<p>The output of <code>gt_snmf</code> is a <code>gt_admix</code> object.
<code>gt_admix</code> objects are designed to make visualising data from
different clustering algorithms swift and easy.</p>
<p>Lets begin by running <code>gt_snmf</code>. We can run K clusters
from k=1 to k=10. We will use just one repeat, but ideally we should run
multiple repetitions for each K:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anole_snmf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_snmf.html">gt_snmf</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">anole_gt</span>,</span>
<span>                         project <span class="op">=</span> <span class="st">"new"</span>,</span>
<span>                         k <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                         entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         n_runs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                         alpha <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                         seed <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can examine the suitability of our K values by plotting the
cross-entropy values for each K value, which is contained in the
<code>cv</code> element of the <code>gt_admix</code> object:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anole_snmf</span><span class="op">$</span><span class="va">cv</span>, cex <span class="op">=</span> <span class="fl">1.2</span>, pch <span class="op">=</span> <span class="fl">19</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-34-1.png" alt="Plot of cross-entropy for each value of K, showing K = 3 has the lowest cross-entropy" width="700"></p>
<p>Here we can see that K = 3 is a sensible choice, as K = 3 represents
the ‘elbow’ in the plot, where cross-entropy is at it’s lowest.</p>
<p>We can quickly plot this data with <code>autoplot</code> by using the
type <code>barplot</code> and selecting our chosen value of
<code>k</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_snmf</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">FALSE</span>, type <span class="op">=</span> <span class="st">"barplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-35-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>This plot is fine, but it is messy, and not a very helpful
visualisation if we want to understand how our populations are
structured.</p>
<p>To re-order this plot, grouping individuals by population, we can use
the <code>annotate_group</code> and <code>arrange_by_group</code>
arguments in autoplot:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_snmf</span>, type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-36-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>This plot is much better. Now the plot is labelled by population we
can see that the individuals from each population are grouped together,
with a few individuals showing mixed ancestry between populations.</p>
<p>However, for a more visually appealing plot, we can re-order the
individuals within each population as follows:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_snmf</span>, type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-37-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources, with individuals arranged according to their dominant ancestry propotion" width="700"></p>
<p>Now individuals are neatly ordered within each population.</p>
<p>For more complex/publication ready plots, we can extract the specific
K value and run that we are interested in plotting, and add this .Q
matrix to our gen_tibble object, so that we can create a custom plot
with <code>ggplot2</code>.</p>
<p>First we extract the q_matrix of interest using the
<code>get_q_matrix</code> function:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">anole_snmf</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Then we can easily add the data with the <code>augment</code>
method:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_sNMF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span><span class="va">q_mat</span>, data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anole_gt_sNMF</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 38</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   population [3]</span></span></span>
<span><span class="co">#&gt;   id            genotypes population .rownames  .fittedPC1 .fittedPC2 .fittedPC3</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> BM288                 1 Eam        BM288            20.8      40.7     -<span style="color: #BB0000;">14.9</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GN71                  2 Eam        GN71             19.5      44.8     -<span style="color: #BB0000;">20.2</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> H1907                 3 Wam        H1907           -<span style="color: #BB0000;">91.2</span>     -<span style="color: #BB0000;">11.3</span>       8.70 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> H1911                 4 Wam        H1911          -<span style="color: #BB0000;">101.</span>      -<span style="color: #BB0000;">18.2</span>     -<span style="color: #BB0000;">30.4</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> H2546                 5 Wam        H2546           -<span style="color: #BB0000;">88.9</span>      -<span style="color: #BB0000;">9.65</span>     12.8  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> IBSPCRIB0361          6 AF         IBSPCRIB0…       43.7     -<span style="color: #BB0000;">30.4</span>       0.908</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 31 more variables: .fittedPC4 &lt;dbl&gt;, .fittedPC5 &lt;dbl&gt;, .fittedPC6 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC7 &lt;dbl&gt;, .fittedPC8 &lt;dbl&gt;, .fittedPC9 &lt;dbl&gt;, .fittedPC10 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC11 &lt;dbl&gt;, .fittedPC12 &lt;dbl&gt;, .fittedPC13 &lt;dbl&gt;, .fittedPC14 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC15 &lt;dbl&gt;, .fittedPC16 &lt;dbl&gt;, .fittedPC17 &lt;dbl&gt;, .fittedPC18 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC19 &lt;dbl&gt;, .fittedPC20 &lt;dbl&gt;, .fittedPC21 &lt;dbl&gt;, .fittedPC22 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC23 &lt;dbl&gt;, .fittedPC24 &lt;dbl&gt;, .fittedPC25 &lt;dbl&gt;, .fittedPC26 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .fittedPC27 &lt;dbl&gt;, .fittedPC28 &lt;dbl&gt;, .fittedPC29 &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>Alternatively, we can convert our q matrix into <code>tidy</code>
format, which is more suitable for plotting:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">q_mat</span>, <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tidy_q</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   id    group q     percentage</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> BM288 Eam   .Q1     1.00    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> BM288 Eam   .Q2     0.000<span style="text-decoration: underline;">100</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> BM288 Eam   .Q3     0.000<span style="text-decoration: underline;">100</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GN71  Eam   .Q1     1.00    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GN71  Eam   .Q2     0.000<span style="text-decoration: underline;">100</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> GN71  Eam   .Q3     0.000<span style="text-decoration: underline;">100</span></span></span></code></pre></div>
<p>And now we can use <code>ggplot2</code> directly to generate our
custom plot:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tidy_q</span> <span class="op">&lt;-</span> <span class="va">tidy_q</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dominant_q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">percentage</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">group</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">dominant_q</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>plot_order <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span>, id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">id</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_q</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">id</span>, y <span class="op">=</span> <span class="va">percentage</span>, fill <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span>, position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>reverse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Population Structure for K = 3"</span>, title <span class="op">=</span> <span class="st">"Clustering at K = 3 for Anolis punctatus using sNMF algorithm"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_distruct.html">theme_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-41-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>In the Prates et al 2018 data, anolis lizards were sampled from three
regions, described as Af, Eam, and Wam. These acronyms correspond to
samples from the Atlantic Forest, Eastern Amazonia, and Western Amazonia
of Brazil.</p>
<p>Lets say we wanted to change the grouping variable of our plot, to
explore how Atlantic Forest individuals differ from all Amazonian
individuals. We can do this by adding a <code>forest</code> group to our
gen_tibble:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span><span class="op">$</span><span class="va">forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span> <span class="op">==</span> <span class="st">"AF"</span>, <span class="st">"Atlantic Forest"</span>, <span class="st">"Amazonian Forest"</span><span class="op">)</span></span></code></pre></div>
<p>Then we can change the grouping variable of our <code>gt_admix</code>
object using <code>gt_admix_reorder_q</code>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_snmf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_admix_reorder_q.html">gt_admix_reorder_q</a></span><span class="op">(</span><span class="va">anole_snmf</span>, group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">forest</span><span class="op">)</span></span></code></pre></div>
<p>And replot the data:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">anole_snmf</span>, type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>, annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-44-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources, labelled Atlantic forest or Amazonian forest" width="700"></p>
<p>And here we can see that the Atlantic Forest individuals clearly
separate from Amazonian individuals. However, there are two clear
clusters among the Amazonian individuals, showing that there is still a
strong distinction between Eastern and Western anolis lizards within the
Amazonian forest.</p>
</div>
<div class="section level2">
<h2 id="handling-q-matrices">Handling Q matrices<a class="anchor" aria-label="anchor" href="#handling-q-matrices"></a>
</h2>
<p>For clustering algorithms that operate outside the R environment(such
as ADMIXTURE), the function <code><a href="../reference/q_matrix.html">q_matrix()</a></code> can take a path to a
directory containing results from multiple runs of a clustering
algorithm for multiple values of K, read the .Q files, and summarise
them in a single <code>q_matrix_list</code> object.</p>
<p>In the analysis above, through <code><a href="../reference/gt_snmf.html">gt_snmf()</a></code>, we ran 1
repetition for K values 1 to 10.</p>
<p>To exemplify how <code><a href="../reference/q_matrix.html">q_matrix()</a></code> can read directly from
output folders, we can write the outputs to a temporary file:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">k_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">q_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="va">k_values</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">q_matrices</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">runs</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">q_matrices</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">anole_snmf</span>, k <span class="op">=</span> <span class="va">x</span>, run <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">q_matrices</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir</span>,<span class="st">"/K"</span>,<span class="va">x</span>,<span class="st">"run"</span>,<span class="va">i</span>,<span class="st">".Q"</span><span class="op">)</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, quote <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>And read them back into our R environment using
<code>q_matrix_list</code>:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_q_files.html">read_q_files</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">q_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; Admixture results for multiple runs:                      </span></span>
<span><span class="co">#&gt; k 1 2 3 4 5 6 7 8 9 10</span></span>
<span><span class="co">#&gt; n 1 1 1 1 1 1 1 1 1  1</span></span>
<span><span class="co">#&gt; with slots:</span></span>
<span><span class="co">#&gt; $Q for Q matrices</span></span></code></pre></div>
<p><code>q_matrix_list</code> has read and summarised the .Q files from
our analysis, and we can access a single matrix from the list by
selecting the run number and K value of interest using
<code>get_q_matrix</code> as before.</p>
<p>For example, if we would like to view the second run of K = 3:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">q_list</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;              .Q1         .Q2         .Q3</span></span>
<span><span class="co">#&gt;  [1,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt;  [2,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt;  [3,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt;  [4,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt;  [5,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt;  [6,] 0.22744500 2.24013e-01 0.548541000</span></span>
<span><span class="co">#&gt;  [7,] 0.16160000 2.11511e-02 0.817248000</span></span>
<span><span class="co">#&gt;  [8,] 0.76707500 9.99910e-05 0.232825000</span></span>
<span><span class="co">#&gt;  [9,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt; [10,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt; [11,] 0.00468650 9.99911e-05 0.995214000</span></span>
<span><span class="co">#&gt; [12,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt; [13,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt; [14,] 0.77902900 1.90400e-01 0.030570500</span></span>
<span><span class="co">#&gt; [15,] 0.98083300 1.90675e-02 0.000099991</span></span>
<span><span class="co">#&gt; [16,] 0.62619600 3.63449e-01 0.010354700</span></span>
<span><span class="co">#&gt; [17,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt; [18,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt; [19,] 0.99980000 9.99900e-05 0.000099990</span></span>
<span><span class="co">#&gt; [20,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt; [21,] 0.96162000 3.82797e-02 0.000099991</span></span>
<span><span class="co">#&gt; [22,] 0.77225200 1.48102e-01 0.079646000</span></span>
<span><span class="co">#&gt; [23,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt; [24,] 0.97051600 2.93841e-02 0.000099991</span></span>
<span><span class="co">#&gt; [25,] 0.78666100 9.99910e-05 0.213239000</span></span>
<span><span class="co">#&gt; [26,] 0.74774700 9.99910e-05 0.252153000</span></span>
<span><span class="co">#&gt; [27,] 0.97680100 2.30986e-02 0.000099991</span></span>
<span><span class="co">#&gt; [28,] 0.66003700 3.39863e-01 0.000099991</span></span>
<span><span class="co">#&gt; [29,] 0.10954200 4.69100e-02 0.843548000</span></span>
<span><span class="co">#&gt; [30,] 0.00009999 9.99900e-05 0.999800000</span></span>
<span><span class="co">#&gt; [31,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [32,] 0.18031500 8.19585e-01 0.000099991</span></span>
<span><span class="co">#&gt; [33,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [34,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [35,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [36,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [37,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [38,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [39,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [40,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [41,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [42,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [43,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [44,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [45,] 0.00009999 9.99800e-01 0.000099990</span></span>
<span><span class="co">#&gt; [46,] 0.18782000 8.12080e-01 0.000099991</span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "q_matrix" "q_matrix" "matrix"   "array"</span></span></code></pre></div>
<p>And, again, we can then autoplot any matrix by selecting from the
<code>q_matrix_list</code> object:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_q_matrix.html">get_q_matrix</a></span><span class="op">(</span><span class="va">q_list</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">anole_gt</span>, annotate_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_group <span class="op">=</span> <span class="cn">TRUE</span>, arrange_by_indiv <span class="op">=</span> <span class="cn">TRUE</span>, reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a03_example_clustering_and_dapc_files/figure-html/unnamed-chunk-48-1.png" alt="Barplot of individuals coloured by predicted ancestry proportion (Q) from each of K ancestral sources" width="700"></p>
<p>In this way, <code>tidypopgen</code> integrates with external
clustering software (such as ADMIXTURE or STRUCTURE) seamlessly for
quick, easy plotting.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
