<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PLINK cheatsheet • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PLINK cheatsheet">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
    <li><a class="dropdown-item" href="../articles/interpolating_hgdp.html">Interpolating Heterozygosity</a></li>
    <li><a class="dropdown-item" href="../articles/interpolating.html">Interpolating genetic data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>PLINK cheatsheet</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/dev/vignettes/a99_plink_cheatsheet.Rmd" class="external-link"><code>vignettes/a99_plink_cheatsheet.Rmd</code></a></small>
      <div class="d-none name"><code>a99_plink_cheatsheet.Rmd</code></div>
    </div>

    
    
<p>The following flags and corresponding tidypopgen functions are based
on plink version 1.9.</p>
<div class="section level2">
<h2 id="file-management-and-reading-data">File management and reading data:<a class="anchor" aria-label="anchor" href="#file-management-and-reading-data"></a>
</h2>
<table class="table">
<colgroup>
<col width="36%">
<col width="63%">
</colgroup>
<thead><tr class="header">
<th>PLINK 1.9</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–make-bed –out</td>
<td><code>gt_as_plink(data, file = my_file, type = "bed")</code></td>
</tr>
<tr class="even">
<td>–recode</td>
<td><code>gt_as_plink(data, file = my_file, type = "ped")</code></td>
</tr>
<tr class="odd">
<td>–recode vcf</td>
<td><code>gt_as_vcf(data, file = my_file)</code></td>
</tr>
<tr class="even">
<td>–allele1234 and –alleleACGT</td>
<td>See <code><a href="../reference/gen_tibble.html">gen_tibble()</a></code> parameter ‘valid_alleles’</td>
</tr>
</tbody>
</table>
<p>PLINK flags –update-alleles, –allele1234, and –alleleACGT, all alter
the coding of alleles. In <code>tidypopgen</code>, valid alleles are
supplied when reading in a <code>gen_tibble</code>.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control:<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–maf</td>
<td><code>data %&gt;% loci_maf()</code></td>
</tr>
<tr class="even">
<td>–geno</td>
<td><code>data %&gt;% loci_missingness()</code></td>
</tr>
<tr class="odd">
<td>–hwe</td>
<td><code>data %&gt;% loci_hwe()</code></td>
</tr>
<tr class="even">
<td>–freq –a2-allele</td>
<td><code>data %&gt;% loci_alt_freq()</code></td>
</tr>
<tr class="odd">
<td>–mind</td>
<td><code>data %&gt;% indiv_missingness()</code></td>
</tr>
<tr class="even">
<td>–het</td>
<td><code>data %&gt;% indiv_het_obs()</code></td>
</tr>
</tbody>
</table>
<p>To filter out variants in <code>tidypopgen</code>, in a similar way
to PLINK flags such as –extract or –autosome, it is necessary to use the
<code>gen_tibble</code> with <code><a href="../reference/select_loci_if.html">select_loci_if()</a></code>. For
example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_chromosomes.html">loci_chromosomes</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 15 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 4</span></span></span>
<span><span class="co">#&gt;   id    population sex    genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> GRC24 pop_a      male   [0,0,...]</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GRC25 pop_a      male   [0,0,...]</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> GRC26 pop_a      male   [0,1,...]</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GRC27 pop_a      male   [0,0,...]</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GRC28 pop_a      male   [0,0,...]</span></span></code></pre></div>
<p>will select autosomal loci in the same way as –autosome. Or
alternatively:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs4477212"</span>, <span class="st">"rs3094315"</span>, <span class="st">"rs3131972"</span>, <span class="st">"rs12124819"</span>, <span class="st">"rs11240777"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_names.html">loci_names</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">my_snps</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name       chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         1 rs4477212  1             <span style="text-decoration: underline;">82</span>154            0 A          <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>         2 rs3094315  1            <span style="text-decoration: underline;">752</span>566            0 A          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>         3 rs3131972  1            <span style="text-decoration: underline;">752</span>721            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>         4 rs12124819 1            <span style="text-decoration: underline;">776</span>546            0 A          <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>         5 rs11240777 1            <span style="text-decoration: underline;">798</span>959            0 G          A</span></span></code></pre></div>
<p>will select loci from a previously defined set in the same way as
–extract.</p>
<p>Similarly, to filter out individuals, as might be performed with
–keep in PLINK, requires using filter:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GRC14300079"</span>, <span class="st">"GRC14300142"</span>, <span class="st">"GRC14300159"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">my_individuals</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 16 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     0 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 variables: id &lt;chr&gt;, population &lt;chr&gt;, sex &lt;fct&gt;, genotypes &lt;vctr_SNP&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="handling-linkage">Handling linkage<a class="anchor" aria-label="anchor" href="#handling-linkage"></a>
</h2>
<p>Linkage disequilibrium is managed through clumping in
<code>tidypopgen</code> with <code><a href="../reference/loci_ld_clump.html">loci_ld_clump()</a></code>.</p>
<p>This option is similar to the –indep-pairwise flag in PLINK, but
results in a more even distribution of loci when compared to LD
pruning.</p>
<p>To explore why clumping is preferable to pruning, see <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="external-link uri">https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</a></p>
</div>
<div class="section level2">
<h2 id="quality-control-for-relatedness-king">Quality control for relatedness (KING)<a class="anchor" aria-label="anchor" href="#quality-control-for-relatedness-king"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>KING</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–kinship</td>
<td><code><a href="../reference/pairwise_king.html">pairwise_king()</a></code></td>
</tr>
<tr class="even">
<td>–distance</td>
<td><code><a href="../reference/pairwise_ibs.html">pairwise_ibs()</a></code></td>
</tr>
<tr class="odd">
<td>–unrelated</td>
<td><code><a href="../reference/filter_high_relatedness.html">filter_high_relatedness()</a></code></td>
</tr>
</tbody>
</table>
<p><code><a href="../reference/pairwise_king.html">pairwise_king()</a></code> implements the KING-robust estimator of
kinship, equivalent to –kinship in KING. To remove related individuals,
the user can pass the kinship matrix of <code><a href="../reference/pairwise_king.html">pairwise_king()</a></code> and
a relatedness threshold (a numeric KING kinship coefficient) to
<code><a href="../reference/filter_high_relatedness.html">filter_high_relatedness()</a></code>, which will return the largest
possible set of individuals with no relationships above the
threshold.</p>
<p><code><a href="../reference/pairwise_king.html">pairwise_king()</a></code> also forms part of
<code><a href="../reference/qc_report_indiv.html">qc_report_indiv()</a></code> through the parameter
<code>kings_threshold</code>. To remove related individuals in
<code><a href="../reference/qc_report_indiv.html">qc_report_indiv()</a></code>, the user can pass either a relatedness
threshold, or a string of either “first” or “second” to remove any first
degree or second degree relationships from the dataset. This second
option is similar to using –unrelated –degree 1 or –unrelated –degree 2
in KING. <code><a href="../reference/qc_report_indiv.html">qc_report_indiv()</a></code> returns a dataframe including
columns ‘id’ and ‘to_keep’, showing the ID of individuals and a logical
column of whether the individual should be removed to retain the largest
possible set of individuals with no relationships above the
threshold.</p>
</div>
<div class="section level2">
<h2 id="merging-datasets">Merging datasets:<a class="anchor" aria-label="anchor" href="#merging-datasets"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–bmerge</td>
<td><code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code></td>
</tr>
<tr class="even">
<td>–flip-scan</td>
<td><code><a href="../reference/rbind_dry_run.html">rbind_dry_run()</a></code></td>
</tr>
<tr class="odd">
<td>–flip</td>
<td>use ‘flip_strand = TRUE’ in <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code>
</td>
</tr>
</tbody>
</table>
<p>In PLINK, data merging can fail due to strand inconsistencies that
are not addressed prior to merging. PLINK documentation suggests to
users to try a ‘trial flip’ of data to address this, and then to
‘unflip’ any errors that remain. In <code>tidypopgen</code>, when data
are merged with rbind, strand inconsistencies are identified and
automatically flipped, avoiding multiple rounds of flipping before
merging.</p>
<p>PLINK does allow users to identify inconsistencies prior to merging
with –flip-scan, and this functionality is included in the
<code>tidypopgen</code> rbind_dry_run(). rbind_dry_run() reports the
numeric overlap of datasets, alongside the number of SNPs to ‘flip’ in
the new target dataset, as well as the number of ambiguous SNPs.</p>
<p>Data are only merged one set at a time, there is no equivalent to
–merge-list.</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis:<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–pca</td>
<td>See <code><a href="../reference/gt_pca.html">gt_pca()</a></code> for pca options</td>
</tr>
<tr class="even">
<td>–fst</td>
<td>
<code><a href="../reference/pairwise_pop_fst.html">pairwise_pop_fst()</a></code> with <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code>
</td>
</tr>
<tr class="odd">
<td>–homozyg</td>
<td><code><a href="../reference/windows_indiv_roh.html">windows_indiv_roh()</a></code></td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
