<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>plink cheatsheet • tidypopgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="plink cheatsheet">
<meta property="og:description" content="tidypopgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidypopgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9014</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_overview.html">overview</a>
    </li>
    <li>
      <a href="../articles/a02_qc.html">QC</a>
    </li>
    <li>
      <a href="../articles/a03_example_clustering_and_dapc.html">example workflow with tidypopgen</a>
    </li>
    <li>
      <a href="../articles/a99_plink_cheatsheet.html">plink cheatsheet</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>plink cheatsheet</h1>
            
      
      
      <div class="hidden name"><code>a99_plink_cheatsheet.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="file-management-and-reading-data">File management and reading data:<a class="anchor" aria-label="anchor" href="#file-management-and-reading-data"></a>
</h2>
<table class="table">
<colgroup>
<col width="43%">
<col width="56%">
</colgroup>
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–make-bed –out</td>
<td><code>gt_as_plink(data, bedfile = my_file)</code></td>
</tr>
<tr class="even">
<td>–recode</td>
<td>`gt_as_plink(data, bedfile = my_file</td>
</tr>
<tr class="odd">
<td>–allele1234 and –alleleACGT</td>
<td>See <code>gen_tibble()\</code> parameter ‘valid_alleles’</td>
</tr>
</tbody>
</table>
<p>PLINK flags –update-alleles, –allele1234, and –alleleACGT, all alter
the coding of alleles. In <code>tidypopgen</code>, valid alleles are
supplied when reading in a <code>gen_tibble</code>.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control:<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–maf</td>
<td><code>data %\&gt;% loci_maf()</code></td>
</tr>
<tr class="even">
<td>–geno</td>
<td><code>data %\&gt;% loci_missingness()</code></td>
</tr>
<tr class="odd">
<td>–mind</td>
<td><code>data %\&gt;% indiv_missingness()</code></td>
</tr>
<tr class="even">
<td>–hwe</td>
<td><code>data %\&gt;% loci_hwe()</code></td>
</tr>
<tr class="odd">
<td>–het</td>
<td><code>data %\&gt;% indiv_het_obs()</code></td>
</tr>
<tr class="even">
<td>–distance</td>
<td><code>data %\&gt;% gt_ibs()</code></td>
</tr>
</tbody>
</table>
<p>To filter out variants in <code>tidypopgen</code>, in a similar way
to PLINK flags such as –extract or –autosome, it is necessary to use the
<code>gen_tibble</code> with <code><a href="../reference/select_loci_if.html">select_loci_if()</a></code>. For
example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_chromosomes.html">loci_chromosomes</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 15 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     5 × 3</span></span></span>
<span><span class="co">#&gt;   id    population  genotypes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> GRC24 pop_a               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GRC25 pop_a               2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> GRC26 pop_a               3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GRC27 pop_a               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GRC28 pop_a               5</span></span></code></pre></div>
<p>will select autosomal loci in the same way as –autosome. Or
alternatively:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs4477212"</span>,<span class="st">"rs3094315"</span>,<span class="st">"rs3131972"</span>,<span class="st">"rs12124819"</span>,<span class="st">"rs11240777"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_names.html">loci_names</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">my_snps</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 7</span></span></span>
<span><span class="co">#&gt;   big_index name       chromosome position genetic_dist allele_ref allele_alt</span></span>
<span><span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         1 rs4477212           1    <span style="text-decoration: underline;">82</span>154            0 A          <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>         2 rs3094315           1   <span style="text-decoration: underline;">752</span>566            0 A          G         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>         3 rs3131972           1   <span style="text-decoration: underline;">752</span>721            0 G          A         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>         4 rs12124819          1   <span style="text-decoration: underline;">776</span>546            0 A          <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>         5 rs11240777          1   <span style="text-decoration: underline;">798</span>959            0 G          A</span></span></code></pre></div>
<p>will select loci from a previously defined set in the same way as
–extract.</p>
<p>Similarly, to filter out individuals, as might be performed with
–keep in PLINK, requires using filter:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GRC14300079"</span>, <span class="st">"GRC14300142"</span>, <span class="st">"GRC14300159"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">my_individuals</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A gen_tibble: 16 loci</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble:     0 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 variables: id &lt;chr&gt;, population &lt;chr&gt;, genotypes &lt;vctr_SNP&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quality-control-for-linkage">Quality control for linkage<a class="anchor" aria-label="anchor" href="#quality-control-for-linkage"></a>
</h2>
<p>Linkage is managed through clumping in <code>tidypopgen</code> with
loci_ld_clump().</p>
<p>This option is similar to the –indep-pairwise flag in PLINK, but
results in a more even distribution of loci when compared to LD
pruning.</p>
<p>To explore why clumping is preferable to pruning, see <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="external-link uri">https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</a></p>
</div>
<div class="section level2">
<h2 id="quality-control-for-relatedness-king">Quality control for relatedness (KING)<a class="anchor" aria-label="anchor" href="#quality-control-for-relatedness-king"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>KING</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–kinship</td>
<td><code><a href="../reference/gt_king.html">gt_king()</a></code></td>
</tr>
<tr class="even">
<td>–unrelated</td>
<td><code><a href="../reference/filter_high_relatedness.html">filter_high_relatedness()</a></code></td>
</tr>
</tbody>
</table>
<p>gt_king() implements the KING-robust estimator of kinship equivalent
to –kinship in KING. To remove related individuals, instead of using
–unrelated –degree, the user can pass the resulting kinship matrix and a
relatedness threshold to filter_high_relatedness, which will return the
largest possible set of unrelated individuals.</p>
</div>
<div class="section level2">
<h2 id="merging-datasets">Merging datasets:<a class="anchor" aria-label="anchor" href="#merging-datasets"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–bmerge</td>
<td><code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code></td>
</tr>
<tr class="even">
<td>–flip-scan</td>
<td><code><a href="../reference/rbind_dry_run.html">rbind_dry_run()</a></code></td>
</tr>
<tr class="odd">
<td>–flip</td>
<td>use ‘flip_strand = TRUE’ in <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code>
</td>
</tr>
</tbody>
</table>
<p>In PLINK, data merging can fail due to strand inconsistencies that
are not addressed prior to merging. PLINK documentation suggests to
users to try a ‘trial flip’ of data to address this, and then to
‘unflip’ any errors that remain. In <code>tidypopgen</code>, when data
are merged with rbind, strand inconsistencies are identified and
automatically flipped, avoiding multiple rounds of flipping before
merging.</p>
<p>PLINK does allow users to identify inconsistencies prior to merging
with –flip-scan, and this functionality is included in the
<code>tidypopgen</code> rbind_dry_run(). rbind_dry_run() reports the
numeric overlap of datasets, alongside the number of SNPs to ‘flip’ in
the new target dataset, as well as the number of ambiguous SNPs.</p>
<p>Data are only merged one set at a time, there is no equivalent to
–merge-list.</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis:<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>–pca</td>
<td>See gt_pca() for pca options</td>
</tr>
<tr class="even">
<td>–fst</td>
<td>pairwise_pop_fst() with group_by()</td>
</tr>
<tr class="odd">
<td>–homozyg</td>
<td>gt_roh_window()</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Evie Carter, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
