<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Working with aDNA pseuodhaploid samples • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working with aDNA pseuodhaploid samples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Working with aDNA pseuodhaploid samples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/vignettes/articles/aDNA_pseudohaploids.Rmd" class="external-link"><code>vignettes/articles/aDNA_pseudohaploids.Rmd</code></a></small>
      <div class="d-none name"><code>aDNA_pseudohaploids.Rmd</code></div>
    </div>

    
    
<p>The following article shows how to handle ancient DNA in tidypopgen,
using example data from Lazaridis et al. (2016). First, we will
demonstrate how to handle pseudohaploid data. Then, we will show how to
project ancient DNA data onto modern data in a Principal Components
Analysis, replicating the workflow illustrated in the vignette for the R
package <code>smartsnp</code>. Finally, we will replicate the pairwise
Fst analysis from Lazaridis et al. (2016).</p>
<p>First, we can download the data from the Reich lab website into a
temporary directory, and unzip the file.</p>
<div class="section level2">
<h2 id="download-the-data">Download the data<a class="anchor" aria-label="anchor" href="#download-the-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">download_url</span> <span class="op">&lt;-</span> <span class="st">"https://reich.hms.harvard.edu/sites/reich.hms.harvard.edu/files/inline-files/NearEastPublic.tar.gz"</span></span>
<span><span class="va">download_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"NearEastPublic.tar.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">download_url</span>, <span class="va">download_path</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="va">download_path</span>, exdir <span class="op">=</span> <span class="va">temp_dir</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-in-the-data">Read in the data<a class="anchor" aria-label="anchor" href="#read-in-the-data"></a>
</h2>
<p>Now that our data are downloaded, we can use <code>tidypopgen</code>
to read the data into <code>gen_tibble</code> objects, beginning with
the modern data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ho_modern</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="st">"./data/NearEastPublic/HumanOriginsPublic2068.geno"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"test_"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Followed by the ancient data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="st">"./data/NearEastPublic/AncientLazaridis2016.geno"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"test_"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="assigning-pseudohaploid-data">Assigning Pseudohaploid data<a class="anchor" aria-label="anchor" href="#assigning-pseudohaploid-data"></a>
</h2>
<p>Due to the damage associated with ancient DNA samples, genotype
datasets of ancient individuals often include pseudohaploid data.
Instead of diploid genotypes, low coverage samples are represented by a
sampling one allele per site. tidypopgen handles gen_tibbles containing
pseudohaploid data by assigning them a specific genotype code (-2) to
denote that some individuals are pseudohaplodified. The ploidy of each
individual is then stored as 1 (denoting pseudohaploid) or 2 (a diploid
individual).</p>
<p>Before we proceed with any analysis, we can therefore use the
function <code>gt_pseudohaploid</code> to assign the ploidy of each
individual in our <code>gen_tibble</code> object (based on whether the
first <code>test_n_loci</code> are all homozygote):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_pseudohaploid.html">gt_pseudohaploid</a></span><span class="op">(</span><span class="va">ancient</span>, test_n_loci <span class="op">=</span> <span class="fl">100000</span><span class="op">)</span></span></code></pre></div>
<p>Once the ploidy has been assigned, pseudohaploid data is handled
automatically by functions that are designed to work with pseudohaploid
data. Some functions are meaningless to use on pseudohaplodified data,
and these functions will return an error.</p>
<p>For example, attempting to run <code>indiv_inbreeding</code> on
pseudohaploid data will tell us that this function only works on diploid
data:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/indiv_inbreeding.html">indiv_inbreeding</a></span><span class="op">(</span><span class="va">ancient</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in stopifnot_diploid(.x): this function only works on diploid data</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>For our PCA analysis, we will be projecting ancient European
individuals. Therefore, we need to subset the modern data to only modern
West Eurasian populations. We can select these populations by filtering
the <code>gen_tibble</code> object using the <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter()</a></code>
function from the <code>dplyr</code> package:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">west_eurasian_pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Abkhasian"</span>, <span class="st">"Adygei"</span>, <span class="st">"Albanian"</span>, <span class="st">"Armenian"</span>, <span class="st">"Assyrian"</span>, <span class="st">"Balkar"</span>, <span class="st">"Basque"</span>,</span>
<span>  <span class="st">"BedouinA"</span>, <span class="st">"BedouinB"</span>, <span class="st">"Belarusian"</span>, <span class="st">"Bulgarian"</span>, <span class="st">"Canary_Islander"</span>,</span>
<span>  <span class="st">"Chechen"</span>, <span class="st">"Croatian"</span>, <span class="st">"Cypriot"</span>, <span class="st">"Czech"</span>, <span class="st">"Druze"</span>, <span class="st">"English"</span>, <span class="st">"Estonian"</span>,</span>
<span>  <span class="st">"Finnish"</span>, <span class="st">"French"</span>, <span class="st">"Georgian"</span>, <span class="st">"German"</span>, <span class="st">"Greek"</span>, <span class="st">"Hungarian"</span>, <span class="st">"Icelandic"</span>,</span>
<span>  <span class="st">"Iranian"</span>, <span class="st">"Irish"</span>, <span class="st">"Irish_Ulster"</span>, <span class="st">"Italian_North"</span>, <span class="st">"Italian_South"</span>,</span>
<span>  <span class="st">"Jew_Ashkenazi"</span>, <span class="st">"Jew_Georgian"</span>, <span class="st">"Jew_Iranian"</span>, <span class="st">"Jew_Iraqi"</span>, <span class="st">"Jew_Libyan"</span>,</span>
<span>  <span class="st">"Jew_Moroccan"</span>, <span class="st">"Jew_Tunisian"</span>, <span class="st">"Jew_Turkish"</span>, <span class="st">"Jew_Yemenite"</span>, <span class="st">"Jordanian"</span>,</span>
<span>  <span class="st">"Kumyk"</span>, <span class="st">"Lebanese_Christian"</span>, <span class="st">"Lebanese"</span>, <span class="st">"Lebanese_Muslim"</span>, <span class="st">"Lezgin"</span>,</span>
<span>  <span class="st">"Lithuanian"</span>, <span class="st">"Maltese"</span>, <span class="st">"Mordovian"</span>, <span class="st">"North_Ossetian"</span>, <span class="st">"Norwegian"</span>,</span>
<span>  <span class="st">"Orcadian"</span>, <span class="st">"Palestinian"</span>, <span class="st">"Polish"</span>, <span class="st">"Romanian"</span>, <span class="st">"Russian"</span>, <span class="st">"Sardinian"</span>,</span>
<span>  <span class="st">"Saudi"</span>, <span class="st">"Scottish"</span>, <span class="st">"Shetlandic"</span>, <span class="st">"Sicilian"</span>, <span class="st">"Sorb"</span>, <span class="st">"Spanish_North"</span>,</span>
<span>  <span class="st">"Spanish"</span>, <span class="st">"Syrian"</span>, <span class="st">"Turkish"</span>, <span class="st">"Ukrainian"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ho_modern</span> <span class="op">&lt;-</span> <span class="va">ho_modern</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">population</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">west_eurasian_pops</span><span class="op">)</span></span></code></pre></div>
<p>To run a PCA on the modern Eurasian individuals, we need to impute
any missing genotypes. But as we have subset our <code>gen_tibble</code>
object, first we need to update our backingfile:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ho_modern</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_update_backingfile.html">gt_update_backingfile</a></span><span class="op">(</span><span class="va">ho_modern</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And then we can impute, using:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ho_modern</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">ho_modern</span>, method <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span></code></pre></div>
<p>We should also remove any monomorphic loci from our data, as these
loci are uninformative. Monomorphic markers must always be removed
before a PCA is computed using any of the <code>gt_pca</code>
functions.</p>
<p>To do this, we use the <code>select_loci_if</code> function together
with <code>loci_maf</code> to select only genotypes where MAF is greater
than 0:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ho_modern</span> <span class="op">&lt;-</span> <span class="va">ho_modern</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_alt_freq.html">loci_maf</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Now we can create our PCA object from the modern data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modern_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span><span class="va">ho_modern</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>We can then use <code>augment</code> to add the PCA coordinates for
each individual to our <code>gen_tibble</code> object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modern_pca_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">modern_pca</span>, data <span class="op">=</span> <span class="va">ho_modern</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>And <code>tidy</code> to extract the proportion of explained
variance:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_variance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">modern_pca</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="projection">Projection<a class="anchor" aria-label="anchor" href="#projection"></a>
</h2>
<p>Before projecting the ancient samples, we remove ancient individuals
and outgroups that are not of interest:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Samples to remove:</span></span>
<span><span class="va">sample_remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Mota"</span>, <span class="st">"Denisovan"</span>, <span class="st">"Chimp"</span>, <span class="st">"Mbuti.DG"</span>, <span class="st">"Altai"</span>,</span>
<span>  <span class="st">"Vi_merge"</span>, <span class="st">"Clovis"</span>, <span class="st">"Kennewick"</span>, <span class="st">"Chuvash"</span>, <span class="st">"Ust_Ishim"</span>,</span>
<span>  <span class="st">"AG2"</span>, <span class="st">"MA1"</span>, <span class="st">"MezE"</span>, <span class="st">"hg19ref"</span>, <span class="st">"Kostenki14"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ancient</span> <span class="op">&lt;-</span> <span class="va">ancient</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">population</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sample_remove</span><span class="op">)</span></span></code></pre></div>
<p>And now we can project our data using the <code>predict</code>
function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">modern_pca</span>,</span>
<span>  new_data <span class="op">=</span> <span class="va">ancient</span>,</span>
<span>  project_method <span class="op">=</span> <span class="st">"least_squares"</span>,</span>
<span>  as_matrix <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here, we use the “least_squares” argument to generate a least squares
projection that will be comparable to the approach used in
<code>smartpca</code> (also implemented in the R package
<code>smartsnp</code>).</p>
<p>Finally, we can tidy up our data ready to plot by converting the
<code>predicted</code> object to a data frame and adding the population
names:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted</span> <span class="op">&lt;-</span> <span class="va">predicted</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="va">ancient</span><span class="op">$</span><span class="va">id</span>,</span>
<span>  population <span class="op">=</span> <span class="va">ancient</span><span class="op">$</span><span class="va">population</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The PCA and predicted individuals can then be visualized using
<code>ggplot2</code> by layering a geom of ancient individuals over
modern individuals, using the same syntax as <code>smartsnp</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">modern_pca_scores</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.fittedPC1</span>, y <span class="op">=</span> <span class="va">.fittedPC2</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">predicted</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.PC1</span>, <span class="va">.PC2</span>, fill <span class="op">=</span> <span class="va">population</span>, shape <span class="op">=</span> <span class="va">population</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">21</span><span class="op">:</span><span class="fl">25</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_label_repel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">predicted</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.PC1</span>, <span class="va">.PC2</span>, label <span class="op">=</span> <span class="va">population</span>, col <span class="op">=</span> <span class="va">population</span><span class="op">)</span>,</span>
<span>    alpha <span class="op">=</span> <span class="fl">0.7</span>, segment.color <span class="op">=</span> <span class="st">"NA"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC 1 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pca_variance</span><span class="op">$</span><span class="va">percent</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%)"</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC 2 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pca_variance</span><span class="op">$</span><span class="va">percent</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%)"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 247 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span></span></code></pre></div>
<div class="float">
<img src="figure/aDNA_pseudohaploids/pca_projected_aDNA-1.png" alt="plot of chunk pca_projected_aDNA"><div class="figcaption">plot of chunk pca_projected_aDNA</div>
</div>
</div>
<div class="section level2">
<h2 id="pairwise-fst">Pairwise Fst<a class="anchor" aria-label="anchor" href="#pairwise-fst"></a>
</h2>
<p>Now let’s replicate the pairwise Fst analysis from Lazaridis et
al. (2016).</p>
<p>First we need to merge the ancient and modern data:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancient_modern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">ho_modern</span>, <span class="va">ancient</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonising loci between two datasets</span></span>
<span><span class="co">#&gt; flip_strand =  FALSE  ; remove_ambiguous =  FALSE </span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: reference </span></span>
<span><span class="co">#&gt; number of SNPs: 548749 reduced to 548749 </span></span>
<span><span class="co">#&gt; ( 0 are ambiguous, of which 0  were removed)</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; dataset: target </span></span>
<span><span class="co">#&gt; number of SNPs: 1233553 reduced to 548749 </span></span>
<span><span class="co">#&gt; ( 0 were flipped to match the reference set)</span></span>
<span><span class="co">#&gt; ( 36301 are ambiguous, of which 36301 were removed)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_tibble saved to /tmp/RtmpptKMxO/gt_merged_42bc4e3e7eb.gt</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpptKMxO/gt_merged_42bc4e3e7eb.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpptKMxO/gt_merged_42bc4e3e7eb.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span>
<span><span class="co">#&gt; to reload the gen_tibble in another session, use:</span></span>
<span><span class="co">#&gt; gt_load('/tmp/RtmpptKMxO/gt_merged_42bc4e3e7eb.gt')</span></span></code></pre></div>
<p>Then, we group the ancient and modern data by population:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancient_modern</span> <span class="op">&lt;-</span> <span class="va">ancient_modern</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></code></pre></div>
<p>As we are calculating Fst between populations, we will remove any
populations that contain only one individual.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancient_modern</span> <span class="op">&lt;-</span> <span class="va">ancient_modern</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>And then we can calculate Fst:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairwise_fst_tidy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pairwise_pop_fst.html">pairwise_pop_fst</a></span><span class="op">(</span><span class="va">ancient_modern</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"Hudson"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"tidy"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can then assign time ranges to the populations based on their
names. This will allow us to group populations by time period when
creating our plot:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairwise_fst_tidy</span> <span class="op">&lt;-</span> <span class="va">pairwise_fst_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>time_range_pop1 <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>    <span class="fu">str_detect</span><span class="op">(</span><span class="va">population_1</span>, <span class="st">"BA"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"Bronze Age"</span>,</span>
<span>    <span class="fu">str_detect</span><span class="op">(</span></span>
<span>      <span class="va">population_1</span>,</span>
<span>      <span class="st">"ChL|_EN|_N|Eneolithic"</span></span>
<span>    <span class="op">)</span> <span class="op">~</span> <span class="st">"European Neolithic to Chalcolithic"</span>,</span>
<span>    <span class="fu">str_detect</span><span class="op">(</span><span class="va">population_1</span>, <span class="st">"HG|Natufian"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"Before European Neolithic"</span>,</span>
<span>    <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"Present"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>time_range_pop2 <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>    <span class="fu">str_detect</span><span class="op">(</span><span class="va">population_2</span>, <span class="st">"BA"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"Bronze Age"</span>,</span>
<span>    <span class="fu">str_detect</span><span class="op">(</span></span>
<span>      <span class="va">population_2</span>,</span>
<span>      <span class="st">"ChL|_EN|_N|Eneolithic"</span></span>
<span>    <span class="op">)</span> <span class="op">~</span> <span class="st">"European Neolithic to Chalcolithic"</span>,</span>
<span>    <span class="fu">str_detect</span><span class="op">(</span><span class="va">population_2</span>, <span class="st">"HG|Natufian"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"Before European Neolithic"</span>,</span>
<span>    <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"Present"</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>After adding the time ranges, we can filter the pairwise Fst data to
only include comparisons between populations from the same time range,
and set these ranges as factor levels to order the plot:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairwise_fst_tidy</span> <span class="op">&lt;-</span> <span class="va">pairwise_fst_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">time_range_pop1</span> <span class="op">==</span> <span class="va">time_range_pop2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>time_range_pop1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">time_range_pop1</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Present"</span>,</span>
<span>    <span class="st">"Bronze Age"</span>,</span>
<span>    <span class="st">"European Neolithic to Chalcolithic"</span>,</span>
<span>    <span class="st">"Before European Neolithic"</span></span>
<span>  <span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we will calculate the median, minimum, and maximum values
for each time range, so that these can be added to the plot:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">medians</span> <span class="op">&lt;-</span> <span class="va">pairwise_fst_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">time_range_pop1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>median_value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">time_range_pop1</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pairwise_fst_tidy</span><span class="op">$</span><span class="va">time_range_pop1</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">min</span> <span class="op">&lt;-</span> <span class="va">pairwise_fst_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">time_range_pop1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>min_value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">max</span> <span class="op">&lt;-</span> <span class="va">pairwise_fst_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">time_range_pop1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>max_value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can create our plot using ggplot2, adding the median, minimum, and
maximum values as lines and text labels:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pairwise_fst_tidy</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, y <span class="op">=</span> <span class="va">time_range_pop1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">medians</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">median_value</span>, y <span class="op">=</span> <span class="va">time_range_pop1</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"red"</span>, linewidth <span class="op">=</span> <span class="fl">1.2</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">min</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">min_value</span>, y <span class="op">=</span> <span class="va">time_range_pop1</span>,</span>
<span>    label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">min_value</span>, <span class="fl">3</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.7</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">max</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">max_value</span>, y <span class="op">=</span> <span class="va">time_range_pop1</span>,</span>
<span>    label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">max_value</span>, <span class="fl">3</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.7</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"F"</span><span class="op">[</span><span class="va">ST</span><span class="op">]</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figure/aDNA_pseudohaploids/pairwise_fst-1.png" alt="plot of chunk pairwise_fst"><div class="figcaption">plot of chunk pairwise_fst</div>
</div>
</div>
<div class="section level2">
<h2 id="f-statistics">F statistics<a class="anchor" aria-label="anchor" href="#f-statistics"></a>
</h2>
<p>To calculate F statistics, <code>tidypopgen</code> integrates with
the R <code>admixtools</code> package (a.k.a. ADMIXTOOLS 2). The
function <code>gt_extract_f2</code> allows users to calculate blocked f2
statistics from a <code>gen_tibble</code> object, operating in the same
way as the <code>extract_f2</code> function of
<code>admixtools</code>.</p>
<p><code>gt_extract_f2</code> can be used as follows:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f2s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_extract_f2.html">gt_extract_f2</a></span><span class="op">(</span><span class="va">ancient_modern</span>,</span>
<span>  outdir <span class="op">=</span> <span class="st">"./data/NearEastPublic/f2"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We could now use any of the functions from <code>admixtools</code>
giving the <code>outdir</code> with the f2 values.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
