<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Benchmark on the HGDP • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Benchmark on the HGDP">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Benchmark on the HGDP</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/dev/vignettes/articles/benchmark_hgdp.Rmd" class="external-link"><code>vignettes/articles/benchmark_hgdp.Rmd</code></a></small>
      <div class="d-none name"><code>benchmark_hgdp.Rmd</code></div>
    </div>

    
    
<p>This vignette is a benchmark of a few key functions in the tidypopgen
package. The Human Genome Diversity Project (HGDP) SNP dataset is used
for this benchmark. This dataset includes 1043 individuals from 51
populations, typed at ~650k loci.</p>
<p>We will run this benchmark on a machine with a GenuineIntel, Intel(R)
Core(TM) Ultra 7 155H, 22 CPU and 31 GiB of RAM. However, we will limit
the number of cores used to 20. Parallelised functions within tidypopgen
use an <code>n_cores</code> argument for the user to set. However, to
prevent a behind-the-scenes inflation of the number of threads used (for
example, in cases where dependency functions may automatically use all
available cores) we need to set up preferences at the beginning of the
session. Specifically, we limit the number of cores used by the
parallelised BLAS library with
<code><a href="https://rdrr.io/pkg/bigparallelr/man/get_blas_ncores.html" class="external-link">bigparallelr::set_blas_ncores()</a></code>, and by the package
data.table with <code><a href="https://rdatatable.gitlab.io/data.table/reference/openmp-utils.html" class="external-link">data.table::setDTthreads()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_cores</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/openmp-utils.html" class="external-link">setDTthreads</a></span><span class="op">(</span><span class="va">n_cores</span><span class="op">)</span></span>
<span><span class="fu">bigparallelr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/get_blas_ncores.html" class="external-link">set_blas_ncores</a></span><span class="op">(</span><span class="va">n_cores</span><span class="op">)</span></span></code></pre></div>
<p>We can now load the necessary libraries:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EvolEcolGroup/tidypopgen" class="external-link">tidypopgen</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: tibble</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>And before running the benchmark, we need to download the relevant
files. We can use the following to download the HGDOP files into a
temporary directory:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">download_url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/15582364/files/hgdp650_id_pop.txt"</span></span>
<span><span class="va">download_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"hgdp650_id_pop.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">download_url</span>, <span class="va">download_path</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span>
<span><span class="va">download_url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/15582364/files/hgdp650.qc.hg19.bed"</span></span>
<span><span class="va">download_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"hgdp650.qc.hg19.bed"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">download_url</span>, <span class="va">download_path</span>, method <span class="op">=</span> <span class="st">"wget"</span><span class="op">)</span></span>
<span><span class="va">download_url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/15582364/files/hgdp650.qc.hg19.bim"</span></span>
<span><span class="va">download_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"hgdp650.qc.hg19.bim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">download_url</span>, <span class="va">download_path</span>, method <span class="op">=</span> <span class="st">"wget"</span><span class="op">)</span></span>
<span><span class="va">download_url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/15582364/files/hgdp650.qc.hg19.fam"</span></span>
<span><span class="va">download_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"hgdp650.qc.hg19.fam"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">download_url</span>, <span class="va">download_path</span>, method <span class="op">=</span> <span class="st">"wget"</span><span class="op">)</span></span></code></pre></div>
<p>In the following vignette, we place these files into the subdirectory
<code>data/hgdp/</code>, and we will use the paths to the bed file
<code>hgdp650.qc.hg19.bed</code> and the metadata file
<code>hgdp650_id_pop.txt</code> to read in our data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_path</span> <span class="op">&lt;-</span> <span class="st">"./data/hgdp/hgdp650.qc.hg19.bed"</span></span>
<span><span class="va">meta_info</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"./data/hgdp/hgdp650_id_pop.txt"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: './data/hgdp/hgdp650_id_pop.txt' does not exist in current working directory ('/home/EJC199/Git/tidypopgen/vignettes/articles').</span></span></code></pre></div>
<div class="section level2">
<h2 id="create-gen_tibble-object">Create gen_tibble object<a class="anchor" aria-label="anchor" href="#create-gen_tibble-object"></a>
</h2>
<p>Our first step is to load the HGDP data into a
<code>gen_tibble</code> object, and add its associated metadata.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span><span class="va">bed_path</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"test_"</span><span class="op">)</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="va">n_cores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>read_plink: 5.3s</p>
<p>Add metadata</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="va">hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>  population <span class="op">=</span> <span class="va">meta_info</span><span class="op">$</span><span class="va">population</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">$</span><span class="va">id</span>, <span class="va">meta_info</span><span class="op">$</span><span class="va">Id</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="va">meta_info</span><span class="op">$</span><span class="va">Region</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">$</span><span class="va">id</span>, <span class="va">meta_info</span><span class="op">$</span><span class="va">Id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `mutate()`:</span></span>
<span><span class="co">#&gt; ℹ In argument: `population = meta_info$population[match(hgdp$id, meta_info$Id)]`.</span></span>
<span><span class="co">#&gt; Caused by error:</span></span>
<span><span class="co">#&gt; ! object 'meta_info' not found</span></span></code></pre></div>
<p>Let’s confirm that we have read all the expected information:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span></span>
<span><span class="co">#&gt; # A gen_tibble: 643733 loci</span></span>
<span><span class="co">#&gt; # A tibble:     1,043 × 2</span></span>
<span><span class="co">#&gt;    id         genotypes</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;vctr_SNP&gt;</span></span>
<span><span class="co">#&gt;  1 HGDP00448          1</span></span>
<span><span class="co">#&gt;  2 HGDP00479          2</span></span>
<span><span class="co">#&gt;  3 HGDP00985          3</span></span>
<span><span class="co">#&gt;  4 HGDP01094          4</span></span>
<span><span class="co">#&gt;  5 HGDP00982          5</span></span>
<span><span class="co">#&gt;  6 HGDP00911          6</span></span>
<span><span class="co">#&gt;  7 HGDP01202          7</span></span>
<span><span class="co">#&gt;  8 HGDP00927          8</span></span>
<span><span class="co">#&gt;  9 HGDP00461          9</span></span>
<span><span class="co">#&gt; 10 HGDP00451         10</span></span>
<span><span class="co">#&gt; # ℹ 1,033 more rows</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loci-report">Loci Report<a class="anchor" aria-label="anchor" href="#loci-report"></a>
</h2>
<p>We can then call <code>qc_report_loci</code>. This function supplies
minor allele frequency, rate of missingness, and a Hardy-Weinberg exact
p-value for each SNP.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loci_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">)</span></span>
<span><span class="co">#&gt; This gen_tibble is not grouped. For Hardy-Weinberg equilibrium, `qc_report_loci()` will assume individuals are part of the same population and HWE test p-values will be calculated across all individuals. If you wish to calculate HWE p-values within populations or groups, please use`group_by()` before calling `qc_report_loci()`.</span></span></code></pre></div>
<p>loci_report: 3.3s The resulting report can be observed using
<code>autoplot</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_report</span>, type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figure/hgdp_benchmark/autoplot_loci_report-1.png" alt="plot of chunk autoplot_loci_report"><div class="figcaption">plot of chunk autoplot_loci_report</div>
</div>
</div>
<div class="section level2">
<h2 id="filter-loci">Filter Loci<a class="anchor" aria-label="anchor" href="#filter-loci"></a>
</h2>
<p>Following this, we filter the loci to only including those with a
minor allele frequency over 0.05, and a missingness rate below 0.05.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep_loci</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">loci_report</span>, <span class="va">loci_report</span><span class="op">$</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">loci_report</span><span class="op">$</span><span class="va">missingness</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="va">hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci.html">select_loci</a></span><span class="op">(</span><span class="va">to_keep_loci</span><span class="op">$</span><span class="va">snp_id</span><span class="op">)</span></span></code></pre></div>
<p>filter_loci: 5.3s</p>
</div>
<div class="section level2">
<h2 id="individual-report">Individual Report<a class="anchor" aria-label="anchor" href="#individual-report"></a>
</h2>
<p>We can then call <code>qc_report_indiv</code> to supply observed
heterozygosity per individual, and rate of missingness per
individual.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">indiv_report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">)</span></span></code></pre></div>
<p>indiv_report: 3.4s</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">indiv_report</span>, type <span class="op">=</span> <span class="st">"scatter"</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figure/hgdp_benchmark/autoplot_indiv_report-1.png" alt="plot of chunk autoplot_indiv_report"><div class="figcaption">plot of chunk autoplot_indiv_report</div>
</div>
</div>
<div class="section level2">
<h2 id="filter-individuals">Filter individuals<a class="anchor" aria-label="anchor" href="#filter-individuals"></a>
</h2>
<p>And we can filter individuals down to only include those with less
than 1% of their genotypes missing.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep_indiv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">indiv_report</span><span class="op">$</span><span class="va">missingness</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="va">hgdp</span><span class="op">[</span><span class="va">to_keep_indiv</span>, <span class="op">]</span></span></code></pre></div>
<p>filter_indiv: 7ms</p>
</div>
<div class="section level2">
<h2 id="update-backingfile">Update backingfile<a class="anchor" aria-label="anchor" href="#update-backingfile"></a>
</h2>
<p>After removing individuals from the dataset, and before imputing, we
need to update the file backing matrix with
<code>gt_update_backingfile</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_update_backingfile.html">gt_update_backingfile</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">)</span></span>
<span><span class="co">#&gt; Genetic distances are not sorted, setting them to zero</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gen_backing files updated, now</span></span>
<span><span class="co">#&gt; using bigSNP file: /tmp/RtmpY30jCS/test_163872732cbd_v2.rds</span></span>
<span><span class="co">#&gt; with backing file: /tmp/RtmpY30jCS/test_163872732cbd_v2.bk</span></span>
<span><span class="co">#&gt; make sure that you do NOT delete those files!</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="impute-data">Impute data<a class="anchor" aria-label="anchor" href="#impute-data"></a>
</h2>
<p>Some functions, such as <code>loci_ld_clump</code> and the
<code>gt_pca</code> functions, require that there is no missingness in
the dataset, so we use <code>gt_impute_simple</code> to impute any
remaining missing genotypes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">hgdp</span>, method <span class="op">=</span> <span class="st">"mode"</span>, n_cores <span class="op">=</span> <span class="va">n_cores</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gt_set_imputed.html">gt_set_imputed</a></span><span class="op">(</span><span class="va">hgdp</span>, <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>impute: 365ms</p>
</div>
<div class="section level2">
<h2 id="ld-clumping">LD clumping<a class="anchor" aria-label="anchor" href="#ld-clumping"></a>
</h2>
<p>LD clumping is then performed to control for linkage
disequilibrium.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgdp</span> <span class="op">&lt;-</span> <span class="va">hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_ld_clump.html">loci_ld_clump</a></span><span class="op">(</span><span class="va">genotypes</span>, thr_r2 <span class="op">=</span> <span class="fl">0.2</span>, n_cores <span class="op">=</span> <span class="va">n_cores</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>ld_clumping: 7.5s</p>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>A principal components analysis can then be computed using the
resulting cleaned and LD clumped dataset.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_pca</span> <span class="op">&lt;-</span> <span class="va">hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>pca: 4s</p>
<p>Plot PCA:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">test_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">hgdp</span><span class="op">$</span><span class="va">region</span>, shape <span class="op">=</span> <span class="va">hgdp</span><span class="op">$</span><span class="va">region</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Unknown or uninitialised column: `region`.</span></span>
<span><span class="co">#&gt; Unknown or uninitialised column: `region`.</span></span></code></pre></div>
<div class="float">
<img src="figure/hgdp_benchmark/PCA-1.png" alt="plot of chunk PCA"><div class="figcaption">plot of chunk PCA</div>
</div>
</div>
<div class="section level2">
<h2 id="dapc">DAPC<a class="anchor" aria-label="anchor" href="#dapc"></a>
</h2>
<p>We can continue with a discriminant analysis of principal components
using <code>gt_dapc</code>, setting 6 groups corresponding to the main
geographic regions covered by the dataset.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_dapc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_dapc.html">gt_dapc</a></span><span class="op">(</span><span class="va">test_pca</span>, pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">hgdp</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Unknown or uninitialised column: `region`.</span></span>
<span><span class="co">#&gt; Error in sweep(x$u, 2, x$d, "*")[, 1:n_pca, drop = FALSE]: only 0's may be mixed with negative subscripts</span></span></code></pre></div>
<p>dapc: 24ms</p>
<p>Plot DAPC:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">test_dapc</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: object 'test_dapc' not found</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-fst">Calculate Fst<a class="anchor" aria-label="anchor" href="#calculate-fst"></a>
</h2>
<p>To examine the differentiation between populations in the global HGDP
set, we calculate pairwise Fst.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grouped_hgdp</span> <span class="op">&lt;-</span> <span class="va">hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `group_by()`:</span></span>
<span><span class="co">#&gt; ! Must group by variables found in `.data`.</span></span>
<span><span class="co">#&gt; ✖ Column `population` is not found.</span></span>
<span><span class="va">pairwise_fsts</span> <span class="op">&lt;-</span> <span class="va">grouped_hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/pairwise_pop_fst.html">pairwise_pop_fst</a></span><span class="op">(</span></span>
<span>  n_cores <span class="op">=</span> <span class="va">n_cores</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"pairwise"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error: object 'grouped_hgdp' not found</span></span></code></pre></div>
<p>pairwise_fst: 129ms</p>
<p>Plot pairwise Fst:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Order by continents</span></span>
<span><span class="va">grouped_hgdp_order</span> <span class="op">&lt;-</span> <span class="va">grouped_hgdp</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">region</span>, <span class="va">population</span><span class="op">)</span></span>
<span><span class="va">regional_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">grouped_hgdp_order</span><span class="op">$</span><span class="va">population</span><span class="op">)</span></span>
<span><span class="va">pairwise_fsts</span> <span class="op">&lt;-</span> <span class="va">pairwise_fsts</span><span class="op">[</span><span class="va">regional_order</span>, <span class="va">regional_order</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">ggheatmap</span><span class="op">(</span><span class="va">pairwise_fsts</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="op">-</span><span class="fl">60</span>, hjust <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span>, hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    fill <span class="op">=</span> <span class="st">"Fst"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="save-in-plink-bed-format">Save in plink bed format<a class="anchor" aria-label="anchor" href="#save-in-plink-bed-format"></a>
</h2>
<p>Finally, we can save the resulting cleaned dataset to a PLINK .bed
file.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gt_as_plink.html">gt_as_plink</a></span><span class="op">(</span><span class="va">hgdp</span>,</span>
<span>  file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"bed"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpY30jCS/file1638575879aa.bed"</span></span></code></pre></div>
<p>plink_save: 485ms</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>Here is a summary of the time taken (in seconds) to perform each step
of the analyses:</p>
<pre><code><span><span class="co">#&gt;            step  time</span></span>
<span><span class="co">#&gt; 1    read_plink  5.29</span></span>
<span><span class="co">#&gt; 2   loci_report  3.29</span></span>
<span><span class="co">#&gt; 3   filter_loci  5.29</span></span>
<span><span class="co">#&gt; 4  indiv_report  3.44</span></span>
<span><span class="co">#&gt; 5  filter_indiv  0.01</span></span>
<span><span class="co">#&gt; 6        impute  0.36</span></span>
<span><span class="co">#&gt; 7   ld_clumping  7.54</span></span>
<span><span class="co">#&gt; 8           pca  3.99</span></span>
<span><span class="co">#&gt; 9          dapc  0.02</span></span>
<span><span class="co">#&gt; 10 pairwise_fst  0.13</span></span>
<span><span class="co">#&gt; 11   plink_save  0.48</span></span>
<span><span class="co">#&gt; 12        Total 29.84</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Eirlys Tysall, Andrea Manica.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
