<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>tidypopgen • tidypopgen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidypopgen">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidypopgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/tidypopgen.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_overview.html">The grammar of population genetics</a></li>
    <li><a class="dropdown-item" href="../articles/a02_qc.html">Quality Control</a></li>
    <li><a class="dropdown-item" href="../articles/a03_example_clustering_and_dapc.html">Population genetic analysis with tidypopgen</a></li>
    <li><a class="dropdown-item" href="../articles/a99_plink_cheatsheet.html">PLINK cheatsheet</a></li>
    <li><a class="dropdown-item" href="../articles/aDNA_pseudohaploids.html">Working with aDNA pseuodhaploid samples</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark_hgdp.html">Benchmark on the HGDP</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EvolEcolGroup/tidypopgen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>tidypopgen</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EvolEcolGroup/tidypopgen/blob/main/vignettes/tidypopgen.Rmd" class="external-link"><code>vignettes/tidypopgen.Rmd</code></a></small>
      <div class="d-none name"><code>tidypopgen.Rmd</code></div>
    </div>

    
    
<p>This vignette gives a short introduction on how to use the
<code>tidypopgen</code> package to perform basic data cleaning and a PCA
on a dataset of European lobsters (<em>Homarus gammarus</em>). The
original data are available at <a href="https://datadryad.org/dataset/doi:10.5061/dryad.2v1kr38" class="external-link uri">https://datadryad.org/dataset/doi:10.5061/dryad.2v1kr38</a>,
but for the purpose of this vignette we will use a subset of the data
stored as a .bed file, available in the
<code>inst/extdata/lobster</code> directory of the package. To install
<code>tidypopgen</code> from r-universe, use:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"tidypopgen"</span>, repos <span class="op">=</span> <span class="st">"https://evolecolgroup.r-universe.dev"</span><span class="op">)</span></span></code></pre></div>
<p>Next, load the <code>tidypopgen</code> package and the
<code>ggplot2</code> package for plotting.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EvolEcolGroup/tidypopgen" class="external-link">tidypopgen</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: tibble</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="creating-a-gen_tibble">Creating a <code>gen_tibble</code><a class="anchor" aria-label="anchor" href="#creating-a-gen_tibble"></a>
</h2>
<p>In <code>tidypopgen</code>, we represent data as a
<code>gen_tibble</code>, a subclass of <code>tibble</code> containing
the columns <code>id</code> and <code>genotypes</code> for each
individual. Genotypes are stored in a compressed format as a File-Backed
Matrix that can be easily accessed by functions in
<code>tidypopgen</code>, therefore the <code>genotypes</code> column is
simply a vector giving the row indices of each individual in the
File-Backed Matrix.</p>
<p>Additionally, if data are loaded from a .bed file with information in
the <code>FID</code> column, this is treated as population information
and is automatically added to the <code>gen_tibble</code> as column
<code>population</code>. As with a normal tibble, this information can
be changed, updated, or removed from the <code>gen_tibble</code> if
needed. <code>tidypopgen</code> can also read data form
<code>packedancestry</code> files and <code>vcf</code>.</p>
<p>Let’s start by creating a <code>gen_tibble</code> from the
<code>lobster.bed</code> file.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_tibble.html">gen_tibble</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/lobster/lobster.bed"</span>, package <span class="op">=</span> <span class="st">"tidypopgen"</span><span class="op">)</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lobsters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A gen_tibble: 79 loci</span></span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble:     6 × 3</span></span></span>
<span><span class="co">##   id    population  genotypes</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Ale04 Ale                 1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Ale05 Ale                 2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Ale06 Ale                 3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Ale08 Ale                 4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Ale13 Ale                 5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Ale15 Ale                 6</span></span></code></pre>
<p>We can see the structure of the <code>gen_tibble</code> above, with
the expected columns. In this case, our .bed file contains population
information corresponding to the sampling site of each lobster in the
dataset.</p>
<p>If we want to take a look at the genotypes of our lobsters, we can
use the <code>show_genotypes</code> function to return a matrix of
genotypes, where rows are individuals and columns are loci. This is a
big table, so we will just look at the first ten loci for the first 5
individuals:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_genotypes.html">show_genotypes</a></span><span class="op">(</span>indiv_indices <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, loci_indices <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></span>
<span><span class="co">## [1,]    0   NA   NA   NA    2   NA   NA   NA    0    NA</span></span>
<span><span class="co">## [2,]    1    0    1    2    1    1    0    1    1     1</span></span>
<span><span class="co">## [3,]   NA    0    0   NA    2    2   NA   NA    0     2</span></span>
<span><span class="co">## [4,]   NA    2    0    2   NA   NA   NA   NA    1    NA</span></span>
<span><span class="co">## [5,]    0   NA    0   NA    2   NA    0   NA   NA    NA</span></span></code></pre>
<p>And, similarly, if we want to see information about the loci we can
use the <code>show_loci</code> function, which returns a tibble with
information about each locus. Again this is a big table, so we will use
<code><a href="https://rdrr.io/r/utils/head.html" class="external-link">head()</a></code> to only look at the first few:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/show_loci.html">show_loci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">##   big_index  name chromosome position genetic_dist allele_ref allele_alt chr_int</span></span>
<span><span class="co">##       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>         1  <span style="text-decoration: underline;">3</span>441 1                 1            1 G          A                1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>         2  <span style="text-decoration: underline;">4</span>173 2                 2            2 C          T                2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>         3  <span style="text-decoration: underline;">6</span>157 3                 3            3 G          C                3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>         4  <span style="text-decoration: underline;">7</span>502 4                 4            4 C          T                4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>         5  <span style="text-decoration: underline;">7</span>892 5                 5            5 A          T                5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>         6  <span style="text-decoration: underline;">9</span>441 7                 7            7 A          G                7</span></span></code></pre>
<p>Now we have a <code>gen_tibble</code> to work with, we can start to
clean the data.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>Let’s start by checking the quality of the data for each individual
lobster in our dataset. We can use the <code>qc_report_indiv</code>
function to generate a report which contains information about
missingness and heterozygosity for each individual.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">indiv_qc_lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/qc_report_indiv.html">qc_report_indiv</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We can take a look at this data using the <code>autoplot</code>
function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">indiv_qc_lobsters</span>, type <span class="op">=</span> <span class="st">"scatter"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-7-1.png" alt="Scatter plot of missingness proportion and observed heterozygosity for each individual" width="700"></p>
<p>We can see that most individuals have low missingness and
heterozygosity, but there are a few individuals missing &gt;20% of their
genotypes. We can remove these individuals using the <code>filter</code>
function, specifying to keep only individuals with missingness under
20%.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/indiv_missingness.html">indiv_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p>Now lets check our loci. We can use the <code>qc_report_loci</code>
function to generate a report of the loci quality. This function will
return another <code>qc_report</code> object, which contains information
about missingness, minor allele frequency, and Hardy-Weinberg
Equilibrium for each locus.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loci_qc_lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## This gen_tibble is not grouped. For Hardy-Weinberg equilibrium, `qc_report_loci()` will assume individuals are part of the same population and HWE test p-values will be calculated across all individuals. If you wish to calculate HWE p-values within populations or groups, please use`group_by()` before calling `qc_report_loci()`.</span></span></code></pre>
<p>Here, we get a message because the <code>qc_report_loci</code>
function calculates Hardy-Weinberg equilibrium assuming that all
individuals are part of a single population. As our dataset contains
multiple lobster populations, we should group our data by population
first:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span>
<span><span class="va">loci_qc_lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/qc_report_loci.html">qc_report_loci</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>That’s better. Now, lets take a look at minor allele frequency for
all loci:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_qc_lobsters</span>, type <span class="op">=</span> <span class="st">"maf"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-11-1.png" alt="Histogram of minor allele frequency" width="700"></p>
<p>And we can see that we don’t have any monomorphic SNPs.</p>
<p>Now let’s look at missingness.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">loci_qc_lobsters</span>, type <span class="op">=</span> <span class="st">"missing"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-12-1.png" alt="Histogram of the proportion of missing data" width="700">
Our data mostly have low missingness, but we can see that some loci have
&gt; 5% missingness across individuals, and we want to remove these
individuals.</p>
<p>In tidypopgen, there are two functions to subset the loci in a
<code>gen_tibble</code> object: <code>select_loci</code> and
<code>select_loci_if</code>. The function <code>select_loci</code>
operates on information about the loci (e.g filtering by chromosome or
by rsID), while <code>select_loci_if</code> operates on the genotypes at
those loci (e.g filtering by minor allele frequency or missingness). In
this case, we want to remove loci with &gt;5% missingness, so we can use
<code>select_loci_if</code> with the <code>loci_missingness</code>
function, operating on the <code>genotypes</code> column of our
<code>gen_tibble</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="va">lobsters</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/select_loci_if.html">select_loci_if</a></span><span class="op">(</span><span class="fu"><a href="../reference/loci_missingness.html">loci_missingness</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>Through our filtering, we have removed a few individuals and loci. We
should now update the file backing matrix to reflect these changes,
using the function <code>gt_update_backingfile</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_update_backingfile.html">gt_update_backingfile</a></span><span class="op">(</span><span class="va">lobsters</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## gen_backing files updated, now</span></span></code></pre>
<pre><code><span><span class="co">## using bigSNP file: /tmp/RtmpNnVbGd/file2d857cd43efd.rds</span></span></code></pre>
<pre><code><span><span class="co">## with backing file: /tmp/RtmpNnVbGd/file2d857cd43efd.bk</span></span></code></pre>
<pre><code><span><span class="co">## make sure that you do NOT delete those files!</span></span></code></pre>
<p>Now our data are clean and the backingfile is updated, we are ready
to create a PCA.</p>
</div>
<div class="section level2">
<h2 id="impute">Impute<a class="anchor" aria-label="anchor" href="#impute"></a>
</h2>
<p>First, we need to impute any remaining missing data using the
<code>gt_impute_simple</code> function.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_impute_simple.html">gt_impute_simple</a></span><span class="op">(</span><span class="va">lobsters</span>, method <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>Then we can run a PCA. There are a number of PCA algorithms, here, we
will use the <code>gt_pca_partialSVD</code> function:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partial_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gt_pca_partialSVD.html">gt_pca_partialSVD</a></span><span class="op">(</span><span class="va">lobsters</span><span class="op">)</span></span></code></pre></div>
<p>And we can create a simple plot using <code>autoplot</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">partial_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-17-1.png" alt="Score plot of individuals across the first and second Principal Components" width="700"></p>
<p>That was easy! The <code>autoplot</code> gives us a quick idea of the
explained variance and rough distribution of the samples, but we need to
see the different populations within our dataset.</p>
<p>For a quick overview, we could add an aesthetic to our plot:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">partial_pca</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">lobsters</span><span class="op">$</span><span class="va">population</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-18-1.png" alt="Score plot of individuals across the first and second Principal Components, with individual samples coloured by population" width="700"></p>
<p>However, if we want to fully customise our plot, we can wrangle the
data directly and use <code>ggplot2</code>.</p>
</div>
<div class="section level2">
<h2 id="plot-with-ggplot2">Plot with ggplot2<a class="anchor" aria-label="anchor" href="#plot-with-ggplot2"></a>
</h2>
<p>For a customised plot, we can extract the information on the scores
of each individual using the <code>augment</code> method for
<code>gt_pca</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/augment.html" class="external-link">augment</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">partial_pca</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span></span></code></pre></div>
<p>Then we can extract the eigenvalues for each principal component with
the <code>tidy</code> function, using the “eigenvalues” argument:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eigenvalues</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">partial_pca</span>, <span class="st">"eigenvalues"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">xlab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Axis 1 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">eigenvalues</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">]</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">" %)"</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ylab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Axis 2 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">eigenvalues</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">]</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">" %)"</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And finally plot:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pcs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.fittedPC1</span>, y <span class="op">=</span> <span class="va">.fittedPC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span>,</span>
<span>    shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">3</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xlab</span>, y <span class="op">=</span> <span class="va">ylab</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Lobster PCA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-21-1.png" alt="Score plot of individuals across the first and second Principal Components, with individual samples coloured by population" width="700"></p>
<p>Or we could create a labelled version of our PCA by determining the
centre of each group to place the labels</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate centre for each population</span></span>
<span><span class="va">centroid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">.fittedPC1</span>, <span class="va">.fittedPC2</span>, <span class="va">.fittedPC3</span><span class="op">)</span> <span class="op">~</span> <span class="va">population</span>,</span>
<span>  data <span class="op">=</span> <span class="va">pcs</span>, FUN <span class="op">=</span> <span class="va">mean</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add these coordinates to our augmented pca object</span></span>
<span><span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">pcs</span>, <span class="va">centroid</span>, by <span class="op">=</span> <span class="st">"population"</span>, suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">".cen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>And then add labels to the plot:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pcs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.fittedPC1</span>, y <span class="op">=</span> <span class="va">.fittedPC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xend <span class="op">=</span> <span class="va">.fittedPC1.cen</span>, yend <span class="op">=</span> <span class="va">.fittedPC2.cen</span><span class="op">)</span>,</span>
<span>    show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span>,</span>
<span>    shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">3</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">centroid</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">population</span>, fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">4</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_fill_distruct.html">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xlab</span>, y <span class="op">=</span> <span class="va">ylab</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Lobster PCA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidypopgen_files/figure-html/unnamed-chunk-23-1.png" alt="Score plot of individuals across the first and second Principal Components. Individual samples are coloured by population, the centroid of points for each group is labelled with the population name, and lines are drawn from the centroid to each point." width="700"></p>
<p>That’s it! There are more functions to run different types of
analyses (DAPC, ADMIXTURE, f statistics with admixtools2, etc.); each
resulting object has <code>autoplot</code>, <code>tidy</code>, and
<code>augment</code> to explore the results and integrated into the
information from the <code>gen_tibble</code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evie Carter, Andrea Manica, Eirlys Tysall.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
