# A function to read geno packedancestrymap files
gen_tibble_packedancestry <- function(x, ...,
                                      valid_alleles = c("A", "T", "C", "G"),
                                      missing_alleles = c("0", "."),
                                      chunk_size = NULL,
                                      backingfile = NULL, quiet = FALSE) {
  if (is.null(chunk_size)) {
    chunk_size <- 100
  }

  # Substitute .ped with .map
  map_file <- sub("\\.geno$", ".snp", x)
  if (!file.exists(map_file)) {
    stop("snp file ", map_file, " does not exist")
  }
  ind_file <- sub("\\.geno$", ".ind", x)
  if (!file.exists(ind_file)) {
    stop("ind file ", ind_file, " does not exist")
  }

  if (is.null(backingfile)) {
    backingfile <- sub("\\.geno$", "", x)
  }
  if (file_ext(backingfile) == "bk") {
    backingfile <- bigstatsr::sub_bk(backingfile, "")
  }

  # read the packed ancestry in chunks
  # count the individuals
  indiv_table <- utils::read.table(ind_file, header = FALSE)
  names(indiv_table) <- c("id", "sex", "population")
  no_individuals <- nrow(indiv_table)

  # count the loci
  loci_table <- utils::read.table(map_file, header = FALSE)
  names(loci_table) <- c(
    "name", "chromosome", "genetic_dist", "position",
    "allele_ref", "allele_alt"
  )
  loci_table$allele_alt[loci_table$allele_alt == "X"] <- NA
  no_variants <- nrow(loci_table)

  # create a matrix to store the data
  file_backed_matrix <- bigstatsr::FBM.code256(
    nrow = no_individuals,
    ncol = no_variants,
    code = bigsnpr::CODE_012,
    backingfile = backingfile
  )


  # set up chunks
  chunks <- split(
    1:no_individuals,
    ceiling(seq_along(1:no_individuals) / chunk_size)
  )

  for (i in chunks) {
    res <- admixtools::read_packedancestrymap(sub("\\.geno$", "", x),
      transpose = TRUE,
      inds = indiv_table$id[i],
      ..., verbose = !quiet
    )
    res$geno[is.na(res$geno)] <- 3
    # now insert the genotypes in the FBM
    file_backed_matrix[
      i,
    ] <- res$geno
  }

  # save the fbm
  file_backed_matrix$save()

  # convert the info from the indiv and loci table
  fam <- tibble(
    family.ID = indiv_table$population,
    sample.ID = indiv_table$id,
    paternal.ID = 0,
    maternal.ID = 0,
    sex = 0, # this needs fixing!
    affection = 0,
    ploidy = 2
  )

  map <- tibble(
    chromosome = loci_table$chromosome,
    marker.ID = loci_table$name,
    genetic.dist = loci_table$genetic_dist,
    physical.pos = loci_table$position,
    allele1 = loci_table$allele_ref,
    allele2 = loci_table$allele_alt
  )

  bigsnp_obj <- structure(list(
    genotypes = file_backed_matrix,
    fam = fam,
    map = map
  ), class = "bigSNP")

  bigsnp_obj <- bigsnpr::snp_save(bigsnp_obj)

  gen_tibble(bigsnp_obj$genotypes$rds,
    valid_alleles = valid_alleles,
    missing_alleles = missing_alleles,
    backingfile = backingfile,
    quiet = TRUE
  ) # silence it as output will be generated by the parent function
}
